
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Literature cited &#8212; fwdpy11 manual</title>
    
  <link rel="stylesheet" href="../_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.40e2e510f6b7d1648584402491bb10fe.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.d3f166471bb80abb5163.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.d31b09fe5c1d09cb49b26a786de4a05d.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Upgrade path" href="upgrade_path.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  
  <h1 class="site-logo" id="site-title">fwdpy11 manual</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../intro.html">
   Introduction
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Installation and setup
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/userenv.html">
   Setting up user environments for installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/deploymenttools.html">
   Deployment tools
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Short vignettes - operations on populations
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/initpops_vignette.html">
   Initializing a population
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/poptour_vignette.html">
   Properties of populations
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Short vignettes - setting up the genetics
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/genetics_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/geneticmaps_vignette.html">
   Genetic maps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/des_vignette.html">
   Distributions of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/mutationdominance_vignette.html">
   Mutations with variable dominance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmuts_vignette.html">
   Neutral mutations during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_fitness.html">
   Genetic values - mutations with direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_traits.html">
   Genetic values - simulating phenotypes/traits.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_fitness.html">
   Working example - parameters for simulating direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_trait.html">
   Working example - parameters for the simulation of a trait
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Short vignettes - simulation with tree sequence recording
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/evolvets_vignette.html">
   Evolving with tree sequence recording
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/recorders_vignette.html">
   Recording data during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmutsafter_vignette.html">
   Adding neutral mutations to a population with existing ancestry
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Short vignettes - demographic modeling
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demes_vignette.html">
   Using
   <code class="docutils literal notranslate">
    <span class="pre">
     demes
    </span>
   </code>
   to specify demographic models.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_debugger_vignette.html">
   Debugging demographic models
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Short vignettes - exporting data to tskit
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/tskitconvert_vignette.html">
   Converting data to tskit format
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Longer vignettes - complete examples
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/bgs_vignette.html">
   Background selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/gss_vignette.html">
   Gaussian stabilizing selection with an optimum shift
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker.html">
   Tracking mutations over time
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker_sqlite.html">
   Tracking mutations over time using
   <code class="docutils literal notranslate">
    <span class="pre">
     sqlite3
    </span>
   </code>
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Concepts
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/definitions.html">
   Definitions
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Tutorials
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/tutorial.html">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/advancedtopics.html">
   Advanced topics
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Objects and concepts
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/softselection.html">
   Soft selection with discrete demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/mvdes.html">
   Different effect sizes of mutations in different demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/tsoverview.html">
   Conceptual overview of tree sequence recording
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/tstypes.html">
   Data structures related to tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/tablefs.html">
   Calculating frequency spectra from tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/recorders.html">
   Time series analysis
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Technical details
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/genetic_values.html">
   Technical overview of genetic value calculations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/writingplugins.html">
   Writing “plugins” with C++
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Data types and fuctions
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/genetic_values.html">
   Genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/datamatrix.html">
   DataMatrix and StateMatrix
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/demographic_models.html">
   The demography debugger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/demographic_models.html#module-fwdpy11.demographic_models">
   Pre-computed demographic models
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/functions.html">
   Function documentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/gslrandom.html">
   Random number distributions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/gvaluenoise.html">
   Adding random noise to genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/gvalue_to_fitness.html">
   Mapping genetic values to fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/model_params.html">
   Model parameters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/regiontypes.html">
   Setting up genomic intervals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/tskit_tools.html">
   Module
   <code class="docutils literal notranslate">
    <span class="pre">
     fwdpy11.tskit_tools
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/types.html">
   Data types related to simulated populations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/types.html#types-related-to-tree-sequence-recording">
   Types related to tree sequence recording
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/types.html#types-related-to-discrete-demographic-events">
   Types related to discrete demographic events
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pages/types.html#miscellaneous-types">
   Miscellaneous types
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Examples
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/gss_divergent_optima.html">
   Adaptation to different optima with multivariate distribution of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/IM.html">
   Isolation-with-migration (IM)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/localadaptation.html">
   Local adaptation of a quantitative trait
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/migtest.html">
   Symmetric migration between two demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/pysnowdrift.html">
   Snowdrift model in python
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/recapitation.html">
   Finishing a simulation with a tree sequence from msprime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/recorder.html">
   Example of time series analysis implemented in C++
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Miscellany
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="changelog.html">
   Changelog
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deprecated.html">
   Deprecated features
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="developersguide.html">
   Developer’s guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pubs.html">
   Publications using fwdpy11
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="upgrade_path.html">
   Upgrade path
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Literature cited
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/misc/bibliography.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/molpopgen/fwdpy11"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/molpopgen/fwdpy11/issues/new?title=Issue%20on%20page%20%2Fmisc/bibliography.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="literature-cited">
<h1>Literature cited<a class="headerlink" href="#literature-cited" title="Permalink to this headline">¶</a></h1>
<p id="id1"><dl class="citation">
<dt class="label" id="id22"><span class="brackets">Burger00</span></dt>
<dd><p>R Bürger. <em>The Mathematical Theory of Selection, Recombination, and Mutation</em>. Wiley, November 2000. ISBN 9780471986539.</p>
</dd>
<dt class="label" id="id11"><span class="brackets">Chr74</span></dt>
<dd><p>Freddy Bugge Christiansen. Sufficient conditions for protected polymorphism in a subdivided population. <em>Am. Nat.</em>, 108(960):157–166, 1974. <a class="reference external" href="https://doi.org/10.1086/282896">doi:10.1086/282896</a>.</p>
</dd>
<dt class="label" id="id12"><span class="brackets">Chr75</span></dt>
<dd><p>Freddy Bugge Christiansen. Hard and soft selection in a subdivided population. <em>Am. Nat.</em>, 109(965):11–16, January 1975. <a class="reference external" href="https://doi.org/10.1086/282970">doi:10.1086/282970</a>.</p>
</dd>
<dt class="label" id="id17"><span class="brackets">DHK04</span></dt>
<dd><p>Michael Doebeli, Christoph Hauert, and Timothy Killingback. The evolutionary origin of cooperators and defectors. <em>Science</em>, 306(5697):859–862, October 2004. <a class="reference external" href="https://doi.org/10.1126/science.1101456">doi:10.1126/science.1101456</a>.</p>
</dd>
<dt class="label" id="id23"><span class="brackets">EW10</span></dt>
<dd><p>Adam Eyre-Walker. Evolution in health and medicine sackler colloquium: genetic architecture of a complex trait and its implications for fitness and genome-wide association studies. <em>Proc. Natl. Acad. Sci. U. S. A.</em>, 107 Suppl 1:1752–1756, January 2010. <a class="reference external" href="https://doi.org/10.1073/pnas.0906182107">doi:10.1073/pnas.0906182107</a>.</p>
</dd>
<dt class="label" id="id15"><span class="brackets">Fel76</span></dt>
<dd><p>J Felsenstein. The theoretical population genetics of variable selection and migration. <em>Annu. Rev. Genet.</em>, 10:253–280, 1976. <a class="reference external" href="https://doi.org/10.1146/annurev.ge.10.120176.001345">doi:10.1146/annurev.ge.10.120176.001345</a>.</p>
</dd>
<dt class="label" id="id24"><span class="brackets">GHWB09</span></dt>
<dd><p>Ryan N Gutenkunst, Ryan D Hernandez, Scott H Williamson, and Carlos D Bustamante. Inferring the joint demographic history of multiple populations from multidimensional SNP frequency data. <em>PLoS Genet.</em>, 5(10):e1000695, October 2009. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1000695">doi:10.1371/journal.pgen.1000695</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets">HGKothers19</span></dt>
<dd><p>B C Haller, J Galloway, J Kelleher, and others. Tree‐sequence recording in SLiM opens new horizons for forward‐time simulation of whole genomes. <em>Mol. Ecol.</em>, 2019. <a class="reference external" href="https://doi.org/10.1111/1755-0998.12968">doi:10.1111/1755-0998.12968</a>.</p>
</dd>
<dt class="label" id="id26"><span class="brackets">Her08</span></dt>
<dd><p>Ryan D Hernandez. A flexible forward simulator for populations subject to selection and demography. <em>Bioinformatics</em>, 24(23):2786–2787, December 2008. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btn522">doi:10.1093/bioinformatics/btn522</a>.</p>
</dd>
<dt class="label" id="id25"><span class="brackets">HK95</span></dt>
<dd><p>R R Hudson and N L Kaplan. Deleterious background selection with recombination. <em>Genetics</em>, 141(4):1605–1617, December 1995.</p>
</dd>
<dt class="label" id="id27"><span class="brackets">Hud02</span></dt>
<dd><p>Richard R Hudson. Generating samples under a Wright–Fisher neutral model of genetic variation. <em>Bioinformatics</em>, 18:337–338, 2002.</p>
</dd>
<dt class="label" id="id16"><span class="brackets">JB05</span></dt>
<dd><p>Toby Johnson and Nick Barton. Theoretical models of selection and mutation on quantitative traits. <em>Philos. Trans. R. Soc. Lond. B Biol. Sci.</em>, 360(1459):1411–1425, July 2005. <a class="reference external" href="https://doi.org/10.1098/rstb.2005.1667">doi:10.1098/rstb.2005.1667</a>.</p>
</dd>
<dt class="label" id="id8"><span class="brackets">JLRG17</span></dt>
<dd><p>Julien Jouganous, Will Long, Aaron P Ragsdale, and Simon Gravel. Inferring the joint demographic history of multiple populations: beyond the diffusion approximation. <em>Genetics</em>, 206(3):1549–1567, July 2017. <a class="reference external" href="https://doi.org/10.1534/genetics.117.200493">doi:10.1534/genetics.117.200493</a>.</p>
</dd>
<dt class="label" id="id21"><span class="brackets">KHL89</span></dt>
<dd><p>N L Kaplan, R R Hudson, and C H Langley. The “hitchhiking effect” revisited. <em>Genetics</em>, 123(4):887–899, December 1989.</p>
</dd>
<dt class="label" id="id28"><span class="brackets">KH88</span></dt>
<dd><p>P D Keightley and W G Hill. Quantitative genetic variability maintained by mutation-stabilizing selection balance in finite populations. <em>Genet. Res.</em>, 52(1):33–43, August 1988.</p>
</dd>
<dt class="label" id="id29"><span class="brackets">KH90</span></dt>
<dd><p>Peter D Keightley and William G Hill. Variation maintained in quantitative traits with Mutation-Selection balance: pleiotropic Side-Effects on fitness traits. <em>Proceedings of the Royal Society of London B: Biological Sciences</em>, 242(1304):95–100, November 1990. <a class="reference external" href="https://doi.org/10.1098/rspb.1990.0110">doi:10.1098/rspb.1990.0110</a>.</p>
</dd>
<dt class="label" id="id19"><span class="brackets">KEM16</span></dt>
<dd><p>Jerome Kelleher, Alison M Etheridge, and Gilean McVean. Efficient coalescent simulation and genealogical analysis for large sample sizes. <em>PLoS Comput. Biol.</em>, 12(5):e1004842, May 2016. <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1004842">doi:10.1371/journal.pcbi.1004842</a>.</p>
</dd>
<dt class="label" id="id10"><span class="brackets">KTAR18</span></dt>
<dd><p>Jerome Kelleher, Kevin R Thornton, Jaime Ashander, and Peter L Ralph. Efficient pedigree recording for fast population genetics simulation. <em>PLoS Comput. Biol.</em>, 14(11):e1006581, November 2018. <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1006581">doi:10.1371/journal.pcbi.1006581</a>.</p>
</dd>
<dt class="label" id="id18"><span class="brackets">KS16</span></dt>
<dd><p>Andrew D Kern and Daniel R Schrider. Discoal: flexible coalescent simulations with selection. <em>Bioinformatics</em>, 32(24):3839–3841, December 2016. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btw556">doi:10.1093/bioinformatics/btw556</a>.</p>
</dd>
<dt class="label" id="id14"><span class="brackets">Lev53</span></dt>
<dd><p>Howard Levene. Genetic equilibrium when more than one ecological niche is available. <em>Am. Nat.</em>, 87(836):331–333, September 1953. <a class="reference external" href="https://doi.org/10.1086/281792">doi:10.1086/281792</a>.</p>
</dd>
<dt class="label" id="id6"><span class="brackets">NKR+20</span></dt>
<dd><p>Dominic Nelson, Jerome Kelleher, Aaron P Ragsdale, Claudia Moreau, Gil McVean, and Simon Gravel. Accounting for long-range correlations in genome-wide simulations of large cohorts. <em>PLoS Genet.</em>, 16(5):e1008619, May 2020. <a class="reference external" href="https://doi.org/10.1371/journal.pgen.1008619">doi:10.1371/journal.pgen.1008619</a>.</p>
</dd>
<dt class="label" id="id2"><span class="brackets">RNGK20</span></dt>
<dd><p>Aaron P Ragsdale, Dominic Nelson, Simon Gravel, and Jerome Kelleher. Lessons learned from bugs in models of human history. <em>Am. J. Hum. Genet.</em>, 107(4):583–588, October 2020. <a class="reference external" href="https://doi.org/10.1016/j.ajhg.2020.08.017">doi:10.1016/j.ajhg.2020.08.017</a>.</p>
</dd>
<dt class="label" id="id5"><span class="brackets">RTK20</span></dt>
<dd><p>Peter Ralph, Kevin Thornton, and Jerome Kelleher. Efficiently summarizing relationships in large samples: a general duality between statistics of genealogies and genomes. <em>Genetics</em>, 215(3):779–797, July 2020. <a class="reference external" href="https://doi.org/10.1534/genetics.120.303253">doi:10.1534/genetics.120.303253</a>.</p>
</dd>
<dt class="label" id="id20"><span class="brackets">TBOConnor+12</span></dt>
<dd><p>Jacob A Tennessen, Abigail W Bigham, Timothy D O'Connor, Wenqing Fu, Eimear E Kenny, Simon Gravel, Sean McGee, Ron Do, Xiaoming Liu, Goo Jun, Hyun Min Kang, Daniel Jordan, Suzanne M Leal, Stacey Gabriel, Mark J Rieder, Goncalo Abecasis, David Altshuler, Deborah A Nickerson, Eric Boerwinkle, Shamil Sunyaev, Carlos D Bustamante, Michael J Bamshad, Joshua M Akey, Broad GO, Seattle GO, and NHLBI Exome Sequencing Project. Evolution and functional impact of rare coding variation from deep sequencing of human exomes. <em>Science</em>, 337(6090):64–69, July 2012. <a class="reference external" href="https://doi.org/10.1126/science.1219240">doi:10.1126/science.1219240</a>.</p>
</dd>
<dt class="label" id="id9"><span class="brackets">Tho19</span></dt>
<dd><p>Kevin R Thornton. Polygenic adaptation to an environmental shift: temporal dynamics of variation under gaussian stabilizing selection and additive effects on a single trait. <em>Genetics</em>, 213(4):1513–1530, December 2019. <a class="reference external" href="https://doi.org/10.1534/genetics.119.302662">doi:10.1534/genetics.119.302662</a>.</p>
</dd>
<dt class="label" id="id3"><span class="brackets">WL18</span></dt>
<dd><p>Bruce Walsh and Michael Lynch. <em>Evolution and Selection of Quantitative Traits</em>. Oxford University Press, August 2018. ISBN 9780198830870.</p>
</dd>
<dt class="label" id="id4"><span class="brackets">XKS00</span></dt>
<dd><p>Peter Xue-Kun Song. Multivariate dispersion models generated from gaussian copula. <em>Scand. Stat. Theory Appl.</em>, 27(2):305–320, June 2000. <a class="reference external" href="https://doi.org/10.1111/1467-9469.00191">doi:10.1111/1467-9469.00191</a>.</p>
</dd>
</dl>
</p>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./misc"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="upgrade_path.html" title="previous page">Upgrade path</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Kevin Thornton<br/>
        
            &copy; Copyright 2020.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.d3f166471bb80abb5163.js"></script>


    
  </body>
</html>