

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Advanced topics &#8212; fwdpy11 manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'pages/advancedtopics';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Different effect sizes of mutations in different demes" href="mvdes.html" />
    <link rel="prev" title="Definitions" href="definitions.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    <p class="title logo__title">fwdpy11 manual</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Introduction
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Installation and setup</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="userenv.html">Setting up user environments for installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="deploymenttools.html">Deployment tools</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - operations on populations</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/initpops_vignette.html">Initializing a population</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/poptour_vignette.html">Properties of populations</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - setting up the genetics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/genetics_vignettes_intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/geneticmaps_vignette.html">Genetic maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/des_vignette.html">Distributions of effect sizes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/mutationdominance_vignette.html">Mutations with variable dominance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/neutralmuts_vignette.html">Neutral mutations during a simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/gvalues_fitness.html">Genetic values - mutations with direct effects on fitness</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/gvalues_traits.html">Genetic values - simulating phenotypes/traits.</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/workingexample_fitness.html">Working example - parameters for simulating direct effects on fitness</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/workingexample_trait.html">Working example - parameters for the simulation of a trait</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - simulation with tree sequence recording</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/evolvets_vignette.html">Evolving with tree sequence recording</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/recorders_vignette.html">Recording data during a simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/recapitation.html">Finishing a simulation with msprime</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/neutralmutsafter_vignette.html">Adding neutral mutations to a population with existing ancestry</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/import_mutations_from_tskit.html">Importing mutations from tskit</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - demographic modeling</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/demography_vignettes_intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/demes_vignette.html">Using <code class="docutils literal notranslate"><span class="pre">demes</span></code> to specify demographic models.</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - varying fitness conditions across demes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/varying_fitness_conditions_across_demes_intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/multivariate_gaussian_effect_sizes_across_demes.html">The multivariate Gaussian distribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/multivariate_lognormal_effect_sizes.html">The multivariate lognormal distribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/custom_multivarate_distributions.html">“Custom” multivariate distributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/special_considerations_demes_models.html">Special considerations for demes models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/genetic_values_multi_deme_models_advanced.html">Advanced topics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - conditional models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/trackmutationfates.html">Tracking user-specified new mutations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/selective_sweep.html">Selective sweeps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/incomplete_sweep.html">Incomplete selective sweeps</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Longer vignettes - exporting data to tskit</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/tskitconvert_vignette.html">Converting data to tskit format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/tskit_metadata_vignette.html">Working with data in tskit format</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Longer vignettes - complete examples</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/bgs_vignette.html">Background selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/gss_vignette.html">Gaussian stabilizing selection with an optimum shift</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/pyfreqtracker.html">Tracking mutations over time</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/pyfreqtracker_sqlite.html">Tracking mutations over time using <code class="docutils literal notranslate"><span class="pre">sqlite3</span></code></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="definitions.html">Definitions</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Advanced topics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Objects and concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="mvdes.html">Different effect sizes of mutations in different demes</a></li>
<li class="toctree-l1"><a class="reference internal" href="tsoverview.html">Conceptual overview of tree sequence recording</a></li>
<li class="toctree-l1"><a class="reference internal" href="tstypes.html">Data structures related to tree sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="tablefs.html">Calculating frequency spectra from tree sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="recorders.html">Time series analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Technical details</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../technical/demes.html">Details of <code class="docutils literal notranslate"><span class="pre">demes</span></code> integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../technical/genetic_values.html">Technical overview of genetic value calculations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../technical/writingplugins.html">Writing “plugins” with C++</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data types and fuctions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_values.html">Genetic values</a></li>
<li class="toctree-l1"><a class="reference internal" href="datamatrix.html">DataMatrix and StateMatrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="functions.html">Function documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="gslrandom.html">Random number distributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="gvaluenoise.html">Adding random noise to genetic values</a></li>
<li class="toctree-l1"><a class="reference internal" href="gvalue_to_fitness.html">Mapping genetic values to fitness</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_params.html">Model parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="regiontypes.html">Setting up genomic intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="tskit_tools.html">Module <code class="docutils literal notranslate"><span class="pre">fwdpy11.tskit_tools</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="conditional_models.html">Module <code class="docutils literal notranslate"><span class="pre">fwdpy11.conditional_models</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="types.html">Data types related to simulated populations</a></li>



</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../examples/gss_divergent_optima.html">Adaptation to different optima with multivariate distribution of effect sizes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/pysnowdrift.html">Snowdrift model in python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/recorder.html">Example of time series analysis implemented in C++</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellany</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../misc/changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/deprecated.html">Deprecated features</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/developersguide.html">Developer’s guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/pubs.html">Publications using fwdpy11</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/upgrade_path.html">Upgrade path</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/bibliography.html">Literature cited</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/molpopgen/fwdpy11" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/molpopgen/fwdpy11/issues/new?title=Issue%20on%20page%20%2Fpages/advancedtopics.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/pages/advancedtopics.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Advanced topics</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#executing-multiple-replicates-in-parallel-using-concurrent-futures">Executing multiple replicates in parallel using <code class="docutils literal notranslate"><span class="pre">concurrent.futures</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-long-to-run-the-simulation">How long to run the simulation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-eyre-walker-model-of-complex-traits">The “Eyre-Walker” model of complex traits</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#writing-genetic-value-classes-in-python">Writing genetic value classes in Python</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#concepts">Concepts</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-dimensionality-of-a-model">The “dimensionality” of a model</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#supporting-time-dependent-models">Supporting time-dependent models</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#performance">Performance</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-issues">Technical issues</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#abstract-base-classes">Abstract base classes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#initializing-the-super-class">Initializing the super class</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#attrs"><code class="docutils literal notranslate"><span class="pre">attrs</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#custom-decorators">Custom decorators</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-value-to-fitness-maps">Genetic value to fitness maps</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-data-type">The data type</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-value-noise">Genetic value “noise”</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">The data type</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#new-genetic-value-models">New genetic value models</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-abstract-base-class">The abstract base class</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#outline-of-user-defined-classes">Outline of user-defined classes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-requirements">Method requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-strictly-additive-effects-on-a-trait">Example: strictly additive effects on a trait</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#more-complex-scenarios-such-as-social-interactions">More complex scenarios (such as social interactions)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-data-types">The data types</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subtleties-of-starting-finishing-tree-sequences-using-msprime">Subtleties of starting/finishing tree sequences using <code class="docutils literal notranslate"><span class="pre">msprime</span></code></a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="advanced-topics">
<span id="advancedtopics"></span><h1>Advanced topics<a class="headerlink" href="#advanced-topics" title="Permalink to this heading">#</a></h1>
<section id="executing-multiple-replicates-in-parallel-using-concurrent-futures">
<h2>Executing multiple replicates in parallel using <code class="docutils literal notranslate"><span class="pre">concurrent.futures</span></code><a class="headerlink" href="#executing-multiple-replicates-in-parallel-using-concurrent-futures" title="Permalink to this heading">#</a></h2>
<p>You have many options when it comes to running many replicates of a model.
For example, you could write a single Python script plus a shell script
that your local cluster understands, sending your simulation to hundreds
of compute nodes.</p>
<p>This section introduces <a class="reference external" href="https://docs.python.org/3/library/concurrent.futures.html#module-concurrent.futures" title="(in Python v3.11)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">concurrent.futures</span></code></a> as a method to execute
many replicates in separate Python <em>processes</em> using a single script.</p>
<p>Fully worked-out examples of varying complexity can be found here:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../long_vignettes/bgs_vignette.html#bgs-vignette"><span class="std std-ref">Background selection</span></a></p></li>
<li><p><a class="reference internal" href="../short_vignettes/initpops_vignette.html#precapitation"><span class="std std-ref">Initializing with ancestry from msprime</span></a></p></li>
<li><p><a class="reference internal" href="../short_vignettes/recapitation.html#recapitation"><span class="std std-ref">Finishing a simulation with msprime</span></a></p></li>
</ul>
<p>A common set of idioms used by these examples are:</p>
<ul class="simple">
<li><p>The function to run the simulation typically does not return
the instance of <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a>.  Such
a return requires pickling, which is very expensive.  If you
need to store these population, write them to files rather
than returning them.</p></li>
<li><p>If you do return something to the main “collector”
process, keep it simple!</p></li>
<li><p>Random number seeds are set in <code class="docutils literal notranslate"><span class="pre">__main__</span></code> using
<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html#numpy.random.randint" title="(in NumPy v1.25)"><code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.random.randint()</span></code></a> after calling
<a class="reference external" href="https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html#numpy.random.seed" title="(in NumPy v1.25)"><code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.random.seed()</span></code></a> using a user-provided seed.</p></li>
</ul>
</section>
<section id="how-long-to-run-the-simulation">
<span id="howlongtorun"></span><h2>How long to run the simulation<a class="headerlink" href="#how-long-to-run-the-simulation" title="Permalink to this heading">#</a></h2>
<p>Deciding how long to run a simulation depends on whether or not
a “burn in” phase is required to get the model to statistical
equilibrium.  For example, if you want to get the steady-state
properties of a model with many selected mutations and linkage,
you will need to burn in.  If you want to know how the short term
dynamics of the neutral site-frequency spectrum are affected by suddenly
introducing strongly-beneficial mutations, then you may not need to
burn in.</p>
<p>So, then, how long to burn in?  Clearly, the answer should be “the
least that you have to”!  As always, there are a few different
things to consider:</p>
<ul class="simple">
<li><p>You want the <em>distribution</em> of final outcomes to be “correct”.
Theory often gives us expressions for expectations and
sometimes for variances of things we care about, like
the number of mutations in a sample. However, comparing
distributions will often require a statistical analysis
comparing the output of independent simulations.
For example, for a neutral model, the distribution of the
number of mutations under an infinitely-many sites model
for a small sample should match <code class="docutils literal notranslate"><span class="pre">msprime</span></code> very well.
For a steady-state model of recurrent hitch-hiking
<span id="id1">[<a class="reference internal" href="../misc/bibliography.html#id22" title="N L Kaplan, R R Hudson, and C H Langley. The “hitchhiking effect” revisited. Genetics, 123(4):887–899, December 1989.">KHL89</a>]</span>, results from small samples
should match coalescent simulations such as <span id="id2">[<a class="reference internal" href="../misc/bibliography.html#id19" title="Andrew D Kern and Daniel R Schrider. Discoal: flexible coalescent simulations with selection. Bioinformatics, 32(24):3839–3841, December 2016.">KS16</a>]</span>.
Clearly, all of the different methods
should match analytical results where possible.</p></li>
<li><p>You probably want all of your trees to be fully coalesced
to a single common ancestor.  The time back to a final
ancestor has a very long tail.  In models with recombination,
it is not uncommon to have a tree or three with multiple roots
at the end of a simulation.</p></li>
</ul>
<p>For the first point, a burn-in of <code class="docutils literal notranslate"><span class="pre">10N</span></code> generations or so
is usually sufficient. Before we simulated with tree sequence
recording, we simulated entire “genomes” (neutral mutations and all),
which was rather slow.  (Here “we” is the field in general.)  We then
(hopefully!) compared our results to something like <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.  The
agreement was really good, so one answer is “about <code class="docutils literal notranslate"><span class="pre">10N</span></code>.  I’m being
vague about <code class="docutils literal notranslate"><span class="pre">N</span></code> here–typically, it would be equal to the starting effective
population size of your population (<em>e.g.</em> in the absence of selection).</p>
<p>With tree sequence recording, we run into the “uncoalesced marginal
trees” issue mentioned above.  We could just simulate much longer
to get rid of this problem, but it is simply easier to start
with a tree sequence from <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.</p>
</section>
<section id="the-eyre-walker-model-of-complex-traits">
<span id="eyre-walker"></span><h2>The “Eyre-Walker” model of complex traits<a class="headerlink" href="#the-eyre-walker-model-of-complex-traits" title="Permalink to this heading">#</a></h2>
<p><span id="id3">[<a class="reference internal" href="../misc/bibliography.html#id24" title="Adam Eyre-Walker. Evolution in health and medicine sackler colloquium: genetic architecture of a complex trait and its implications for fitness and genome-wide association studies. Proc. Natl. Acad. Sci. U. S. A., 107 Suppl 1:1752–1756, January 2010.">EW10</a>]</span> describes a model relating mutations with
fitness effect <span class="math notranslate nohighlight">\(S = |2Ns|\)</span> to <span class="math notranslate nohighlight">\(z\)</span>, their effect on
a phenotype/trait according to <span class="math notranslate nohighlight">\(z = \delta S^\tau(1+\epsilon)\)</span>.
Here, <span class="math notranslate nohighlight">\(\epsilon\)</span> is a draw from a Gaussian distribution with mean zero,
<span class="math notranslate nohighlight">\(\delta\)</span> is <span class="math notranslate nohighlight">\(1\)</span> or <span class="math notranslate nohighlight">\(-1\)</span> with equal probability, and
<span class="math notranslate nohighlight">\(\tau\)</span> “tunes” the correlation between fitness effect and effect on the trait.</p>
<p>Implementing this model is quite straightforward, as the trait values do not affect
the dynamics of the model.  For this flavor of a “pure pleiotropy” model,
the technical details reduce to a standard population genetic model
where we tack on the trait values at the end.</p>
<p>Let’s set up a model where <span class="math notranslate nohighlight">\(S\)</span> is exponentially distributed with mean <code class="docutils literal notranslate"><span class="pre">-20</span></code>.
We’ll run a small population for a few generations.  This model will be nowhere
near equilibrium, but we’re just using it as an example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy11</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">N</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mean</span><span class="o">=-</span><span class="mi">20</span><span class="p">,</span> <span class="n">scaling</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="n">N</span><span class="p">)]</span>
<span class="n">recregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">PoissonInterval</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)]</span>
<span class="n">gvalue</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Multiplicative</span><span class="p">(</span><span class="n">scaling</span><span class="o">=</span><span class="mf">2.0</span><span class="p">)</span>
<span class="n">pdict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;nregions&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;sregions&quot;</span><span class="p">:</span> <span class="n">sregions</span><span class="p">,</span>
    <span class="s2">&quot;recregions&quot;</span><span class="p">:</span> <span class="n">recregions</span><span class="p">,</span>
    <span class="s2">&quot;rates&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1e-3</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
    <span class="s2">&quot;gvalue&quot;</span><span class="p">:</span> <span class="n">gvalue</span><span class="p">,</span>
    <span class="s2">&quot;prune_selected&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s2">&quot;simlen&quot;</span><span class="p">:</span> <span class="mi">150</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">params</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ModelParams</span><span class="p">(</span><span class="o">**</span><span class="n">pdict</span><span class="p">)</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
<span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">54321</span><span class="p">)</span>
<span class="n">fwdpy11</span><span class="o">.</span><span class="n">evolvets</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span> <span class="n">pop</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>16
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_6026/116052154.py:20: UserWarning: Applying a default demographic model where deme sizes are [1000] and the burn-in length is 10*1000
  fwdpy11.evolvets(rng, pop, params, 100)
</pre></div>
</div>
</div>
</div>
<p>Define a function relating <span class="math notranslate nohighlight">\(S\)</span> to <span class="math notranslate nohighlight">\(z\)</span>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">getz</span><span class="p">(</span><span class="n">S</span><span class="p">,</span> <span class="n">tau</span><span class="p">,</span> <span class="n">sigma</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="n">delta</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">delta</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">epsilon</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">sigma</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">delta</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">power</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">S</span><span class="p">),</span> <span class="n">tau</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">+</span> <span class="n">epsilon</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Apply the function and look at the results:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">101010</span><span class="p">)</span>
<span class="n">zvals</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">m</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">mutations</span><span class="p">):</span>
    <span class="n">zvals</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">getz</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">N</span> <span class="o">*</span> <span class="n">pop</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>

<span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">zvals</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>-0.009545936425350269 12.57306047991845
-0.010349104565670205 -2.2118130398591935
-0.011602139413926871 -7.084718822750244
-0.0032279509968761435 -1.737328013826212
-0.015449207258832993 -5.131892545676068
-0.0077168400201644664 4.512341872600712
-0.0022531736886723442 -4.096721449329499
-0.006044053890735736 -8.652700112679995
-0.0023465570676289957 0.6320311886584215
-0.0007746986635068075 2.2164750099025436
-0.021502039265014553 9.597464398344142
-0.009716042965590406 -7.443653191909565
-0.0008560999958330585 0.26396822983492146
-0.01597600966131924 4.994348870871331
-0.003548690338568886 1.3665737289445181
-0.005819804232148941 8.384444628613952
</pre></div>
</div>
</div>
</div>
<p>Traverse the tree sequence to get individual phenotypes under a
strictly additive model:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">phenotypes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">N</span><span class="p">)</span>
<span class="n">node_to_individual</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">diploid_metadata</span><span class="p">):</span>
    <span class="k">assert</span> <span class="n">j</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">node_to_individual</span>
    <span class="k">assert</span> <span class="n">j</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">node_to_individual</span>
    <span class="n">node_to_individual</span><span class="p">[</span><span class="n">j</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">i</span>
    <span class="n">node_to_individual</span><span class="p">[</span><span class="n">j</span><span class="o">.</span><span class="n">nodes</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">i</span>
<span class="n">ti</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">TreeIterator</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">tables</span><span class="p">,</span> <span class="n">pop</span><span class="o">.</span><span class="n">alive_nodes</span><span class="p">,</span> <span class="n">update_samples</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">ti</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">mutations</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">samples_below</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">node</span><span class="p">):</span>
            <span class="n">phenotypes</span><span class="p">[</span><span class="n">node_to_individual</span><span class="p">[</span><span class="n">n</span><span class="p">]]</span> <span class="o">+=</span> <span class="n">zvals</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>The trait value distribution is:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(array([-17.30540023, -15.73741894, -11.18144027, -10.39002813,
         -8.82204684,  -8.65270011,  -7.44365319,  -7.08471882,
         -5.13189255,  -4.09672145,  -2.57237695,  -2.21181304,
         -1.73732801,   0.        ,   0.26396823,   0.63203119,
          1.26406238,   1.36657373,   1.63054196,   2.21647501,
          4.51234187,   4.99434887,   5.62638006,   5.8789156 ,
          8.38444463,  11.81393941,  12.57306048]),
 array([  1,   1,   1,   1,   2,  11,  22,  39,   2,  17,   1,  16,  28,
        799,   6,  12,   1,  13,   1,   7,  13,   1,   1,   1,   1,   1,
          1]))
</pre></div>
</div>
</div>
</div>
<p>The mean trait value and the genetic variance are:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">phenotypes</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">phenotypes</span><span class="o">.</span><span class="n">var</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5.8498620701229225
</pre></div>
</div>
</div>
</div>
<p>For our final trick, let’s store them in the individual metadata:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">md</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">diploid_metadata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">md</span><span class="p">[</span><span class="s2">&quot;g&quot;</span><span class="p">][:]</span> <span class="o">=</span> <span class="n">phenotypes</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">md</span><span class="p">[</span><span class="s2">&quot;g&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">],</span> <span class="n">phenotypes</span><span class="p">[:</span><span class="mi">5</span><span class="p">]):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>-1.737328013826212 -1.737328013826212
0.0 0.0
0.0 0.0
0.0 0.0
-7.084718822750244 -7.084718822750244
</pre></div>
</div>
</div>
</div>
<p>The trick is that the <code class="docutils literal notranslate"><span class="pre">numpy</span></code> array is a <em>non-owning</em> array, meaning
that it is a simple “view” of the underlying <code class="docutils literal notranslate"><span class="pre">C++</span></code> data.  Further,
it happens to be read/write, and thus we can modify it.  (Try not
to abuse this.  More often than not, you’ll just break stuff.)</p>
<p>I feel that some comments about this model are warranted:</p>
<ul class="simple">
<li><p>The model is a simplification of earlier work by Keightley and Hill
(<span id="id4">[<a class="reference internal" href="../misc/bibliography.html#id29" title="P D Keightley and W G Hill. Quantitative genetic variability maintained by mutation-stabilizing selection balance in finite populations. Genet. Res., 52(1):33–43, August 1988.">KH88</a>]</span>, <span id="id5">[<a class="reference internal" href="../misc/bibliography.html#id30" title="Peter D Keightley and William G Hill. Variation maintained in quantitative traits with Mutation-Selection balance: pleiotropic Side-Effects on fitness traits. Proceedings of the Royal Society of London B: Biological Sciences, 242(1304):95–100, November 1990.">KH90</a>]</span>), which should be cited
alongside <span id="id6">[<a class="reference internal" href="../misc/bibliography.html#id24" title="Adam Eyre-Walker. Evolution in health and medicine sackler colloquium: genetic architecture of a complex trait and its implications for fitness and genome-wide association studies. Proc. Natl. Acad. Sci. U. S. A., 107 Suppl 1:1752–1756, January 2010.">EW10</a>]</span>.</p></li>
<li><p><span id="id7">[<a class="reference internal" href="../misc/bibliography.html#id29" title="P D Keightley and W G Hill. Quantitative genetic variability maintained by mutation-stabilizing selection balance in finite populations. Genet. Res., 52(1):33–43, August 1988.">KH88</a>]</span> and <span id="id8">[<a class="reference internal" href="../misc/bibliography.html#id30" title="Peter D Keightley and William G Hill. Variation maintained in quantitative traits with Mutation-Selection balance: pleiotropic Side-Effects on fitness traits. Proceedings of the Royal Society of London B: Biological Sciences, 242(1304):95–100, November 1990.">KH90</a>]</span>) show that this model predicts heritability
(the genetic variance) is linear-ish with <code class="docutils literal notranslate"><span class="pre">N</span></code>, with the details depending
somewhat on the model parameters.  The biological reasonableness
of that prediction is dubious.  <span id="id9">[<a class="reference internal" href="../misc/bibliography.html#id17" title="Toby Johnson and Nick Barton. Theoretical models of selection and mutation on quantitative traits. Philos. Trans. R. Soc. Lond. B Biol. Sci., 360(1459):1411–1425, July 2005.">JB05</a>]</span> discuss this point
in some detail, and give other relevant references.</p></li>
<li><p>Depending somewhat on the parameters of the <code class="docutils literal notranslate"><span class="pre">getz</span></code> function, one will eventually
generate an intermediate-frequency variant with a massive <span class="math notranslate nohighlight">\(z\)</span>,
meaning that it would explain a considerable amount of the genic
variance for the trait (high <span class="math notranslate nohighlight">\(2pqz^2\)</span>).  Such outcomes are contrary
to the results of human GWAS.</p></li>
<li><p>It becomes tempting to start doing arbitrary things when applying this model
to simulation output.  For example, only treating mutations in certain frequency
ranges as affecting trait values.  However, doing so means the predictions
made by such procedures are not natural outcomes of evolutionary models.
Consider only applying the <code class="docutils literal notranslate"><span class="pre">getz</span></code> function to mutations with frequency
<span class="math notranslate nohighlight">\(&lt; x\)</span> in order to say something about “rare alleles”.
This treatment of the data actually changes the model: mutations that are common
now (at the end of the simulation) were rare at some point in the past,
and had frequencies <span class="math notranslate nohighlight">\(&lt; x\)</span>.  Therefore, the model is one where mutations
suddenly stop affecting the trait once they hit some critical frequency.
Presumably, the same variants would affect the trait again should they drift to
a frequency below <span class="math notranslate nohighlight">\(x\)</span> if the simulation is run a bit longer.</p></li>
</ul>
</section>
<section id="writing-genetic-value-classes-in-python">
<span id="gvalues-python"></span><h2>Writing genetic value classes in Python<a class="headerlink" href="#writing-genetic-value-classes-in-python" title="Permalink to this heading">#</a></h2>
<p>Version <code class="docutils literal notranslate"><span class="pre">0.9.0</span></code> added the ability to write custom genetic value types in
Python.  Such classes are useful for prototyping new models
and/or generating example models for teaching/illustrating concepts.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
<section id="concepts">
<h3>Concepts<a class="headerlink" href="#concepts" title="Permalink to this heading">#</a></h3>
<section id="the-dimensionality-of-a-model">
<span id="python-genetic-value-dimensions"></span><h4>The “dimensionality” of a model<a class="headerlink" href="#the-dimensionality-of-a-model" title="Permalink to this heading">#</a></h4>
<p><code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> supports both univariate and multivariate trait simulations.
Thus, a genetic value implicitly has a dimensionality. For the
case of a standard population-genetic simulation where mutations
directly affect fitness, that dimensionality is <code class="docutils literal notranslate"><span class="pre">1</span></code>.  Simply put,
the genetic value is fitness.</p>
<p>For a simulation of mutations affecting <code class="docutils literal notranslate"><span class="pre">n</span></code> traits, the dimensionality
is <code class="docutils literal notranslate"><span class="pre">n</span></code>.</p>
<p>The dimensionality matters because we need to define a single genetic value
to populate the <a class="reference internal" href="types.html#fwdpy11.DiploidMetadata.g" title="fwdpy11.DiploidMetadata.g"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata.g</span></code></a> field for an offspring.
Doing so is simple for a one-dimensional trait.  For a simulation with
pleiotropy, we have a decision to make.  Should <a class="reference internal" href="types.html#fwdpy11.DiploidMetadata.g" title="fwdpy11.DiploidMetadata.g"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata.g</span></code></a>
store the Euclidean distance from the optima?  Or should it store the
individual’s value for some “focal” trait?  The <code class="docutils literal notranslate"><span class="pre">API</span></code> described here is
flexible enough to accommodate both.</p>
<p>The dimensionality of your genetic value to fitness map and your genetic value
object must match.  For the trait examples discussed above, this requirement
should make good sense.  For the case of modeling correlated effect sizes
of mutations in different demes (see <a class="reference internal" href="mvdes.html#mvdes"><span class="std std-ref">here</span></a>), we still require
this although it makes less sense. (Such simulations are limited to either
the standard population genetic case or the case of selection on a single
trait.  But relaxing this requirement for some models but not others becomes
problematic.)</p>
<p>The current version of this section only discusses one-dimensional genetic
values.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Update for multi-dimensional genetic values once we get documentation
in place for simulations of Gaussian stabilizing selection with pleiotropy.</p>
</div>
</section>
<section id="supporting-time-dependent-models">
<span id="python-genetic-value-update-desc"></span><h4>Supporting time-dependent models<a class="headerlink" href="#supporting-time-dependent-models" title="Permalink to this heading">#</a></h4>
<p>Genetic value methods may have dynamic states that depend on some
state of the population.  For example, something may change once
<code class="docutils literal notranslate"><span class="pre">pop.generation</span> <span class="pre">&gt;</span> <span class="pre">x</span></code>.</p>
<p>To support such time-dependent states, classes have a method with
the following signature:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">update</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pop</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="k">pass</span>
</pre></div>
</div>
<p>A type with no time-dependent state can in fact use the function shown
above. A concrete example is shown <a class="reference internal" href="#python-gvalue-to-fitness"><span class="std std-ref">below</span></a>,
in an example class called <code class="docutils literal notranslate"><span class="pre">PyGSSRandomOptimum</span></code>.  That example does not
use any information from <code class="docutils literal notranslate"><span class="pre">pop</span></code> to update its state.</p>
</section>
</section>
<section id="performance">
<h3>Performance<a class="headerlink" href="#performance" title="Permalink to this heading">#</a></h3>
<p>It will be hard to get “awesome” performance with Python genetic
value classes.  The culprit are abstract methods that get called
once per offspring.  That’s simply a lot of Python/C++ boundary
crossing.  The <code class="docutils literal notranslate"><span class="pre">update</span></code> function described above is called only
once per generation, making it much less of a performance killer.</p>
<p>We’ve tried hard to make the data exchange from C++ to your Python
code as efficient as possible.  We accomplished this by carefully
considering how we implement the <code class="docutils literal notranslate"><span class="pre">data</span></code> argument that gets passed
in.  Benchmarking showed that carefully designed proxies to the
underlying C++ data greatly improved run times.</p>
<p>Importantly, you do not have to write every component in Python!
If you are interested in additive effects models but a new “noise”
model, then use <a class="reference internal" href="genetic_values.html#fwdpy11.Additive" title="fwdpy11.Additive"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Additive</span></code></a> along with your custom
type.</p>
<p><code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> provides the following helper functions to improve efficiency:</p>
<dl class="py function">
<dt class="sig sig-object py" id="fwdpy11.strict_additive_effects">
<span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">strict_additive_effects</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.strict_additive_effects" title="Permalink to this definition">#</a></dt>
<dd><p>This function computes the sum of effect sizes in a
diploid genome, ignoring dominance.  The return value
is an offset from zero, so add 1.0 to convert to a
fitness if needed.</p>
<p>:param pop: The population
:type pop: fwdpy11.DiploidPopulation
:param metadata: The offspring metadata
:type metadata: fwdpy11.DiploidMetadata
:returns: Sum of effect sizes
:rtype: float</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fwdpy11.additive_effects">
<span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">additive_effects</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.additive_effects" title="Permalink to this definition">#</a></dt>
<dd><p>This function computes the sum of effect sizes in a
diploid genome, accounting for dominance.  The return value
is an offset from zero, so add 1.0 to convert to a
fitness if needed.</p>
<p>:param pop: The population
:type pop: fwdpy11.DiploidPopulation
:param metadata: The offspring metadata
:type metadata: fwdpy11.DiploidMetadata
:type scaling: float
:returns: Sum of effect sizes (accounting for heterozygous effects)
:rtype: float</p>
</dd></dl>

<p>The first function is equivalent to the following Python code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">strict_additive_effects</span><span class="p">(</span>
    <span class="n">pop</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">,</span>
    <span class="n">metadata</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidMetadata</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="n">g</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">dip</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">diploids</span><span class="p">[</span><span class="n">metadata</span><span class="o">.</span><span class="n">label</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="n">dip</span><span class="o">.</span><span class="n">first</span><span class="p">,</span> <span class="n">dip</span><span class="o">.</span><span class="n">second</span><span class="p">]:</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">pop</span><span class="o">.</span><span class="n">haploid_genomes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">smutations</span><span class="p">:</span>
            <span class="n">g</span> <span class="o">+=</span> <span class="n">pop</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">s</span>
    <span class="k">return</span> <span class="n">g</span>
</pre></div>
</div>
<p>The second function uses C++ code from <code class="docutils literal notranslate"><span class="pre">fwdpp</span></code> to do the calculation
accounting for dominance effects.</p>
</section>
<section id="technical-issues">
<h3>Technical issues<a class="headerlink" href="#technical-issues" title="Permalink to this heading">#</a></h3>
<p>This section discusses some important technicalities.</p>
<section id="abstract-base-classes">
<h4>Abstract base classes<a class="headerlink" href="#abstract-base-classes" title="Permalink to this heading">#</a></h4>
<p>In object oriented lingo, the <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> base classes here are <em>abstract
base classes</em> or <code class="docutils literal notranslate"><span class="pre">ABCs</span></code>.  This means that they simply describe a required
interface.  When you inherit from an ABC, you must fill in the details
of that interface.</p>
</section>
<section id="initializing-the-super-class">
<h4>Initializing the super class<a class="headerlink" href="#initializing-the-super-class" title="Permalink to this heading">#</a></h4>
<p>When writing object-oriented code, it is important that a derived
class properly initialize the base class.  In Python, base
classes are often called <code class="docutils literal notranslate"><span class="pre">superclasses</span></code> and derived classes are
<code class="docutils literal notranslate"><span class="pre">subclasses</span></code>.</p>
<p>The idiomatic approach in <code class="docutils literal notranslate"><span class="pre">Python</span> <span class="pre">3</span></code> is:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">Base</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
   <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
       <span class="bp">self</span><span class="o">.</span><span class="n">b</span> <span class="o">=</span> <span class="s2">&quot;I am a Base&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">Derived</span><span class="p">(</span><span class="n">Base</span><span class="p">):</span>
   <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
       <span class="bp">self</span><span class="o">.</span><span class="n">d</span> <span class="o">=</span> <span class="s2">&quot;I am a Derived&quot;</span>
       <span class="nb">super</span><span class="p">(</span><span class="n">Derived</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">d</span> <span class="o">=</span> <span class="n">Derived</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">b</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>I am a Base
</pre></div>
</div>
</div>
</div>
<p>Failure to do that <code class="docutils literal notranslate"><span class="pre">super(...)</span></code> business would result in an
<code class="docutils literal notranslate"><span class="pre">AttributeError</span></code> when trying to print <code class="docutils literal notranslate"><span class="pre">d.b</span></code>.</p>
<p>However, the classes that you are writing are <strong>not</strong> straightforward
Python classes!  While <code class="docutils literal notranslate"><span class="pre">super(...)</span></code> <em>may</em> work, it is not
guaranteed.  To quote the <code class="docutils literal notranslate"><span class="pre">pybind11</span></code> documentation:</p>
<blockquote>
<div><p>Note that a direct <strong>init</strong> constructor should be called,
and super() should not be used. For simple cases of linear inheritance,
super() may work, but once you begin mixing Python and C++ multiple inheritance,
things will fall apart due to differences between Python’s MRO and C++’s mechanisms.</p>
</div></blockquote>
<p>The source of this quote is the <code class="docutils literal notranslate"><span class="pre">pybind11</span></code> manual section on
<a class="reference external" href="https://pybind11.readthedocs.io/en/stable/advanced/classes.html#">Classes</a>.</p>
<p>The examples that follow take their advice.</p>
</section>
<section id="attrs">
<h4><code class="docutils literal notranslate"><span class="pre">attrs</span></code><a class="headerlink" href="#attrs" title="Permalink to this heading">#</a></h4>
<p>The <a class="reference external" href="https://www.attrs.org">attrs</a> package provides a library
of convenient class decorators that significantly reduce the amount
of “boiler plate” code needed to write Python
classes.  Some of the examples below will use this library.  You must
<em>not</em> use <code class="docutils literal notranslate"><span class="pre">slots=True</span></code>!  The <code class="docutils literal notranslate"><span class="pre">superclasses</span></code> already use the slots mechanism,
and it seems difficult for a derived class to also use slots and properly
initialize the base.</p>
<p>Other than this caveat regarding slots, we enthusiastically recommend
<code class="docutils literal notranslate"><span class="pre">attrs</span></code>.</p>
</section>
<section id="custom-decorators">
<h4>Custom decorators<a class="headerlink" href="#custom-decorators" title="Permalink to this heading">#</a></h4>
<p>Speaking of decorators, we provide several that make writing genetic
value types a bit easier.  You will see concrete examples using them
later.</p>
<dl class="py function">
<dt class="sig sig-object py" id="fwdpy11.custom_genetic_value_decorators.genetic_value_is_trait_default_clone">
<span class="sig-prename descclassname"><span class="pre">&#64;</span></span><span class="sig-prename descclassname"><span class="pre">fwdpy11.custom_genetic_value_decorators.</span></span><span class="sig-name descname"><span class="pre">genetic_value_is_trait_default_clone</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ndim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.custom_genetic_value_decorators.genetic_value_is_trait_default_clone" title="Permalink to this definition">#</a></dt>
<dd><p>This decorator is a <em>class</em> that adds the ability
for custom types to be passed to classes derived from
:class:<code class="docutils literal notranslate"><span class="pre">fwdpy11.DiploidGeneticValue</span></code>.</p>
<p>Because this decorator is a class, you need the <code class="docutils literal notranslate"><span class="pre">()</span></code>
to apply it.</p>
<p>This decorator is required because of the very different
object models of C++ and Python.
See <code class="docutils literal notranslate"><span class="pre">here</span> <span class="pre">&lt;https://github.com/pybind/pybind11/issues/1049&gt;</span></code>_
for some of the gory details.</p>
<p>.. py:method:: <strong>init</strong></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>:param ndim: The dimensionality of the model
:type ndim: int

The default is ``ndim=1``.
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone">
<span class="sig-prename descclassname"><span class="pre">&#64;</span></span><span class="sig-prename descclassname"><span class="pre">fwdpy11.custom_genetic_value_decorators.</span></span><span class="sig-name descname"><span class="pre">genetic_value_noise_default_clone</span></span><a class="headerlink" href="#fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone" title="Permalink to this definition">#</a></dt>
<dd><p>This is the analog to
:attr:<code class="docutils literal notranslate"><span class="pre">fwdpy11.custom_genetic_value_decorators.genetic_value_is_trait_default_clone</span></code>
for noise objects inheriting from :class:<code class="docutils literal notranslate"><span class="pre">fwdpy11.GeneticValueNoise</span></code>.</p>
<p>Unlike its analog, this decorator is a <em>function</em> and not a class.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="fwdpy11.custom_genetic_value_decorators.default_update">
<span class="sig-prename descclassname"><span class="pre">&#64;</span></span><span class="sig-prename descclassname"><span class="pre">fwdpy11.custom_genetic_value_decorators.</span></span><span class="sig-name descname"><span class="pre">default_update</span></span><a class="headerlink" href="#fwdpy11.custom_genetic_value_decorators.default_update" title="Permalink to this definition">#</a></dt>
<dd><p>This decorator adds the following function to <code class="docutils literal notranslate"><span class="pre">cls</span></code>,
which is a no-op function:</p>
<p>.. code-block:: python</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>def update(pop: fwdpy11.DiploidPopulation):
    pass
</pre></div>
</div>
<p>.. note::</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>In the future, we will experiment with adding a C++
implementation of this no-op.
</pre></div>
</div>
</dd></dl>

</section>
</section>
<section id="genetic-value-to-fitness-maps">
<span id="python-gvalue-to-fitness"></span><h3>Genetic value to fitness maps<a class="headerlink" href="#genetic-value-to-fitness-maps" title="Permalink to this heading">#</a></h3>
<p>Here, one derives a new class from <a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GeneticValueIsTrait" title="fwdpy11.GeneticValueIsTrait"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GeneticValueIsTrait</span></code></a>.
The basic form of such a class is:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy11</span>
<span class="kn">import</span> <span class="nn">fwdpy11.custom_genetic_value_decorators</span>


<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">genetic_value_is_trait_default_clone</span><span class="p">()</span>
<span class="k">class</span> <span class="nc">MyGeneticValueToFitness</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Do NOT call super()!</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidGeneticValueToFitnessData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">pass</span>

    <span class="k">def</span> <span class="nf">update</span><span class="p">(</span><span class="n">pop</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">pass</span>
</pre></div>
</div>
<p>If the object does not need to manage an internal state that changes over time,
we can simplify a bit with another decorator:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy11</span>
<span class="kn">import</span> <span class="nn">fwdpy11.custom_genetic_value_decorators</span>


<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">default_update</span>
<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">genetic_value_is_trait_default_clone</span><span class="p">()</span>
<span class="k">class</span> <span class="nc">MyGeneticValueToFitness</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Do NOT call super()!</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidGeneticValueToFitnessData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">pass</span>
</pre></div>
</div>
<p>Two complete examples come from the test suite.  The first example reimplements
<a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GSS" title="fwdpy11.GSS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GSS</span></code></a> in Python.  The second example implements a model
where the optimum changes to a new value each generation. (Note that
the second model requires that <a class="reference external" href="https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html#numpy.random.seed" title="(in NumPy v1.25)"><code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.random.seed()</span></code></a> be called elsewhere
so that results are reproducible.)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">math</span>

<span class="kn">import</span> <span class="nn">attr</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">fwdpy11</span>
<span class="kn">import</span> <span class="nn">fwdpy11.custom_genetic_value_decorators</span>


<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">default_update</span>
<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">genetic_value_is_trait_default_clone</span><span class="p">()</span>
<span class="nd">@attr</span><span class="o">.</span><span class="n">s</span><span class="p">()</span>
<span class="k">class</span> <span class="nc">PyGSS</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="p">):</span>
    <span class="n">opt</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">()</span>
    <span class="n">VS</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">__attrs_post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidGeneticValueToFitnessData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">math</span><span class="o">.</span><span class="n">e</span> <span class="o">**</span> <span class="p">(</span>
            <span class="o">-</span><span class="p">((</span><span class="n">data</span><span class="o">.</span><span class="n">offspring_metadata</span><span class="o">.</span><span class="n">g</span> <span class="o">+</span> <span class="n">data</span><span class="o">.</span><span class="n">offspring_metadata</span><span class="o">.</span><span class="n">e</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">opt</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
            <span class="o">/</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">VS</span><span class="p">)</span>
        <span class="p">)</span>


<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">genetic_value_is_trait_default_clone</span><span class="p">()</span>
<span class="nd">@attr</span><span class="o">.</span><span class="n">s</span><span class="p">()</span>
<span class="k">class</span> <span class="nc">PyGSSRandomOptimum</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="p">):</span>
    <span class="n">opt</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">()</span>
    <span class="n">VS</span> <span class="o">=</span> <span class="n">attr</span><span class="o">.</span><span class="n">ib</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">__attrs_post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueIsTrait</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">optima</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidGeneticValueToFitnessData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">math</span><span class="o">.</span><span class="n">e</span> <span class="o">**</span> <span class="p">(</span>
            <span class="o">-</span><span class="p">((</span><span class="n">data</span><span class="o">.</span><span class="n">offspring_metadata</span><span class="o">.</span><span class="n">g</span> <span class="o">+</span> <span class="n">data</span><span class="o">.</span><span class="n">offspring_metadata</span><span class="o">.</span><span class="n">e</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">opt</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
            <span class="o">/</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">VS</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">update</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pop</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">opt</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">optima</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">pop</span><span class="o">.</span><span class="n">generation</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">opt</span><span class="p">))</span>
</pre></div>
</div>
<section id="the-data-type">
<h4>The data type<a class="headerlink" href="#the-data-type" title="Permalink to this heading">#</a></h4>
<p>The type passed into the <code class="docutils literal notranslate"><span class="pre">__call__</span></code> function is:</p>
<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.DiploidGeneticValueToFitnessData">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">DiploidGeneticValueToFitnessData</span></span><a class="headerlink" href="#fwdpy11.DiploidGeneticValueToFitnessData" title="Permalink to this definition">#</a></dt>
<dd><p>.. versionadded:: 0.9.0</p>
<p>This class supports the buffer protocol, which exposes the
genetic values array.  The most efficient access will
be via a :class:<code class="docutils literal notranslate"><span class="pre">memoryview</span></code>.</p>
<p>Instances of this class have the following attributes:</p>
<p>.. py:attribute:: offspring_metadata</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>An instance of :class:`fwdpy11.DiploidMetadata`, giving you
access to those fields that have been assigned to the offspring.
(This is a *copy* of the metadata from the C++ side.)
</pre></div>
</div>
<p>.. py:attribute:: offspring_metadata_index</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>A 64 bit integer that gives the location (index) of
``offspring_metadata`` in :attr:`fwdpy11.DiploidPopulation.diploid_metadata`.
This index is useful in the event of mass migrations via copies,
which can cause a mismatch between :attr:`fwdpy11.DiploidMetadata.label`
and this value.
</pre></div>
</div>
<p>.. py:attribute:: parent1_metadata</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>The first parent&#39;s metadata (:class:`fwdpy11.DiploidMetadata`).
</pre></div>
</div>
<p>.. py:attribute:: parent2_metadata</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>The second parent&#39;s metadata (:class:`fwdpy11.DiploidMetadata`).
</pre></div>
</div>
</dd></dl>

</section>
</section>
<section id="genetic-value-noise">
<h3>Genetic value “noise”<a class="headerlink" href="#genetic-value-noise" title="Permalink to this heading">#</a></h3>
<p>A custom “noise” class inherits from <a class="reference internal" href="gvaluenoise.html#fwdpy11.GeneticValueNoise" title="fwdpy11.GeneticValueNoise"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GeneticValueNoise</span></code></a>.
A minimal implementation has the following form:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">genetic_value_noise_default_clone</span>
<span class="k">class</span> <span class="nc">MyGeneticValueNoise</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueNoise</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidGeneticValueNoiseData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">pass</span>

    <span class="k">def</span> <span class="nf">update</span><span class="p">(</span><span class="n">pop</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">pass</span>
</pre></div>
</div>
<p>If your model can allow a no-op <code class="docutils literal notranslate"><span class="pre">update</span></code> function, you may apply
the decorator
<a class="reference internal" href="#fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone" title="fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone</span></code></a>.</p>
<p>A simple example from the test suite implements Gaussian noise with mean zero
and standard deviation <code class="docutils literal notranslate"><span class="pre">0.1</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy11</span>
<span class="kn">import</span> <span class="nn">fwdpy11.custom_genetic_value_decorators</span>


<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">default_update</span>
<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">genetic_value_noise_default_clone</span>
<span class="k">class</span> <span class="nc">PyNoise</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueNoise</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GeneticValueNoise</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidGeneticValueNoiseData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">gsl_ran_gaussian_ziggurat</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">rng</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span>
</pre></div>
</div>
<p>See <a class="reference internal" href="gslrandom.html#fwdpy11.gsl_ran_gaussian_ziggurat" title="fwdpy11.gsl_ran_gaussian_ziggurat"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.gsl_ran_gaussian_ziggurat()</span></code></a> for details on that function.</p>
<section id="id10">
<h4>The data type<a class="headerlink" href="#id10" title="Permalink to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.DiploidGeneticValueNoiseData">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">DiploidGeneticValueNoiseData</span></span><a class="headerlink" href="#fwdpy11.DiploidGeneticValueNoiseData" title="Permalink to this definition">#</a></dt>
<dd><p>.. versionadded:: 0.9.0</p>
<p>Instances of this class have the following attributes:</p>
<p>.. py:attribute:: rng</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>The simulation&#39;s random number generation, an
instance of :class:`fwdpy11.GSLrng`
</pre></div>
</div>
<p>.. py:attribute:: offspring_metadata</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>An instance of :class:`fwdpy11.DiploidMetadata`, giving you
access to those fields that have been assigned to the offspring.
(This is a *copy* of the metadata from the C++ side.)
</pre></div>
</div>
<p>.. py:attribute:: offspring_metadata_index</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>A 64 bit integer that gives the location (index) of
``offspring_metadata`` in :attr:`fwdpy11.DiploidPopulation.diploid_metadata`.
This index is useful in the event of mass migrations via copies,
which can cause a mismatch between :attr:`fwdpy11.DiploidMetadata.label`
and this value.
</pre></div>
</div>
<p>.. py:attribute:: parent1_metadata</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>The first parent&#39;s metadata (:class:`fwdpy11.DiploidMetadata`).
</pre></div>
</div>
<p>.. py:attribute:: parent2_metadata</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>The first parent&#39;s metadata (:class:`fwdpy11.DiploidMetadata`).
</pre></div>
</div>
</dd></dl>

</section>
</section>
<section id="new-genetic-value-models">
<h3>New genetic value models<a class="headerlink" href="#new-genetic-value-models" title="Permalink to this heading">#</a></h3>
<section id="the-abstract-base-class">
<h4>The abstract base class<a class="headerlink" href="#the-abstract-base-class" title="Permalink to this heading">#</a></h4>
<p>The ABC type is <a class="reference internal" href="#fwdpy11.PyDiploidGeneticValue" title="fwdpy11.PyDiploidGeneticValue"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.PyDiploidGeneticValue</span></code></a>:</p>
<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.PyDiploidGeneticValue">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">PyDiploidGeneticValue</span></span><a class="headerlink" href="#fwdpy11.PyDiploidGeneticValue" title="Permalink to this definition">#</a></dt>
<dd><p>.. py:method:: <strong>init</strong>()</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>:param ndim:
:type ndim: int
:param genetic_value_to_fitness:
:type genetic_value_to_fitness: fwdpy11.GeneticValueIsTrait or None
:param noise:
:type noise: fwdpy11.GeneticValueNoise or None
</pre></div>
</div>
<p>.. py:method:: calculate_gvalue
:abstractmethod:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>:param data: Input data
:type data: fwdpy11.PyDiploidGeneticValueData
:returns: The value to be stored in the offspring&#39;s
          :attr:`fwdpy11.DiploidMetadata.g`
:rtype: float
</pre></div>
</div>
<p>.. py:method:: genetic_value_to_fitness</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>:param data: Input data
:type data: fwdpy11.DiploidGeneticValueToFitnessData

:returns: fitness
:rtype: float

.. note::

    This function does not need to be defined
    by derived classes most of the time.
    The default behavior is to apply
    the instance of :class:`fwdpy11.GeneticValueToFitnessMap`
    stored by the instance of :class:`fwdpy11.DiploidGeneticValue`
    (or :class:`fwdpy11.PyDiploidGeneticValue`).  Defining
    this function in a derived class skips calling held instance
    in favor of the derived class implementation.
    In general, one only needs to derive this class for models
    where either individual genetic values depend on genotypes of the
    rest of the population. See :ref:`here &lt;more-complex-gvalue-models&gt;`.
</pre></div>
</div>
<p>.. py:method:: update</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>:param pop: The population being simulated
:type pop: fwdpy11.DiploidPopulation
:rtype: None

.. note::

    A default implementation can be defined using
    the decorator
    :attr:`fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone`.
</pre></div>
</div>
</dd></dl>

</section>
<section id="outline-of-user-defined-classes">
<h4>Outline of user-defined classes<a class="headerlink" href="#outline-of-user-defined-classes" title="Permalink to this heading">#</a></h4>
<p>The form of a custom genetic value type inherits from the ABC
described above and would look something like this:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">MyGeneticValueObject</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">PyDiploidGeneticValue</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">):</span>
        <span class="c1"># Do not call super()!</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PyDiploidGeneticValue</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">calculate_gvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PyDiploidGeneticValueData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="k">pass</span>

    <span class="k">def</span> <span class="nf">update</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pop</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">pass</span>
</pre></div>
</div>
</section>
<section id="method-requirements">
<h4>Method requirements<a class="headerlink" href="#method-requirements" title="Permalink to this heading">#</a></h4>
<p><a class="reference internal" href="#fwdpy11.PyDiploidGeneticValue" title="fwdpy11.PyDiploidGeneticValue"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.PyDiploidGeneticValue</span></code></a> <em>does</em> provide a default <code class="docutils literal notranslate"><span class="pre">update</span></code>
function that updates the genetic-value-to-fitness-map and noise objects.</p>
<p>If your model can allow a no-op <code class="docutils literal notranslate"><span class="pre">update</span></code> function, you may apply
the decorator
<a class="reference internal" href="#fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone" title="fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone</span></code></a>.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">genetic_value_to_fitness</span></code> function must completely replace the
functionality of an instance of <a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GeneticValueIsTrait" title="fwdpy11.GeneticValueIsTrait"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GeneticValueIsTrait</span></code></a>.
See <a class="reference internal" href="#python-gvalue-to-fitness"><span class="std std-ref">here</span></a> for details.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">calculate_gvalue</span></code> function is more complex because it has more
responsibilities:</p>
<ol class="arabic simple">
<li><p>It must return a value that will populate the genetic value
field of the offspring’s metadata (<a class="reference internal" href="types.html#fwdpy11.DiploidMetadata.g" title="fwdpy11.DiploidMetadata.g"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata.g</span></code></a>).</p></li>
<li><p>It also needs to populate an array of <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a> that represents
the genetic values for all trait dimensions. See <a class="reference internal" href="#python-genetic-value-dimensions"><span class="std std-ref">here</span></a>.
For the common case of a one-dimensional genetic value,
you populate element <code class="docutils literal notranslate"><span class="pre">0</span></code> of this array with value that you
will return from this function. The “genetic values array” is
accessible via the Python buffer protocol.
The next section shows a concrete example.</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The steps described in <code class="docutils literal notranslate"><span class="pre">2</span></code> are populating
<a class="reference internal" href="genetic_values.html#fwdpy11.DiploidGeneticValue.genetic_values" title="fwdpy11.DiploidGeneticValue.genetic_values"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidGeneticValue.genetic_values</span></code></a>.  In the
future, we may provide read-write access to that field, which
would constitute an API change.  However, the path of least
resistance in the short term was to implement the current approach.</p>
</div>
</section>
<section id="example-strictly-additive-effects-on-a-trait">
<h4>Example: strictly additive effects on a trait<a class="headerlink" href="#example-strictly-additive-effects-on-a-trait" title="Permalink to this heading">#</a></h4>
<p>Here is an example of a strict additive model of a trait,
taken from the test suite. You’ll notice that we are accessing
arrays on the <code class="docutils literal notranslate"><span class="pre">C++</span></code> using Python <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#memoryview" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">memoryview</span></code></a> objects.
This is a <em>very</em> efficient method, and much faster
than creating a temporary, non-owning <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.25)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy11</span>
<span class="kn">import</span> <span class="nn">fwdpy11.custom_genetic_value_decorators</span>


<span class="nd">@fwdpy11</span><span class="o">.</span><span class="n">custom_genetic_value_decorators</span><span class="o">.</span><span class="n">default_update</span>
<span class="k">class</span> <span class="nc">PyAdditive</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">PyDiploidGeneticValue</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gvalue_to_fitness</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">noise</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PyDiploidGeneticValue</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span>
                                               <span class="n">gvalue_to_fitness</span><span class="p">,</span> <span class="n">noise</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">calculate_gvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                         <span class="n">data</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PyDiploidGeneticValueData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">strict_additive_effects</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">pop</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">offspring_metadata</span><span class="p">)</span>
        <span class="nb">memoryview</span><span class="p">(</span><span class="n">data</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">s</span>
        <span class="k">return</span> <span class="n">s</span>
</pre></div>
</div>
<p>If you want to account for the heterozygous effect of mutations, then the
most explicit approach is probably to use <a class="reference external" href="https://docs.python.org/3/library/collections.html#collections.Counter" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">collections.Counter</span></code></a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">calculate_gvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data</span><span class="p">):</span>
    <span class="n">s</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">()</span>
    <span class="n">pop</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">pop</span>
    <span class="n">dip</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">diploids</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">offspring_metadata</span><span class="o">.</span><span class="n">label</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="p">[</span><span class="n">dip</span><span class="o">.</span><span class="n">first</span><span class="p">,</span> <span class="n">dip</span><span class="o">.</span><span class="n">second</span><span class="p">]:</span>
        <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">pop</span><span class="o">.</span><span class="n">haploid_genomes</span><span class="p">[</span><span class="n">g</span><span class="p">]</span><span class="o">.</span><span class="n">smutations</span><span class="p">:</span>
            <span class="n">m</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">mutations</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
            <span class="n">c</span><span class="o">.</span><span class="n">update</span><span class="p">([(</span><span class="n">m</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="n">m</span><span class="o">.</span><span class="n">s</span><span class="p">,</span> <span class="n">m</span><span class="o">.</span><span class="n">h</span><span class="p">)])</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">j</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># Aa</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># aa</span>
            <span class="c1"># This is a hard-coded &quot;scaling&quot;</span>
            <span class="c1"># of 2</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="c1"># Use a memory view to update</span>
    <span class="c1"># the genetic values array</span>
    <span class="nb">memoryview</span><span class="p">(</span><span class="n">data</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">s</span>
    <span class="k">return</span> <span class="n">s</span>
</pre></div>
</div>
<p>For additive cases like this, though, you’ll get better performance
with <a class="reference internal" href="#fwdpy11.additive_effects" title="fwdpy11.additive_effects"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.additive_effects()</span></code></a>.</p>
</section>
<section id="more-complex-scenarios-such-as-social-interactions">
<span id="more-complex-gvalue-models"></span><h4>More complex scenarios (such as social interactions)<a class="headerlink" href="#more-complex-scenarios-such-as-social-interactions" title="Permalink to this heading">#</a></h4>
<p>Some models are not easily implemented with the genetic value +
genetic value to fitness map + noise trio of types defined here.
The reason has to do with how these types are applied during a simulation,
which is described in more detail <a class="reference internal" href="../technical/genetic_values.html#gvalue-technical-details"><span class="std std-ref">here</span></a>, and
in <a class="reference internal" href="../technical/genetic_values.html#gvalue-simulation-behavior"><span class="std std-ref">this section</span></a> specifically.</p>
</section>
<section id="the-data-types">
<h4>The data types<a class="headerlink" href="#the-data-types" title="Permalink to this heading">#</a></h4>
<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.PyDiploidGeneticValueData">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">PyDiploidGeneticValueData</span></span><a class="headerlink" href="#fwdpy11.PyDiploidGeneticValueData" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<div class="versionadded">
<p><span class="versionmodified added">New in version 0.9.0.</span></p>
</div>
<p>This class supports the buffer protocol, which exposes the
genetic values array.  The most efficient access will
be via a <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#memoryview" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">memoryview</span></code></a>.</p>
<p>Instances of this class have the following attributes:</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="rng">
<span class="sig-name descname"><span class="pre">rng</span></span><a class="headerlink" href="#rng" title="Permalink to this definition">#</a></dt>
<dd><p>The simulation’s random number generation, an
instance of <a class="reference internal" href="types.html#fwdpy11.GSLrng" title="fwdpy11.GSLrng"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GSLrng</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pop">
<span class="sig-name descname"><span class="pre">pop</span></span><a class="headerlink" href="#pop" title="Permalink to this definition">#</a></dt>
<dd><p>The population, an instance of <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="offspring_metadata">
<span class="sig-name descname"><span class="pre">offspring_metadata</span></span><a class="headerlink" href="#offspring_metadata" title="Permalink to this definition">#</a></dt>
<dd><p>The offspring’s <a class="reference internal" href="types.html#fwdpy11.DiploidMetadata" title="fwdpy11.DiploidMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="parent1_metadata">
<span class="sig-name descname"><span class="pre">parent1_metadata</span></span><a class="headerlink" href="#parent1_metadata" title="Permalink to this definition">#</a></dt>
<dd><p>The first parent’s <a class="reference internal" href="types.html#fwdpy11.DiploidMetadata" title="fwdpy11.DiploidMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="parent2_metadata">
<span class="sig-name descname"><span class="pre">parent2_metadata</span></span><a class="headerlink" href="#parent2_metadata" title="Permalink to this definition">#</a></dt>
<dd><p>The second parent’s <a class="reference internal" href="types.html#fwdpy11.DiploidMetadata" title="fwdpy11.DiploidMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata</span></code></a></p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="offspring_metadata_index">
<span class="sig-name descname"><span class="pre">offspring_metadata_index</span></span><a class="headerlink" href="#offspring_metadata_index" title="Permalink to this definition">#</a></dt>
<dd><p>A 64 bit integer that gives the location (index) of
<code class="docutils literal notranslate"><span class="pre">offspring_metadata</span></code> in <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation.diploid_metadata" title="fwdpy11.DiploidPopulation.diploid_metadata"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation.diploid_metadata</span></code></a>.
This index is useful in the event of mass migrations via copies,
which can cause a mismatch between <a class="reference internal" href="types.html#fwdpy11.DiploidMetadata.label" title="fwdpy11.DiploidMetadata.label"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata.label</span></code></a>
and this value.</p>
</dd></dl>

</section>
</section>
</section>
<section id="subtleties-of-starting-finishing-tree-sequences-using-msprime">
<span id="msprime-subtleties"></span><h2>Subtleties of starting/finishing tree sequences using <code class="docutils literal notranslate"><span class="pre">msprime</span></code><a class="headerlink" href="#subtleties-of-starting-finishing-tree-sequences-using-msprime" title="Permalink to this heading">#</a></h2>
<p>Some care is required starting a simulation from a tree sequence (see <a class="reference internal" href="../short_vignettes/initpops_vignette.html#precapitation"><span class="std std-ref">here</span></a>) or finishing a simulation with one (see <a class="reference internal" href="../short_vignettes/recapitation.html#recapitation"><span class="std std-ref">here</span></a>).</p>
<p>If simulating with discrete recombination positions, it may be important to realize that the current stable releases of <code class="docutils literal notranslate"><span class="pre">msprime</span></code> model continuous recombination breakpoints.
A future <code class="docutils literal notranslate"><span class="pre">msprime</span></code> release will support discretized positions.</p>
<p>When simulating with genome lengths longer than unity (1.0), determining the correct recombination rate for <code class="docutils literal notranslate"><span class="pre">msprime</span></code> needs additional work.
The examples shown <a class="reference internal" href="../short_vignettes/recapitation.html#recapitation"><span class="std std-ref">here</span></a> and <a class="reference internal" href="../short_vignettes/initpops_vignette.html#precapitation"><span class="std std-ref">here</span></a> show the following method for determining the <code class="docutils literal notranslate"><span class="pre">msprime</span></code> recombination rate:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">recombination_rate</span> <span class="o">=</span> <span class="n">rho</span> <span class="o">/</span> <span class="n">Ne</span> <span class="o">/</span> <span class="mi">4</span>
</pre></div>
</div>
<p>That code implies a genome length of 1.0.
A more general example, for a genome of length <code class="docutils literal notranslate"><span class="pre">L</span></code> would look like:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">recombination_rate</span> <span class="o">=</span> <span class="n">rho</span> <span class="o">/</span> <span class="n">Ne</span> <span class="o">/</span> <span class="mi">4</span> <span class="o">/</span> <span class="n">L</span>
</pre></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./pages"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="definitions.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Definitions</p>
      </div>
    </a>
    <a class="right-next"
       href="mvdes.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Different effect sizes of mutations in different demes</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#executing-multiple-replicates-in-parallel-using-concurrent-futures">Executing multiple replicates in parallel using <code class="docutils literal notranslate"><span class="pre">concurrent.futures</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-long-to-run-the-simulation">How long to run the simulation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-eyre-walker-model-of-complex-traits">The “Eyre-Walker” model of complex traits</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#writing-genetic-value-classes-in-python">Writing genetic value classes in Python</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#concepts">Concepts</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-dimensionality-of-a-model">The “dimensionality” of a model</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#supporting-time-dependent-models">Supporting time-dependent models</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#performance">Performance</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-issues">Technical issues</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#abstract-base-classes">Abstract base classes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#initializing-the-super-class">Initializing the super class</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#attrs"><code class="docutils literal notranslate"><span class="pre">attrs</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#custom-decorators">Custom decorators</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-value-to-fitness-maps">Genetic value to fitness maps</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-data-type">The data type</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#genetic-value-noise">Genetic value “noise”</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">The data type</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#new-genetic-value-models">New genetic value models</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-abstract-base-class">The abstract base class</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#outline-of-user-defined-classes">Outline of user-defined classes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-requirements">Method requirements</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-strictly-additive-effects-on-a-trait">Example: strictly additive effects on a trait</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#more-complex-scenarios-such-as-social-interactions">More complex scenarios (such as social interactions)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#the-data-types">The data types</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#subtleties-of-starting-finishing-tree-sequences-using-msprime">Subtleties of starting/finishing tree sequences using <code class="docutils literal notranslate"><span class="pre">msprime</span></code></a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            <div class="bd-footer-content__inner">
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Kevin Thornton
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div></div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>