

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Genetic values &#8212; fwdpy11 manual</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'pages/genetic_values';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="DataMatrix and StateMatrix" href="datamatrix.html" />
    <link rel="prev" title="Writing “plugins” with C++" href="../technical/writingplugins.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    <p class="title logo__title">fwdpy11 manual</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Introduction
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Installation and setup</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="userenv.html">Setting up user environments for installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="deploymenttools.html">Deployment tools</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - operations on populations</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/initpops_vignette.html">Initializing a population</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/poptour_vignette.html">Properties of populations</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - setting up the genetics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/genetics_vignettes_intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/geneticmaps_vignette.html">Genetic maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/des_vignette.html">Distributions of effect sizes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/mutationdominance_vignette.html">Mutations with variable dominance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/neutralmuts_vignette.html">Neutral mutations during a simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/gvalues_fitness.html">Genetic values - mutations with direct effects on fitness</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/gvalues_traits.html">Genetic values - simulating phenotypes/traits.</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/workingexample_fitness.html">Working example - parameters for simulating direct effects on fitness</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/workingexample_trait.html">Working example - parameters for the simulation of a trait</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - simulation with tree sequence recording</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/evolvets_vignette.html">Evolving with tree sequence recording</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/recorders_vignette.html">Recording data during a simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/recapitation.html">Finishing a simulation with msprime</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/neutralmutsafter_vignette.html">Adding neutral mutations to a population with existing ancestry</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/import_mutations_from_tskit.html">Importing mutations from tskit</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - demographic modeling</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/demography_vignettes_intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/demes_vignette.html">Using <code class="docutils literal notranslate"><span class="pre">demes</span></code> to specify demographic models.</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - varying fitness conditions across demes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/varying_fitness_conditions_across_demes_intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/multivariate_gaussian_effect_sizes_across_demes.html">The multivariate Gaussian distribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/multivariate_lognormal_effect_sizes.html">The multivariate lognormal distribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/custom_multivarate_distributions.html">“Custom” multivariate distributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/special_considerations_demes_models.html">Special considerations for demes models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/genetic_values_multi_deme_models_advanced.html">Advanced topics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Short vignettes - conditional models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/trackmutationfates.html">Tracking user-specified new mutations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/selective_sweep.html">Selective sweeps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../short_vignettes/incomplete_sweep.html">Incomplete selective sweeps</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Longer vignettes - exporting data to tskit</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/tskitconvert_vignette.html">Converting data to tskit format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/tskit_metadata_vignette.html">Working with data in tskit format</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Longer vignettes - complete examples</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/bgs_vignette.html">Background selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/gss_vignette.html">Gaussian stabilizing selection with an optimum shift</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/pyfreqtracker.html">Tracking mutations over time</a></li>
<li class="toctree-l1"><a class="reference internal" href="../long_vignettes/pyfreqtracker_sqlite.html">Tracking mutations over time using <code class="docutils literal notranslate"><span class="pre">sqlite3</span></code></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="definitions.html">Definitions</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="advancedtopics.html">Advanced topics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Objects and concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="mvdes.html">Different effect sizes of mutations in different demes</a></li>
<li class="toctree-l1"><a class="reference internal" href="tsoverview.html">Conceptual overview of tree sequence recording</a></li>
<li class="toctree-l1"><a class="reference internal" href="tstypes.html">Data structures related to tree sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="tablefs.html">Calculating frequency spectra from tree sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="recorders.html">Time series analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Technical details</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../technical/demes.html">Details of <code class="docutils literal notranslate"><span class="pre">demes</span></code> integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../technical/genetic_values.html">Technical overview of genetic value calculations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../technical/writingplugins.html">Writing “plugins” with C++</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data types and fuctions</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Genetic values</a></li>
<li class="toctree-l1"><a class="reference internal" href="datamatrix.html">DataMatrix and StateMatrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="functions.html">Function documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="gslrandom.html">Random number distributions</a></li>
<li class="toctree-l1"><a class="reference internal" href="gvaluenoise.html">Adding random noise to genetic values</a></li>
<li class="toctree-l1"><a class="reference internal" href="gvalue_to_fitness.html">Mapping genetic values to fitness</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_params.html">Model parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="regiontypes.html">Setting up genomic intervals</a></li>
<li class="toctree-l1"><a class="reference internal" href="tskit_tools.html">Module <code class="docutils literal notranslate"><span class="pre">fwdpy11.tskit_tools</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="conditional_models.html">Module <code class="docutils literal notranslate"><span class="pre">fwdpy11.conditional_models</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="types.html">Data types related to simulated populations</a></li>



</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../examples/gss_divergent_optima.html">Adaptation to different optima with multivariate distribution of effect sizes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/pysnowdrift.html">Snowdrift model in python</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples/recorder.html">Example of time series analysis implemented in C++</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Miscellany</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../misc/changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/deprecated.html">Deprecated features</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/developersguide.html">Developer’s guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/pubs.html">Publications using fwdpy11</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/upgrade_path.html">Upgrade path</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/bibliography.html">Literature cited</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/molpopgen/fwdpy11" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/molpopgen/fwdpy11/issues/new?title=Issue%20on%20page%20%2Fpages/genetic_values.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/pages/genetic_values.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Genetic values</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="genetic-values">
<h1>Genetic values<a class="headerlink" href="#genetic-values" title="Permalink to this heading">#</a></h1>
<p>For some background, see <a class="reference internal" href="definitions.html#definitions"><span class="std std-ref">Definitions</span></a>.</p>
<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.DiploidGeneticValue">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">DiploidGeneticValue</span></span><a class="headerlink" href="#fwdpy11.DiploidGeneticValue" title="Permalink to this definition">#</a></dt>
<dd><p>ABC for genetic value calculations for diploid members of <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a></p>
<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.DiploidGeneticValue.genetic_values">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genetic_values</span></span><a class="headerlink" href="#fwdpy11.DiploidGeneticValue.genetic_values" title="Permalink to this definition">#</a></dt>
<dd><p>Return the list of genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.DiploidGeneticValue.maps_to_fitness">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_fitness</span></span><a class="headerlink" href="#fwdpy11.DiploidGeneticValue.maps_to_fitness" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a mapping directly to fitness, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.DiploidGeneticValue.maps_to_trait_value">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_trait_value</span></span><a class="headerlink" href="#fwdpy11.DiploidGeneticValue.maps_to_trait_value" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a trait value, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.DiploidGeneticValue.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#fwdpy11.DiploidGeneticValue.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Return the dimensions of the genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.Additive">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">Additive</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scaling</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gvalue_to_fitness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ndemes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Additive" title="Permalink to this definition">#</a></dt>
<dd><p>Additive effects on genetic values.</p>
<p>This class has the following attributes, whose names
are also <cite>kwargs</cite> for intitialization.  The attribute names
also determine the order of positional arguments:</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scaling</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – How to treat mutant homozygotes.</p></li>
<li><p><strong>gvalue_to_fitness</strong> (<a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GeneticValueIsTrait" title="fwdpy11.GeneticValueIsTrait"><em>fwdpy11.GeneticValueIsTrait</em></a>) – How to map trait value to fitness</p></li>
<li><p><strong>noise</strong> (<a class="reference internal" href="gvaluenoise.html#fwdpy11.GeneticValueNoise" title="fwdpy11.GeneticValueNoise"><em>fwdpy11.GeneticValueNoise</em></a>) – Random effects on trait values</p></li>
</ul>
</dd>
</dl>
<p>When <cite>gvalue_to_fitness</cite> is <cite>None</cite>, then we are
modeling additive effects on fitness.</p>
<p>For a model of fitness, the genetic value is 1, 1+e*h,
1+`scaling`*e for genotypes AA, Aa, and aa, respectively,
where <cite>e</cite> and <cite>h</cite> are the effect size and dominance, respectively.</p>
<p>For a model of a trait (phenotype), meaning <cite>gvalue_to_fitness</cite>
is not <cite>None</cite>, the values for the three genotypes are 0, e*h,
and e, respectively.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8.0: </span>Refactored to use attrs and inherit from
low-level C++ class</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.Additive.asblack">
<span class="sig-name descname"><span class="pre">asblack</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Additive.asblack" title="Permalink to this definition">#</a></dt>
<dd><p>Return a string representation formatted with black</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.Additive.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Additive.asdict" title="Permalink to this definition">#</a></dt>
<dd><p>Return dict representation</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.Additive.fromdict">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fromdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Additive.fromdict" title="Permalink to this definition">#</a></dt>
<dd><p>Build an instance from a dictionary</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Additive.genetic_values">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genetic_values</span></span><a class="headerlink" href="#fwdpy11.Additive.genetic_values" title="Permalink to this definition">#</a></dt>
<dd><p>Return the list of genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Additive.maps_to_fitness">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_fitness</span></span><a class="headerlink" href="#fwdpy11.Additive.maps_to_fitness" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a mapping directly to fitness, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Additive.maps_to_trait_value">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_trait_value</span></span><a class="headerlink" href="#fwdpy11.Additive.maps_to_trait_value" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a trait value, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Additive.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#fwdpy11.Additive.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Return the dimensions of the genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">Multiplicative</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scaling</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gvalue_to_fitness</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ndemes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Multiplicative" title="Permalink to this definition">#</a></dt>
<dd><p>Multiplicative effects on genetic values.</p>
<p>This class has the following attributes, whose names
are also <cite>kwargs</cite> for intitialization.  The attribute names
also determine the order of positional arguments:</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scaling</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><em>float</em></a>) – How to treat mutant homozygotes.</p></li>
<li><p><strong>gvalue_to_fitness</strong> (<a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GeneticValueIsTrait" title="fwdpy11.GeneticValueIsTrait"><em>fwdpy11.GeneticValueIsTrait</em></a>) – How to map trait value to fitness</p></li>
<li><p><strong>noise</strong> (<a class="reference internal" href="gvaluenoise.html#fwdpy11.GeneticValueNoise" title="fwdpy11.GeneticValueNoise"><em>fwdpy11.GeneticValueNoise</em></a>) – Random effects on trait values</p></li>
</ul>
</dd>
</dl>
<p>When <cite>gvalue_to_fitness</cite> is <cite>None</cite>, then we are
modeling multiplicative effects on fitness.</p>
<p>For a model of fitness, the genetic value is 1, 1+e*h,
1+`scaling`*e for genotypes AA, Aa, and aa, respectively,
where <cite>e</cite> and <cite>h</cite> are the effect size and dominance, respectively.</p>
<p>For a model of a trait (phenotype), meaning <cite>gvalue_to_fitness</cite>
is not <cite>None</cite>, the values for the three genotypes are 0, e*h,
and e, respectively.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8.0: </span>Refactored to use attrs and inherit from
low-level C++ class</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative.asblack">
<span class="sig-name descname"><span class="pre">asblack</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Multiplicative.asblack" title="Permalink to this definition">#</a></dt>
<dd><p>Return a string representation formatted with black</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Multiplicative.asdict" title="Permalink to this definition">#</a></dt>
<dd><p>Return dict representation</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative.fromdict">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fromdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.Multiplicative.fromdict" title="Permalink to this definition">#</a></dt>
<dd><p>Build an instance from a dictionary</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative.genetic_values">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genetic_values</span></span><a class="headerlink" href="#fwdpy11.Multiplicative.genetic_values" title="Permalink to this definition">#</a></dt>
<dd><p>Return the list of genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative.maps_to_fitness">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_fitness</span></span><a class="headerlink" href="#fwdpy11.Multiplicative.maps_to_fitness" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a mapping directly to fitness, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative.maps_to_trait_value">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_trait_value</span></span><a class="headerlink" href="#fwdpy11.Multiplicative.maps_to_trait_value" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a trait value, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.Multiplicative.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#fwdpy11.Multiplicative.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Return the dimensions of the genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.GBR">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">GBR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gvalue_to_fitness</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.GBR" title="Permalink to this definition">#</a></dt>
<dd><p>The “gene-based recessive” trait model described in Thornton et al.
2013 <a class="reference external" href="http://dx.doi.org/10.1371/journal.pgen.1003258">http://dx.doi.org/10.1371/journal.pgen.1003258</a> and Sanjak et al. 2017
<a class="reference external" href="http://dx.doi.org/10.1371/journal.pgen.1006573">http://dx.doi.org/10.1371/journal.pgen.1006573</a>.</p>
<p>The trait value is the geometric mean of the sum of effect
sizes on each haplotype.
It is undefined for the case where these sums are negative.</p>
<p>This class has the following attributes, whose names
are also <cite>kwargs</cite> for intitialization.  The attribute names
also determine the order of positional arguments:</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gvalue_to_fitness</strong> (<a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GeneticValueIsTrait" title="fwdpy11.GeneticValueIsTrait"><em>fwdpy11.GeneticValueIsTrait</em></a>) – How to map trait value to fitness</p></li>
<li><p><strong>noise</strong> (<a class="reference internal" href="gvaluenoise.html#fwdpy11.GeneticValueNoise" title="fwdpy11.GeneticValueNoise"><em>fwdpy11.GeneticValueNoise</em></a>) – Random effects on trait values</p></li>
</ul>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8.0: </span>Refactored to use attrs and inherit from
low-level C++ class</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.GBR.asblack">
<span class="sig-name descname"><span class="pre">asblack</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.GBR.asblack" title="Permalink to this definition">#</a></dt>
<dd><p>Return a string representation formatted with black</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.GBR.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.GBR.asdict" title="Permalink to this definition">#</a></dt>
<dd><p>Return dict representation</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.GBR.fromdict">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fromdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.GBR.fromdict" title="Permalink to this definition">#</a></dt>
<dd><p>Build an instance from a dictionary</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.GBR.genetic_values">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genetic_values</span></span><a class="headerlink" href="#fwdpy11.GBR.genetic_values" title="Permalink to this definition">#</a></dt>
<dd><p>Return the list of genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.GBR.maps_to_fitness">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_fitness</span></span><a class="headerlink" href="#fwdpy11.GBR.maps_to_fitness" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a mapping directly to fitness, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.GBR.maps_to_trait_value">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_trait_value</span></span><a class="headerlink" href="#fwdpy11.GBR.maps_to_trait_value" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a trait value, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.GBR.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#fwdpy11.GBR.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Return the dimensions of the genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fwdpy11.</span></span><span class="sig-name descname"><span class="pre">StrictAdditiveMultivariateEffects</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ndimensions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">focal_trait</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gvalue_to_fitness</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">noise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects" title="Permalink to this definition">#</a></dt>
<dd><p>Multivariate trait values under strictly additive effects.</p>
<p>Calculate the trait value for a diploid in a
<a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a> for a multidimensional trait.</p>
<p>This class is restricted to the case of simple additive effects, meaning
that any dominance terms associated with mutations are ignored.</p>
<p>During a simulation, <a class="reference internal" href="types.html#fwdpy11.DiploidMetadata.g" title="fwdpy11.DiploidMetadata.g"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata.g</span></code></a> is filled with the
genetic value corresponding to a “focal” trait specified upon
object construction. This class has the following attributes, whose names
are also <cite>kwargs</cite> for intitialization.  The attribute names
also determine the order of positional arguments:</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ndimensions</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Number of trait dimensions</p></li>
<li><p><strong>focal_trait</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Index of the focal trait</p></li>
<li><p><strong>gvalue_to_fitness</strong> (<a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GeneticValueToFitnessMap" title="fwdpy11.GeneticValueToFitnessMap"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GeneticValueToFitnessMap</span></code></a>) – Function mapping trait value to fitness</p></li>
<li><p><strong>noise</strong> (<a class="reference internal" href="gvaluenoise.html#fwdpy11.GeneticValueNoise" title="fwdpy11.GeneticValueNoise"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GeneticValueNoise</span></code></a>) – Function adding random additive noise to trait value</p></li>
</ul>
</dd>
</dl>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 0.8.0: </span>Refactored to use attrs and inherit from
low-level C++ class</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects.asblack">
<span class="sig-name descname"><span class="pre">asblack</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects.asblack" title="Permalink to this definition">#</a></dt>
<dd><p>Return a string representation formatted with black</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects.asdict" title="Permalink to this definition">#</a></dt>
<dd><p>Return dict representation</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects.fromdict">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fromdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects.fromdict" title="Permalink to this definition">#</a></dt>
<dd><p>Build an instance from a dictionary</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects.genetic_values">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">genetic_values</span></span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects.genetic_values" title="Permalink to this definition">#</a></dt>
<dd><p>Return the list of genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects.maps_to_fitness">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_fitness</span></span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects.maps_to_fitness" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a mapping directly to fitness, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects.maps_to_trait_value">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">maps_to_trait_value</span></span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects.maps_to_trait_value" title="Permalink to this definition">#</a></dt>
<dd><p>Returns True if object represents a trait value, and
False otherwise.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.7.0.</span></p>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fwdpy11.StrictAdditiveMultivariateEffects.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#fwdpy11.StrictAdditiveMultivariateEffects.shape" title="Permalink to this definition">#</a></dt>
<dd><p>Return the dimensions of the genetic values.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
</dd></dl>

</dd></dl>

</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./pages"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../technical/writingplugins.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Writing “plugins” with C++</p>
      </div>
    </a>
    <a class="right-next"
       href="datamatrix.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">DataMatrix and StateMatrix</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            <div class="bd-footer-content__inner">
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Kevin Thornton
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div></div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>