
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Data structures related to tree sequences &#8212; fwdpy11 manual</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Calculating frequency spectra from tree sequences" href="tablefs.html" />
    <link rel="prev" title="Conceptual overview of tree sequence recording" href="tsoverview.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">fwdpy11 manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Introduction
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Installation and setup
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="userenv.html">
   Setting up user environments for installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deploymenttools.html">
   Deployment tools
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - operations on populations
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/initpops_vignette.html">
   Initializing a population
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/poptour_vignette.html">
   Properties of populations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - setting up the genetics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/genetics_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/geneticmaps_vignette.html">
   Genetic maps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/des_vignette.html">
   Distributions of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/mutationdominance_vignette.html">
   Mutations with variable dominance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmuts_vignette.html">
   Neutral mutations during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_fitness.html">
   Genetic values - mutations with direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_traits.html">
   Genetic values - simulating phenotypes/traits.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_fitness.html">
   Working example - parameters for simulating direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_trait.html">
   Working example - parameters for the simulation of a trait
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - simulation with tree sequence recording
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/evolvets_vignette.html">
   Evolving with tree sequence recording
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/recorders_vignette.html">
   Recording data during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/recapitation.html">
   Finishing a simulation with msprime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmutsafter_vignette.html">
   Adding neutral mutations to a population with existing ancestry
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - demographic modeling
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demes_vignette.html">
   Using
   <code class="docutils literal notranslate">
    <span class="pre">
     demes
    </span>
   </code>
   to specify demographic models.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_debugger_vignette.html">
   Debugging demographic models
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - varying fitness conditions across demes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/varying_fitness_conditions_across_demes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/multivariate_gaussian_effect_sizes_across_demes.html">
   The multivariate Gaussian distribution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/multivariate_lognormal_effect_sizes.html">
   The multivariate lognormal distribution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/custom_multivarate_distributions.html">
   “Custom” multivariate distributions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/special_considerations_demes_models.html">
   Special considerations for demes models
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - conditional models
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/trackmutationfates.html">
   Tracking user-specified new mutations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/selective_sweep.html">
   Selective sweeps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/incomplete_sweep.html">
   Incomplete selective sweeps
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Longer vignettes - exporting data to tskit
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/tskitconvert_vignette.html">
   Converting data to tskit format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/tskit_metadata_vignette.html">
   Working with data in tskit format
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Longer vignettes - complete examples
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/import_mutations_from_tskit.html">
   Importing mutations from tskit
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/bgs_vignette.html">
   Background selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/gss_vignette.html">
   Gaussian stabilizing selection with an optimum shift
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker.html">
   Tracking mutations over time
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker_sqlite.html">
   Tracking mutations over time using
   <code class="docutils literal notranslate">
    <span class="pre">
     sqlite3
    </span>
   </code>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Concepts
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="definitions.html">
   Definitions
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorials
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="advancedtopics.html">
   Advanced topics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Objects and concepts
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="softselection.html">
   Soft selection with discrete demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mvdes.html">
   Different effect sizes of mutations in different demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tsoverview.html">
   Conceptual overview of tree sequence recording
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Data structures related to tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tablefs.html">
   Calculating frequency spectra from tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="recorders.html">
   Time series analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Technical details
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/demes.html">
   Details of
   <code class="docutils literal notranslate">
    <span class="pre">
     demes
    </span>
   </code>
   integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/genetic_values.html">
   Technical overview of genetic value calculations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/writingplugins.html">
   Writing “plugins” with C++
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data types and fuctions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="genetic_values.html">
   Genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datamatrix.html">
   DataMatrix and StateMatrix
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="demographic_models.html">
   The demography debugger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="functions.html">
   Function documentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gslrandom.html">
   Random number distributions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gvaluenoise.html">
   Adding random noise to genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gvalue_to_fitness.html">
   Mapping genetic values to fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_params.html">
   Model parameters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="regiontypes.html">
   Setting up genomic intervals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tskit_tools.html">
   Module
   <code class="docutils literal notranslate">
    <span class="pre">
     fwdpy11.tskit_tools
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="conditional_models.html">
   Module
   <code class="docutils literal notranslate">
    <span class="pre">
     fwdpy11.conditional_models
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="types.html">
   Data types related to simulated populations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Examples
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/gss_divergent_optima.html">
   Adaptation to different optima with multivariate distribution of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/IM.html">
   Isolation-with-migration (IM)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/localadaptation.html">
   Local adaptation of a quantitative trait
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/migtest.html">
   Symmetric migration between two demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/pysnowdrift.html">
   Snowdrift model in python
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/recorder.html">
   Example of time series analysis implemented in C++
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellany
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/changelog.html">
   Changelog
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/deprecated.html">
   Deprecated features
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/developersguide.html">
   Developer’s guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/pubs.html">
   Publications using fwdpy11
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/upgrade_path.html">
   Upgrade path
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/bibliography.html">
   Literature cited
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/molpopgen/fwdpy11"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/molpopgen/fwdpy11/issues/new?title=Issue%20on%20page%20%2Fpages/tstypes.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/pages/tstypes.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download notebook file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-code"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        <a href="../_sources/pages/tstypes.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#low-level-data-types">
   Low-level data types
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#table-collections">
   Table collections
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#representing-trees">
   Representing trees
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#leaves-of-a-tree">
     “Leaves” of a tree
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#children-and-siblings">
     Children and siblings
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multiply-linked-lists">
     Multiply-linked lists
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Data structures related to tree sequences</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#low-level-data-types">
   Low-level data types
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#table-collections">
   Table collections
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#representing-trees">
   Representing trees
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#leaves-of-a-tree">
     “Leaves” of a tree
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#children-and-siblings">
     Children and siblings
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multiply-linked-lists">
     Multiply-linked lists
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="data-structures-related-to-tree-sequences">
<span id="ts-data-types"></span><h1>Data structures related to tree sequences<a class="headerlink" href="#data-structures-related-to-tree-sequences" title="Permalink to this headline">#</a></h1>
<section id="low-level-data-types">
<h2>Low-level data types<a class="headerlink" href="#low-level-data-types" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="types.html#fwdpy11.Node" title="fwdpy11.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Node</span></code></a> defines nodes</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.Edge" title="fwdpy11.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Edge</span></code></a> defines edges</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.Site" title="fwdpy11.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Site</span></code></a> defines a site (genomic location) where a mutation is present.</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.MutationRecord" title="fwdpy11.MutationRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.MutationRecord</span></code></a> defines mutations associated with <a class="reference internal" href="types.html#fwdpy11.Site" title="fwdpy11.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Site</span></code></a>
objects on trees and in <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a> objects.</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.NodeTable" title="fwdpy11.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.NodeTable</span></code></a> represents a node table</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.EdgeTable" title="fwdpy11.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.EdgeTable</span></code></a> represents an edge table</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.MutationTable" title="fwdpy11.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.MutationTable</span></code></a> represents a mutation table</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.SiteTable" title="fwdpy11.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.SiteTable</span></code></a> represents a site table</p></li>
</ul>
<p>These types are all analogous to the corresponding <a class="reference external" href="http://tskit.readthedocs.io">tskit</a> types.
Here, <a class="reference internal" href="types.html#fwdpy11.MutationRecord" title="fwdpy11.MutationRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.MutationRecord</span></code></a> plays the role of <code class="docutils literal notranslate"><span class="pre">tskit.Mutation</span></code>,
but with additional data fields that are useful for the forward-time simulations.</p>
</section>
<section id="table-collections">
<h2>Table collections<a class="headerlink" href="#table-collections" title="Permalink to this headline">#</a></h2>
<p>The above data types are encapsulated into the Python class <a class="reference internal" href="types.html#fwdpy11.TableCollection" title="fwdpy11.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.TableCollection</span></code></a>.  Instances of this
class are data fields of populations, via <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation.tables" title="fwdpy11.DiploidPopulation.tables"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation.tables</span></code></a>.</p>
</section>
<section id="representing-trees">
<span id="id1"></span><h2>Representing trees<a class="headerlink" href="#representing-trees" title="Permalink to this headline">#</a></h2>
<p>An edge table and a node table contain all the data required to generate what we may call “marginal trees”, which are the trees
corresponding to genomic intervals <span class="math notranslate nohighlight">\([left, right)\)</span>. In fwdpy11, <a class="reference internal" href="types.html#fwdpy11.TreeIterator" title="fwdpy11.TreeIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator</span></code></a> allows
left-to-tright traversal across marginal trees as well as access to their data.</p>
<p>The intent of this section is to describe the structure of the marginal trees and get some terminology out of the way.  I will use Jerome Kelleher’s
<a class="reference external" href="http://tskit.readthedocs.io">tskit</a> package to illustrate the concepts.  fwdpy11 uses the same basic data model, and it is more convenient to use trees
generated via a coalescent simulation that to run forward simulations.  Plus, we can use the great “draw” methods from
<a class="reference external" href="http://tskit.readthedocs.io">tskit</a>!</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Do not mistake the <a class="reference external" href="http://tskit.readthedocs.io">tskit</a> operations used below for the operations needed in fwdpy11.  They differ in some
important ways.  Namely, the <a class="reference external" href="http://tskit.readthedocs.io">tskit</a> API is currently richer and the fwdpy11 API is currently more low-level.
The main goal here is to define terms and visualize them.</p>
</div>
<p>We are going to look at a single tree, but the concepts apply to any of the trees generated by a simulation with nonzero
recombination rates.  First, let’s get a tree sequence using the linear-time algorithm of Hudson (1990),
which involves repeated swapping of the IDs of potential ancestors until the MRCA is reached.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import tskit
import numpy as np

# The following makes use of the Kirk Lohmueller seed.
np.random.seed(101 * 405 * 10 * 110)

tc = tskit.TableCollection(1)

nsam = 10

for i in range(nsam):
    tc.nodes.add_row(time=0, flags=1)

    nodes = np.arange(2 * nsam - 1, dtype=np.int32)
    time = 0.0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># The citation for this algorithm is
# Hudson, Richard R. 1990.
# “Gene Genealogies and the Coalescent Process.”
# Oxford Surveys in Evolutionary Biology 7 (1): 44.
n = nsam
while n &gt; 1:
    rcoal = (n * (n - 1)) / 2.0
    tcoal = np.random.exponential(1.0 / rcoal)
    time += tcoal
    tc.nodes.add_row(time=time)
    ancestor = 2 * nsam - n
    p = np.random.choice(n, 1)[0]
    tc.edges.add_row(left=0.0, right=1.0, parent=ancestor, child=nodes[p])
    nodes[p] = nodes[n - 1]
    p = np.random.choice(n - 1, 1)[0]
    tc.edges.add_row(left=0.0, right=1.0, parent=ancestor, child=nodes[p])
    nodes[p] = nodes[2 * nsam - n]
    n -= 1
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>tc.sort()
ts = tc.tree_sequence()
print(ts.first().draw(format=&quot;unicode&quot;))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>       18          
  ┏━━━━━┻━━━━━┓    
  ┃          17    
  ┃       ┏━━━┻━━┓ 
 16       ┃      ┃ 
 ┏┻━┓     ┃      ┃ 
 ┃  ┃    15      ┃ 
 ┃  ┃   ┏━┻━━┓   ┃ 
 ┃  ┃   ┃   14   ┃ 
 ┃  ┃   ┃   ┏┻┓  ┃ 
 ┃  ┃   ┃   ┃ ┃ 13 
 ┃  ┃   ┃   ┃ ┃ ┏┻┓
12  ┃   ┃   ┃ ┃ ┃ ┃
┏┻┓ ┃   ┃   ┃ ┃ ┃ ┃
┃ ┃ ┃  11   ┃ ┃ ┃ ┃
┃ ┃ ┃ ┏━┻┓  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ 10  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃
0 4 1 2 6 7 3 5 8 9
</pre></div>
</div>
</div>
</div>
<p>The above tree corresponds to a sample size of 10 haplotypes.  The tree shows the node labels.  The nodes labelled 0
through 9 correspond to the present time point–these are “alive nodes” or the “current generation” if we are thinking
about a forward simulation.  Further, we can describe the branches leading to these sample nodes as the “tips” or
“leaves” of a tree.</p>
<section id="leaves-of-a-tree">
<h3>“Leaves” of a tree<a class="headerlink" href="#leaves-of-a-tree" title="Permalink to this headline">#</a></h3>
<p>One thing that we often want to know is “how many samples descend from node <code class="docutils literal notranslate"><span class="pre">i</span></code>?” To do that, we
may look at the leaf counts attribute of a marginal tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Let&#39;s store our
# tree in a variable now
t = ts.first()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>def get_leaf_counts(tree, i):
    return len([j for j in tree.leaves(i)])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Map node ids to their leaf counts using a dict
lcmap = {i: &quot;&quot;.format(get_leaf_counts(t, i)) for i in range(len(ts.tables.nodes))}
print(t.draw(format=&quot;unicode&quot;, node_labels=lcmap))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>         
 ┏━━┻━┓  
 ┃    ┃  
 ┃   ┏┻━┓
 ┃   ┃  ┃
┏┻┓  ┃  ┃
┃ ┃  ┃  ┃
┃ ┃ ┏┻┓ ┃
┃ ┃ ┃ ┃ ┃
┃ ┃ ┃ ┏┓┃
┃ ┃ ┃ ┃┃┃
┃ ┃ ┃ ┃┃┏
┃ ┃ ┃ ┃┃┃
┏┓┃ ┃ ┃┃┃
┃┃┃ ┃ ┃┃┃
┃┃┃┏┓ ┃┃┃
┃┃┃┃┃ ┃┃┃
┃┃┃┃┏┓┃┃┃
┃┃┃┃┃┃┃┃┃
</pre></div>
</div>
</div>
</div>
</section>
<section id="children-and-siblings">
<h3>Children and siblings<a class="headerlink" href="#children-and-siblings" title="Permalink to this headline">#</a></h3>
<p>Let’s take another look at our tree, labelled with node ids:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>print(t.draw(format=&quot;unicode&quot;))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>       18          
  ┏━━━━━┻━━━━━┓    
  ┃          17    
  ┃       ┏━━━┻━━┓ 
 16       ┃      ┃ 
 ┏┻━┓     ┃      ┃ 
 ┃  ┃    15      ┃ 
 ┃  ┃   ┏━┻━━┓   ┃ 
 ┃  ┃   ┃   14   ┃ 
 ┃  ┃   ┃   ┏┻┓  ┃ 
 ┃  ┃   ┃   ┃ ┃ 13 
 ┃  ┃   ┃   ┃ ┃ ┏┻┓
12  ┃   ┃   ┃ ┃ ┃ ┃
┏┻┓ ┃   ┃   ┃ ┃ ┃ ┃
┃ ┃ ┃  11   ┃ ┃ ┃ ┃
┃ ┃ ┃ ┏━┻┓  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ 10  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃
0 4 1 2 6 7 3 5 8 9
</pre></div>
</div>
</div>
</div>
<p>Let’s ponder node 15 for a moment.  It has two immediate descendants, nodes 11 and 14.  We may consider
these the left and right children, respectively, of node 15.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>def get_children(tree, i):
    lc = tree.left_child(i)
    rc = tree.right_child(i)
    if lc == tskit.NULL and rc == tskit.NULL:
        return &quot;-&gt;NULL&quot;.format(i)
    return &quot;-&gt;&quot;.format(i) + str((lc, rc))


cmap = {i: get_children(t, i) for i in range(len(ts.tables.nodes))}
print(t.draw(format=&quot;unicode&quot;, node_labels=cmap))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                          -&gt;(16, 17)                                 
            ┏━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━┓                  
            ┃                                -&gt;(13, 15)              
            ┃                         ┏━━━━━━━━━━━┻━━━━━━━━━━━┓      
        -&gt;(1, 12)                     ┃                       ┃      
      ┏━━━━━┻━━━━┓                    ┃                       ┃      
      ┃          ┃               -&gt;(11, 14)                   ┃      
      ┃          ┃           ┏━━━━━━━━┻━━━━━━━━━┓             ┃      
      ┃          ┃           ┃              -&gt;(3, 5)          ┃      
      ┃          ┃           ┃               ┏━━┻━━━┓         ┃      
      ┃          ┃           ┃               ┃      ┃     -&gt;(8, 9)   
      ┃          ┃           ┃               ┃      ┃      ┏━━┻━━━┓  
  -&gt;(0, 4)       ┃           ┃               ┃      ┃      ┃      ┃  
   ┏━━┻━━━┓      ┃           ┃               ┃      ┃      ┃      ┃  
   ┃      ┃      ┃       -&gt;(2, 10)           ┃      ┃      ┃      ┃  
   ┃      ┃      ┃      ┏━━━━┻━━━━┓          ┃      ┃      ┃      ┃  
   ┃      ┃      ┃      ┃     -&gt;(6, 7)       ┃      ┃      ┃      ┃  
   ┃      ┃      ┃      ┃      ┏━━┻━━━┓      ┃      ┃      ┃      ┃  
-&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL
</pre></div>
</div>
</div>
</div>
<p>Likewise, we may look at the sibling relationships amongst nodes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>def get_sibs(tree, i):
    ls = tree.left_sib(i)
    rs = tree.right_sib(i)
    if ls == tskit.NULL and rs == tskit.NULL:
        return &quot;-&gt;NULL&quot;.format(i)
    return &quot;-&gt;&quot;.format(i) + str((ls, rs))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>smap = {i: get_sibs(t, i) for i in range(len(ts.tables.nodes))}
print(t.draw(format=&quot;unicode&quot;, node_labels=smap))
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                                           -&gt;NULL                                                    
                 ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━┓                          
                 ┃                                                   -&gt;(16, -1)                      
                 ┃                                        ┏━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┓         
            -&gt;(-1, 17)                                    ┃                                ┃         
         ┏━━━━━━━┻━━━━━━━┓                                ┃                                ┃         
         ┃               ┃                           -&gt;(13, -1)                            ┃         
         ┃               ┃                  ┏━━━━━━━━━━━━━┻━━━━━━━━━━━━┓                   ┃         
         ┃               ┃                  ┃                     -&gt;(11, -1)               ┃         
         ┃               ┃                  ┃                     ┏━━━━┻━━━━┓              ┃         
         ┃               ┃                  ┃                     ┃         ┃         -&gt;(-1, 15)     
         ┃               ┃                  ┃                     ┃         ┃         ┏━━━━┻━━━━┓    
     -&gt;(1, -1)           ┃                  ┃                     ┃         ┃         ┃         ┃    
    ┏━━━━┻━━━━┓          ┃                  ┃                     ┃         ┃         ┃         ┃    
    ┃         ┃          ┃             -&gt;(-1, 14)                 ┃         ┃         ┃         ┃    
    ┃         ┃          ┃          ┏━━━━━━━┻━━━━━━┓              ┃         ┃         ┃         ┃    
    ┃         ┃          ┃          ┃          -&gt;(2, -1)          ┃         ┃         ┃         ┃    
    ┃         ┃          ┃          ┃         ┏━━━━┻━━━━┓         ┃         ┃         ┃         ┃    
-&gt;(-1, 4) -&gt;(0, -1) -&gt;(-1, 12) -&gt;(-1, 10) -&gt;(-1, 7) -&gt;(6, -1) -&gt;(-1, 5) -&gt;(3, -1) -&gt;(-1, 9) -&gt;(8, -1)
</pre></div>
</div>
</div>
</div>
</section>
<section id="multiply-linked-lists">
<h3>Multiply-linked lists<a class="headerlink" href="#multiply-linked-lists" title="Permalink to this headline">#</a></h3>
<p>Under the hood, the data structures representing marginal trees consist of several arrays
represening a multiply-linked list allowing traversal up/down/left/right along a marginal tree.
By convention, the value -1 is taken as a NULL value, signifying that there are no more nodes in
“that” direction along the tree.</p>
<p>We can look directly at what theses arrays look like in our tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>nnodes = len(ts.tables.nodes)
# First, let&#39;s write down our node ids:
print([i for i in range(nnodes)])
# Now, get the parents of each node, moving &quot;up&quot; the tree
print([t.parent(i) for i in range(nnodes)])
# The left child list allows moving &quot;down left&quot; along a tree
print([t.left_child(i) for i in range(nnodes)])
# The right child list allows moving &quot;down right&quot; along a tree
print([t.right_child(i) for i in range(nnodes)])
# The left sib list allows moving &quot;left&quot; along a tree
print([t.left_sib(i) for i in range(nnodes)])
# The right sib list allows moving &quot;right&quot; along a tree
print([t.right_sib(i) for i in range(nnodes)])
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
[12, 16, 11, 14, 12, 14, 10, 10, 13, 13, 11, 15, 16, 17, 15, 17, 18, 18, -1]
[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, 2, 0, 8, 3, 11, 1, 13, 16]
[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 7, 10, 4, 9, 5, 14, 12, 15, 17]
[-1, -1, -1, -1, 0, 3, -1, 6, -1, 8, 2, -1, 1, -1, 11, 13, -1, 16, -1]
[4, 12, 10, 5, -1, -1, 7, -1, 9, -1, -1, 14, -1, 15, -1, -1, 17, -1, -1]
</pre></div>
</div>
</div>
</div>
<p>These lists are interpreted a lists of nodes referring to other nodes.  For example, the value 10 in position 0 of the
parents list means, “The parent of the node with index 0 has index 10”, where the indexes refer to the node table.</p>
<p>Given the above lists, you may start at any valid node id (<em>e.g.</em> a value not equal to -1 in the first array),
and then “walk” in any direction you choose along the tree until you hit a value of -1, meaning that you cannot proceed
any further.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The trees generated by <code class="docutils literal notranslate"><span class="pre">msprime.simulate</span></code> are bifurcating, which is a consequence of simulating from the Kingman
coalescent.  In forward-time simulations, it is not uncommon to have more than two descendants of a node.  When that
happens, left_child and right_child refer to the left-most and right-most children, respectively. Thus, to “walk”
along the descendants of a node, you proceed to left_child, and then march along right_sib until a value of -1 is
seen.  This “walking” method is the same as what you would do for a bifurcating tree, but I want to point out that
counting the number of immediate descendants of a node requires counting the number of steps that the walk requires,
and that it may be longer than two steps.</p>
</div>
<p>The last five arrays show above correspond to the following attributes of <a class="reference internal" href="types.html#fwdpy11.TreeIterator" title="fwdpy11.TreeIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.parent" title="fwdpy11.TreeIterator.parent"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.parent</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.left_child" title="fwdpy11.TreeIterator.left_child"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.left_child</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.right_child" title="fwdpy11.TreeIterator.right_child"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.right_child</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.left_sib" title="fwdpy11.TreeIterator.left_sib"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.left_sib</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.right_sib" title="fwdpy11.TreeIterator.right_sib"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.right_sib</span></code></a></p></li>
</ul>
<p>The key to efficiency is how these linked lists are updated as you move from tree <span class="math notranslate nohighlight">\(i\)</span> to tree <span class="math notranslate nohighlight">\(i+1\)</span>.  Given
an approprate set of indexes, these lists are only updated at the positions that differ between the two trees.  In
practice, adjacent trees are highly-correlated, meaning that very few values need updating. The
construction of these indexes is described immediatlely above the description of “Algorithm T” in the Kelleher et al. (2016) paper describing <a class="reference external" href="http://msprime.readthedocs.io">msprime</a>.</p>
<p>The building of those indexes is a key determinant of performance for algorithms on tree sequences.  Fortunately, you
don’t need to worry about that, as that machinery is hidden in the internals of fwdpp (and tskit).</p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pages"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="tsoverview.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Conceptual overview of tree sequence recording</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="tablefs.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Calculating frequency spectra from tree sequences</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Kevin Thornton<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>