
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Data structures related to tree sequences &#8212; fwdpy11 manual</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet">
  <link href="../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Calculating frequency spectra from tree sequences" href="tablefs.html" />
    <link rel="prev" title="Conceptual overview of tree sequence recording" href="tsoverview.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">fwdpy11 manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../intro.html">
   Introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Installation and setup
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="userenv.html">
   Setting up user environments for installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deploymenttools.html">
   Deployment tools
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - operations on populations
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/initpops_vignette.html">
   Initializing a population
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/poptour_vignette.html">
   Properties of populations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - setting up the genetics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/genetics_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/geneticmaps_vignette.html">
   Genetic maps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/des_vignette.html">
   Distributions of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/mutationdominance_vignette.html">
   Mutations with variable dominance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmuts_vignette.html">
   Neutral mutations during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_fitness.html">
   Genetic values - mutations with direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_traits.html">
   Genetic values - simulating phenotypes/traits.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_fitness.html">
   Working example - parameters for simulating direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_trait.html">
   Working example - parameters for the simulation of a trait
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - simulation with tree sequence recording
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/evolvets_vignette.html">
   Evolving with tree sequence recording
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/recorders_vignette.html">
   Recording data during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/recapitation.html">
   Finishing a simulation with msprime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmutsafter_vignette.html">
   Adding neutral mutations to a population with existing ancestry
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - demographic modeling
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demes_vignette.html">
   Using
   <code class="docutils literal notranslate">
    <span class="pre">
     demes
    </span>
   </code>
   to specify demographic models.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_debugger_vignette.html">
   Debugging demographic models
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - special case models
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/trackmutationfates.html">
   Need title
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/selective_sweep.html">
   Selective sweeps
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Longer vignettes - exporting data to tskit
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/tskitconvert_vignette.html">
   Converting data to tskit format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/tskit_metadata_vignette.html">
   Working with data in tskit format
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Longer vignettes - complete examples
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/bgs_vignette.html">
   Background selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/gss_vignette.html">
   Gaussian stabilizing selection with an optimum shift
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker.html">
   Tracking mutations over time
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker_sqlite.html">
   Tracking mutations over time using
   <code class="docutils literal notranslate">
    <span class="pre">
     sqlite3
    </span>
   </code>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Concepts
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="definitions.html">
   Definitions
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorials
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial.html">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="advancedtopics.html">
   Advanced topics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Objects and concepts
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="softselection.html">
   Soft selection with discrete demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mvdes.html">
   Different effect sizes of mutations in different demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tsoverview.html">
   Conceptual overview of tree sequence recording
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Data structures related to tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tablefs.html">
   Calculating frequency spectra from tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="recorders.html">
   Time series analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Technical details
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/demes.html">
   Details of
   <code class="docutils literal notranslate">
    <span class="pre">
     demes
    </span>
   </code>
   integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/genetic_values.html">
   Technical overview of genetic value calculations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/writingplugins.html">
   Writing “plugins” with C++
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data types and fuctions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="genetic_values.html">
   Genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datamatrix.html">
   DataMatrix and StateMatrix
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="demographic_models.html">
   The demography debugger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="functions.html">
   Function documentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gslrandom.html">
   Random number distributions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gvaluenoise.html">
   Adding random noise to genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gvalue_to_fitness.html">
   Mapping genetic values to fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_params.html">
   Model parameters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="regiontypes.html">
   Setting up genomic intervals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tskit_tools.html">
   Module
   <code class="docutils literal notranslate">
    <span class="pre">
     fwdpy11.tskit_tools
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="conditional_models.html">
   Module
   <code class="docutils literal notranslate">
    <span class="pre">
     fwdpy11.conditional_models
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="types.html">
   Data types related to simulated populations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Examples
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/gss_divergent_optima.html">
   Adaptation to different optima with multivariate distribution of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/IM.html">
   Isolation-with-migration (IM)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/localadaptation.html">
   Local adaptation of a quantitative trait
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/migtest.html">
   Symmetric migration between two demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/pysnowdrift.html">
   Snowdrift model in python
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/recorder.html">
   Example of time series analysis implemented in C++
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellany
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/changelog.html">
   Changelog
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/deprecated.html">
   Deprecated features
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/developersguide.html">
   Developer’s guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/pubs.html">
   Publications using fwdpy11
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/upgrade_path.html">
   Upgrade path
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/bibliography.html">
   Literature cited
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="../_sources/pages/tstypes.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/pages/tstypes.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/molpopgen/fwdpy11"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/molpopgen/fwdpy11/issues/new?title=Issue%20on%20page%20%2Fpages/tstypes.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#low-level-data-types">
   Low-level data types
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#table-collections">
   Table collections
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#representing-trees">
   Representing trees
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#leaves-of-a-tree">
     “Leaves” of a tree
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#children-and-siblings">
     Children and siblings
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multiply-linked-lists">
     Multiply-linked lists
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="data-structures-related-to-tree-sequences">
<span id="ts-data-types"></span><h1>Data structures related to tree sequences<a class="headerlink" href="#data-structures-related-to-tree-sequences" title="Permalink to this headline">¶</a></h1>
<div class="section" id="low-level-data-types">
<h2>Low-level data types<a class="headerlink" href="#low-level-data-types" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="types.html#fwdpy11.Node" title="fwdpy11.Node"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Node</span></code></a> defines nodes</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.Edge" title="fwdpy11.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Edge</span></code></a> defines edges</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.Site" title="fwdpy11.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Site</span></code></a> defines a site (genomic location) where a mutation is present.</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.MutationRecord" title="fwdpy11.MutationRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.MutationRecord</span></code></a> defines mutations associated with <a class="reference internal" href="types.html#fwdpy11.Site" title="fwdpy11.Site"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Site</span></code></a>
objects on trees and in <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a> objects.</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.NodeTable" title="fwdpy11.NodeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.NodeTable</span></code></a> represents a node table</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.EdgeTable" title="fwdpy11.EdgeTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.EdgeTable</span></code></a> represents an edge table</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.MutationTable" title="fwdpy11.MutationTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.MutationTable</span></code></a> represents a mutation table</p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.SiteTable" title="fwdpy11.SiteTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.SiteTable</span></code></a> represents a site table</p></li>
</ul>
<p>These types are all analogous to the corresponding <a class="reference external" href="http://tskit.readthedocs.io">tskit</a> types.
Here, <a class="reference internal" href="types.html#fwdpy11.MutationRecord" title="fwdpy11.MutationRecord"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.MutationRecord</span></code></a> plays the role of <code class="docutils literal notranslate"><span class="pre">tskit.Mutation</span></code>,
but with additional data fields that are useful for the forward-time simulations.</p>
</div>
<div class="section" id="table-collections">
<h2>Table collections<a class="headerlink" href="#table-collections" title="Permalink to this headline">¶</a></h2>
<p>The above data types are encapsulated into the Python class <a class="reference internal" href="types.html#fwdpy11.TableCollection" title="fwdpy11.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.TableCollection</span></code></a>.  Instances of this
class are data fields of populations, via <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation.tables" title="fwdpy11.DiploidPopulation.tables"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation.tables</span></code></a>.</p>
</div>
<div class="section" id="representing-trees">
<span id="id1"></span><h2>Representing trees<a class="headerlink" href="#representing-trees" title="Permalink to this headline">¶</a></h2>
<p>An edge table and a node table contain all the data required to generate what we may call “marginal trees”, which are the trees
corresponding to genomic intervals <span class="math notranslate nohighlight">\([left, right)\)</span>. In fwdpy11, <a class="reference internal" href="types.html#fwdpy11.TreeIterator" title="fwdpy11.TreeIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator</span></code></a> allows
left-to-tright traversal across marginal trees as well as access to their data.</p>
<p>The intent of this section is to describe the structure of the marginal trees and get some terminology out of the way.  I will use Jerome Kelleher’s
<a class="reference external" href="http://tskit.readthedocs.io">tskit</a> package to illustrate the concepts.  fwdpy11 uses the same basic data model, and it is more convenient to use trees
generated via a coalescent simulation that to run forward simulations.  Plus, we can use the great “draw” methods from
<a class="reference external" href="http://tskit.readthedocs.io">tskit</a>!</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Do not mistake the <a class="reference external" href="http://tskit.readthedocs.io">tskit</a> operations used below for the operations needed in fwdpy11.  They differ in some
important ways.  Namely, the <a class="reference external" href="http://tskit.readthedocs.io">tskit</a> API is currently richer and the fwdpy11 API is currently more low-level.
The main goal here is to define terms and visualize them.</p>
</div>
<p>We are going to look at a single tree, but the concepts apply to any of the trees generated by a simulation with nonzero
recombination rates.  First, let’s get a tree sequence using the linear-time algorithm of Hudson (1990),
which involves repeated swapping of the IDs of potential ancestors until the MRCA is reached.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">tskit</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># The following makes use of the Kirk Lohmueller seed.</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">101</span> <span class="o">*</span> <span class="mi">405</span> <span class="o">*</span> <span class="mi">10</span> <span class="o">*</span> <span class="mi">110</span><span class="p">)</span>

<span class="n">tc</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">TableCollection</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

<span class="n">nsam</span> <span class="o">=</span> <span class="mi">10</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nsam</span><span class="p">):</span>
    <span class="n">tc</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">nodes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">nsam</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
    <span class="n">time</span> <span class="o">=</span> <span class="mf">0.0</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># The citation for this algorithm is</span>
<span class="c1"># Hudson, Richard R. 1990.</span>
<span class="c1"># “Gene Genealogies and the Coalescent Process.”</span>
<span class="c1"># Oxford Surveys in Evolutionary Biology 7 (1): 44.</span>
<span class="n">n</span> <span class="o">=</span> <span class="n">nsam</span>
<span class="k">while</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
    <span class="n">rcoal</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">))</span> <span class="o">/</span> <span class="mf">2.0</span>
    <span class="n">tcoal</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">exponential</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">rcoal</span><span class="p">)</span>
    <span class="n">time</span> <span class="o">+=</span> <span class="n">tcoal</span>
    <span class="n">tc</span><span class="o">.</span><span class="n">nodes</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">time</span><span class="o">=</span><span class="n">time</span><span class="p">)</span>
    <span class="n">ancestor</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">nsam</span> <span class="o">-</span> <span class="n">n</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">tc</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">ancestor</span><span class="p">,</span> <span class="n">child</span><span class="o">=</span><span class="n">nodes</span><span class="p">[</span><span class="n">p</span><span class="p">])</span>
    <span class="n">nodes</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span> <span class="n">nodes</span><span class="p">[</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">tc</span><span class="o">.</span><span class="n">edges</span><span class="o">.</span><span class="n">add_row</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">ancestor</span><span class="p">,</span> <span class="n">child</span><span class="o">=</span><span class="n">nodes</span><span class="p">[</span><span class="n">p</span><span class="p">])</span>
    <span class="n">nodes</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span> <span class="n">nodes</span><span class="p">[</span><span class="mi">2</span> <span class="o">*</span> <span class="n">nsam</span> <span class="o">-</span> <span class="n">n</span><span class="p">]</span>
    <span class="n">n</span> <span class="o">-=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tc</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">tree_sequence</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>       18          
  ┏━━━━━┻━━━━━┓    
  ┃          17    
  ┃       ┏━━━┻━━┓ 
 16       ┃      ┃ 
 ┏┻━┓     ┃      ┃ 
 ┃  ┃    15      ┃ 
 ┃  ┃   ┏━┻━━┓   ┃ 
 ┃  ┃   ┃   14   ┃ 
 ┃  ┃   ┃   ┏┻┓  ┃ 
 ┃  ┃   ┃   ┃ ┃ 13 
 ┃  ┃   ┃   ┃ ┃ ┏┻┓
12  ┃   ┃   ┃ ┃ ┃ ┃
┏┻┓ ┃   ┃   ┃ ┃ ┃ ┃
┃ ┃ ┃  11   ┃ ┃ ┃ ┃
┃ ┃ ┃ ┏━┻┓  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ 10  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃
0 4 1 2 6 7 3 5 8 9
</pre></div>
</div>
</div>
</div>
<p>The above tree corresponds to a sample size of 10 haplotypes.  The tree shows the node labels.  The nodes labelled 0
through 9 correspond to the present time point–these are “alive nodes” or the “current generation” if we are thinking
about a forward simulation.  Further, we can describe the branches leading to these sample nodes as the “tips” or
“leaves” of a tree.</p>
<div class="section" id="leaves-of-a-tree">
<h3>“Leaves” of a tree<a class="headerlink" href="#leaves-of-a-tree" title="Permalink to this headline">¶</a></h3>
<p>One thing that we often want to know is “how many samples descend from node <code class="docutils literal notranslate"><span class="pre">i</span></code>?” To do that, we
may look at the leaf counts attribute of a marginal tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Let&#39;s store our</span>
<span class="c1"># tree in a variable now</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">ts</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_leaf_counts</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">([</span><span class="n">j</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">leaves</span><span class="p">(</span><span class="n">i</span><span class="p">)])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Map node ids to their leaf counts using a dict</span>
<span class="n">lcmap</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">get_leaf_counts</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">))}</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">,</span> <span class="n">node_labels</span><span class="o">=</span><span class="n">lcmap</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>         
 ┏━━┻━┓  
 ┃    ┃  
 ┃   ┏┻━┓
 ┃   ┃  ┃
┏┻┓  ┃  ┃
┃ ┃  ┃  ┃
┃ ┃ ┏┻┓ ┃
┃ ┃ ┃ ┃ ┃
┃ ┃ ┃ ┏┓┃
┃ ┃ ┃ ┃┃┃
┃ ┃ ┃ ┃┃┏
┃ ┃ ┃ ┃┃┃
┏┓┃ ┃ ┃┃┃
┃┃┃ ┃ ┃┃┃
┃┃┃┏┓ ┃┃┃
┃┃┃┃┃ ┃┃┃
┃┃┃┃┏┓┃┃┃
┃┃┃┃┃┃┃┃┃
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="children-and-siblings">
<h3>Children and siblings<a class="headerlink" href="#children-and-siblings" title="Permalink to this headline">¶</a></h3>
<p>Let’s take another look at our tree, labelled with node ids:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>       18          
  ┏━━━━━┻━━━━━┓    
  ┃          17    
  ┃       ┏━━━┻━━┓ 
 16       ┃      ┃ 
 ┏┻━┓     ┃      ┃ 
 ┃  ┃    15      ┃ 
 ┃  ┃   ┏━┻━━┓   ┃ 
 ┃  ┃   ┃   14   ┃ 
 ┃  ┃   ┃   ┏┻┓  ┃ 
 ┃  ┃   ┃   ┃ ┃ 13 
 ┃  ┃   ┃   ┃ ┃ ┏┻┓
12  ┃   ┃   ┃ ┃ ┃ ┃
┏┻┓ ┃   ┃   ┃ ┃ ┃ ┃
┃ ┃ ┃  11   ┃ ┃ ┃ ┃
┃ ┃ ┃ ┏━┻┓  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ 10  ┃ ┃ ┃ ┃
┃ ┃ ┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃
0 4 1 2 6 7 3 5 8 9
</pre></div>
</div>
</div>
</div>
<p>Let’s ponder node 15 for a moment.  It has two immediate descendants, nodes 11 and 14.  We may consider
these the left and right children, respectively, of node 15.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_children</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
    <span class="n">lc</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">left_child</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">rc</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">right_child</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">lc</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span> <span class="ow">and</span> <span class="n">rc</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;-&gt;NULL&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;-&gt;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="nb">str</span><span class="p">((</span><span class="n">lc</span><span class="p">,</span> <span class="n">rc</span><span class="p">))</span>


<span class="n">cmap</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="n">get_children</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">))}</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">,</span> <span class="n">node_labels</span><span class="o">=</span><span class="n">cmap</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                          -&gt;(16, 17)                                 
            ┏━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━┓                  
            ┃                                -&gt;(13, 15)              
            ┃                         ┏━━━━━━━━━━━┻━━━━━━━━━━━┓      
        -&gt;(1, 12)                     ┃                       ┃      
      ┏━━━━━┻━━━━┓                    ┃                       ┃      
      ┃          ┃               -&gt;(11, 14)                   ┃      
      ┃          ┃           ┏━━━━━━━━┻━━━━━━━━━┓             ┃      
      ┃          ┃           ┃              -&gt;(3, 5)          ┃      
      ┃          ┃           ┃               ┏━━┻━━━┓         ┃      
      ┃          ┃           ┃               ┃      ┃     -&gt;(8, 9)   
      ┃          ┃           ┃               ┃      ┃      ┏━━┻━━━┓  
  -&gt;(0, 4)       ┃           ┃               ┃      ┃      ┃      ┃  
   ┏━━┻━━━┓      ┃           ┃               ┃      ┃      ┃      ┃  
   ┃      ┃      ┃       -&gt;(2, 10)           ┃      ┃      ┃      ┃  
   ┃      ┃      ┃      ┏━━━━┻━━━━┓          ┃      ┃      ┃      ┃  
   ┃      ┃      ┃      ┃     -&gt;(6, 7)       ┃      ┃      ┃      ┃  
   ┃      ┃      ┃      ┃      ┏━━┻━━━┓      ┃      ┃      ┃      ┃  
-&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL -&gt;NULL
</pre></div>
</div>
</div>
</div>
<p>Likewise, we may look at the sibling relationships amongst nodes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">get_sibs</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
    <span class="n">ls</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">left_sib</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">rs</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">right_sib</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">ls</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span> <span class="ow">and</span> <span class="n">rs</span> <span class="o">==</span> <span class="n">tskit</span><span class="o">.</span><span class="n">NULL</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;-&gt;NULL&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;-&gt;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="nb">str</span><span class="p">((</span><span class="n">ls</span><span class="p">,</span> <span class="n">rs</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">smap</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="n">get_sibs</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">))}</span>
<span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">,</span> <span class="n">node_labels</span><span class="o">=</span><span class="n">smap</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                                           -&gt;NULL                                                    
                 ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━┓                          
                 ┃                                                   -&gt;(16, -1)                      
                 ┃                                        ┏━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┓         
            -&gt;(-1, 17)                                    ┃                                ┃         
         ┏━━━━━━━┻━━━━━━━┓                                ┃                                ┃         
         ┃               ┃                           -&gt;(13, -1)                            ┃         
         ┃               ┃                  ┏━━━━━━━━━━━━━┻━━━━━━━━━━━━┓                   ┃         
         ┃               ┃                  ┃                     -&gt;(11, -1)               ┃         
         ┃               ┃                  ┃                     ┏━━━━┻━━━━┓              ┃         
         ┃               ┃                  ┃                     ┃         ┃         -&gt;(-1, 15)     
         ┃               ┃                  ┃                     ┃         ┃         ┏━━━━┻━━━━┓    
     -&gt;(1, -1)           ┃                  ┃                     ┃         ┃         ┃         ┃    
    ┏━━━━┻━━━━┓          ┃                  ┃                     ┃         ┃         ┃         ┃    
    ┃         ┃          ┃             -&gt;(-1, 14)                 ┃         ┃         ┃         ┃    
    ┃         ┃          ┃          ┏━━━━━━━┻━━━━━━┓              ┃         ┃         ┃         ┃    
    ┃         ┃          ┃          ┃          -&gt;(2, -1)          ┃         ┃         ┃         ┃    
    ┃         ┃          ┃          ┃         ┏━━━━┻━━━━┓         ┃         ┃         ┃         ┃    
-&gt;(-1, 4) -&gt;(0, -1) -&gt;(-1, 12) -&gt;(-1, 10) -&gt;(-1, 7) -&gt;(6, -1) -&gt;(-1, 5) -&gt;(3, -1) -&gt;(-1, 9) -&gt;(8, -1)
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="multiply-linked-lists">
<h3>Multiply-linked lists<a class="headerlink" href="#multiply-linked-lists" title="Permalink to this headline">¶</a></h3>
<p>Under the hood, the data structures representing marginal trees consist of several arrays
represening a multiply-linked list allowing traversal up/down/left/right along a marginal tree.
By convention, the value -1 is taken as a NULL value, signifying that there are no more nodes in
“that” direction along the tree.</p>
<p>We can look directly at what theses arrays look like in our tree:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nnodes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">nodes</span><span class="p">)</span>
<span class="c1"># First, let&#39;s write down our node ids:</span>
<span class="nb">print</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nnodes</span><span class="p">)])</span>
<span class="c1"># Now, get the parents of each node, moving &quot;up&quot; the tree</span>
<span class="nb">print</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nnodes</span><span class="p">)])</span>
<span class="c1"># The left child list allows moving &quot;down left&quot; along a tree</span>
<span class="nb">print</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">left_child</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nnodes</span><span class="p">)])</span>
<span class="c1"># The right child list allows moving &quot;down right&quot; along a tree</span>
<span class="nb">print</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">right_child</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nnodes</span><span class="p">)])</span>
<span class="c1"># The left sib list allows moving &quot;left&quot; along a tree</span>
<span class="nb">print</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">left_sib</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nnodes</span><span class="p">)])</span>
<span class="c1"># The right sib list allows moving &quot;right&quot; along a tree</span>
<span class="nb">print</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">right_sib</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nnodes</span><span class="p">)])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
[12, 16, 11, 14, 12, 14, 10, 10, 13, 13, 11, 15, 16, 17, 15, 17, 18, 18, -1]
[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, 2, 0, 8, 3, 11, 1, 13, 16]
[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 7, 10, 4, 9, 5, 14, 12, 15, 17]
[-1, -1, -1, -1, 0, 3, -1, 6, -1, 8, 2, -1, 1, -1, 11, 13, -1, 16, -1]
[4, 12, 10, 5, -1, -1, 7, -1, 9, -1, -1, 14, -1, 15, -1, -1, 17, -1, -1]
</pre></div>
</div>
</div>
</div>
<p>These lists are interpreted a lists of nodes referring to other nodes.  For example, the value 10 in position 0 of the
parents list means, “The parent of the node with index 0 has index 10”, where the indexes refer to the node table.</p>
<p>Given the above lists, you may start at any valid node id (<em>e.g.</em> a value not equal to -1 in the first array),
and then “walk” in any direction you choose along the tree until you hit a value of -1, meaning that you cannot proceed
any further.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The trees generated by <code class="docutils literal notranslate"><span class="pre">msprime.simulate</span></code> are bifurcating, which is a consequence of simulating from the Kingman
coalescent.  In forward-time simulations, it is not uncommon to have more than two descendants of a node.  When that
happens, left_child and right_child refer to the left-most and right-most children, respectively. Thus, to “walk”
along the descendants of a node, you proceed to left_child, and then march along right_sib until a value of -1 is
seen.  This “walking” method is the same as what you would do for a bifurcating tree, but I want to point out that
counting the number of immediate descendants of a node requires counting the number of steps that the walk requires,
and that it may be longer than two steps.</p>
</div>
<p>The last five arrays show above correspond to the following attributes of <a class="reference internal" href="types.html#fwdpy11.TreeIterator" title="fwdpy11.TreeIterator"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator</span></code></a>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.parent" title="fwdpy11.TreeIterator.parent"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.parent</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.left_child" title="fwdpy11.TreeIterator.left_child"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.left_child</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.right_child" title="fwdpy11.TreeIterator.right_child"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.right_child</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.left_sib" title="fwdpy11.TreeIterator.left_sib"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.left_sib</span></code></a></p></li>
<li><p><a class="reference internal" href="types.html#fwdpy11.TreeIterator.right_sib" title="fwdpy11.TreeIterator.right_sib"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TreeIterator.right_sib</span></code></a></p></li>
</ul>
<p>The key to efficiency is how these linked lists are updated as you move from tree <span class="math notranslate nohighlight">\(i\)</span> to tree <span class="math notranslate nohighlight">\(i+1\)</span>.  Given
an approprate set of indexes, these lists are only updated at the positions that differ between the two trees.  In
practice, adjacent trees are highly-correlated, meaning that very few values need updating. The
construction of these indexes is described immediatlely above the description of “Algorithm T” in the Kelleher et al. (2016) paper describing <a class="reference external" href="http://msprime.readthedocs.io">msprime</a>.</p>
<p>The building of those indexes is a key determinant of performance for algorithms on tree sequences.  Fortunately, you
don’t need to worry about that, as that machinery is hidden in the internals of fwdpp (and tskit).</p>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./pages"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="tsoverview.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Conceptual overview of tree sequence recording</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="tablefs.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Calculating frequency spectra from tree sequences</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By Kevin Thornton<br/>
        
            &copy; Copyright 2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>