
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Tutorial &#8212; fwdpy11 manual</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Advanced topics" href="advancedtopics.html" />
    <link rel="prev" title="Definitions" href="definitions.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
      
      
      <h1 class="site-logo" id="site-title">fwdpy11 manual</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    Introduction
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Installation and setup
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="userenv.html">
   Setting up user environments for installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deploymenttools.html">
   Deployment tools
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - operations on populations
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/initpops_vignette.html">
   Initializing a population
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/poptour_vignette.html">
   Properties of populations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - setting up the genetics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/genetics_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/geneticmaps_vignette.html">
   Genetic maps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/des_vignette.html">
   Distributions of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/mutationdominance_vignette.html">
   Mutations with variable dominance
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmuts_vignette.html">
   Neutral mutations during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_fitness.html">
   Genetic values - mutations with direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/gvalues_traits.html">
   Genetic values - simulating phenotypes/traits.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_fitness.html">
   Working example - parameters for simulating direct effects on fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/workingexample_trait.html">
   Working example - parameters for the simulation of a trait
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - simulation with tree sequence recording
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/evolvets_vignette.html">
   Evolving with tree sequence recording
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/recorders_vignette.html">
   Recording data during a simulation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/recapitation.html">
   Finishing a simulation with msprime
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/neutralmutsafter_vignette.html">
   Adding neutral mutations to a population with existing ancestry
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - demographic modeling
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_vignettes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demes_vignette.html">
   Using
   <code class="docutils literal notranslate">
    <span class="pre">
     demes
    </span>
   </code>
   to specify demographic models.
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/demography_debugger_vignette.html">
   Debugging demographic models
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - varying fitness conditions across demes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/varying_fitness_conditions_across_demes_intro.html">
   Introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/multivariate_gaussian_effect_sizes_across_demes.html">
   The multivariate Gaussian distribution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/multivariate_lognormal_effect_sizes.html">
   The multivariate lognormal distribution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/custom_multivarate_distributions.html">
   “Custom” multivariate distributions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/special_considerations_demes_models.html">
   Special considerations for demes models
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Short vignettes - conditional models
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/trackmutationfates.html">
   Tracking user-specified new mutations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/selective_sweep.html">
   Selective sweeps
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../short_vignettes/incomplete_sweep.html">
   Incomplete selective sweeps
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Longer vignettes - exporting data to tskit
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/tskitconvert_vignette.html">
   Converting data to tskit format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/tskit_metadata_vignette.html">
   Working with data in tskit format
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Longer vignettes - complete examples
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/bgs_vignette.html">
   Background selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/gss_vignette.html">
   Gaussian stabilizing selection with an optimum shift
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker.html">
   Tracking mutations over time
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../long_vignettes/pyfreqtracker_sqlite.html">
   Tracking mutations over time using
   <code class="docutils literal notranslate">
    <span class="pre">
     sqlite3
    </span>
   </code>
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Concepts
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="definitions.html">
   Definitions
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Tutorials
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Tutorial
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="advancedtopics.html">
   Advanced topics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Objects and concepts
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="softselection.html">
   Soft selection with discrete demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mvdes.html">
   Different effect sizes of mutations in different demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tsoverview.html">
   Conceptual overview of tree sequence recording
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tstypes.html">
   Data structures related to tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tablefs.html">
   Calculating frequency spectra from tree sequences
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="recorders.html">
   Time series analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Technical details
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/demes.html">
   Details of
   <code class="docutils literal notranslate">
    <span class="pre">
     demes
    </span>
   </code>
   integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/genetic_values.html">
   Technical overview of genetic value calculations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../technical/writingplugins.html">
   Writing “plugins” with C++
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data types and fuctions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="genetic_values.html">
   Genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datamatrix.html">
   DataMatrix and StateMatrix
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="demographic_models.html">
   The demography debugger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="functions.html">
   Function documentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gslrandom.html">
   Random number distributions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gvaluenoise.html">
   Adding random noise to genetic values
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gvalue_to_fitness.html">
   Mapping genetic values to fitness
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_params.html">
   Model parameters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="regiontypes.html">
   Setting up genomic intervals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tskit_tools.html">
   Module
   <code class="docutils literal notranslate">
    <span class="pre">
     fwdpy11.tskit_tools
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="conditional_models.html">
   Module
   <code class="docutils literal notranslate">
    <span class="pre">
     fwdpy11.conditional_models
    </span>
   </code>
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="types.html">
   Data types related to simulated populations
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Examples
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/gss_divergent_optima.html">
   Adaptation to different optima with multivariate distribution of effect sizes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/IM.html">
   Isolation-with-migration (IM)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/localadaptation.html">
   Local adaptation of a quantitative trait
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/migtest.html">
   Symmetric migration between two demes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/pysnowdrift.html">
   Snowdrift model in python
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../examples/recorder.html">
   Example of time series analysis implemented in C++
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Miscellany
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/changelog.html">
   Changelog
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/deprecated.html">
   Deprecated features
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/developersguide.html">
   Developer’s guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/pubs.html">
   Publications using fwdpy11
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/upgrade_path.html">
   Upgrade path
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../misc/bibliography.html">
   Literature cited
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/molpopgen/fwdpy11"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/molpopgen/fwdpy11/issues/new?title=Issue%20on%20page%20%2Fpages/tutorial.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/pages/tutorial.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download notebook file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-code"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        <a href="../_sources/pages/tutorial.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#initializing-a-population">
   Initializing a population
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#looking-at-individual-metadata">
     Looking at individual metadata
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#defining-distributions-of-mutation-effect-sizes">
   Defining distributions of mutation effect sizes
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#region-weighting">
     Region weighting
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#setting-the-dominance-of-new-mutations">
     Setting the dominance of new mutations
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#built-in-distributions-of-effect-sizes">
     Built-in distributions of effect sizes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#labelling-mutations-from-different-regions">
     Labelling mutations from different regions
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modeling-recombination">
   Modeling recombination
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#method-1-regions-and-weights">
     Method 1: regions and weights
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#method-2-using-genetic-map-classes">
     Method 2: using “genetic map” classes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#general-comments">
     General comments
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modeling-mutations-having-direct-effects-on-fitness">
   Modeling mutations having direct effects on fitness
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modeling-mutations-affecting-phenotypes">
   Modeling mutations affecting phenotypes
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#changing-the-optimum-phenotype-during-a-simulation">
     Changing the optimum phenotype during a simulation
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#adding-demographic-events-involving-discrete-demes">
   Adding demographic events involving discrete demes
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-note-of-caution">
     A note of caution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#debugging-a-demographic-model">
     Debugging a demographic model
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setting-up-the-parameters-for-a-simulation">
   Setting up the parameters for a simulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mutation-and-recombination-rates">
     Mutation and recombination rates
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#running-a-simulation-with-tree-sequence-recording">
   Running a simulation with tree sequence recording
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#recording-ancient-samples-during-a-simulation">
   Recording “ancient samples” during a simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tracking-features-of-the-population-during-a-simulation">
   Tracking features of the population during a simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#starting-a-simulation-with-the-output-of-a-coalescent-simulation">
   Starting a simulation with the output of a coalescent simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#converting-the-data-to-a-tskit-treesequence">
   Converting the data to a
   <code class="xref py py-class docutils literal notranslate">
    <span class="pre">
     tskit.TreeSequence
    </span>
   </code>
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#saving-the-results-of-a-simulation-to-disk">
   Saving the results of a simulation to disk
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#outputting-a-tskit-trees-file">
     Outputting a
     <code class="docutils literal notranslate">
      <span class="pre">
       tskit
      </span>
     </code>
     “trees file”
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Tutorial</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#initializing-a-population">
   Initializing a population
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#looking-at-individual-metadata">
     Looking at individual metadata
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#defining-distributions-of-mutation-effect-sizes">
   Defining distributions of mutation effect sizes
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#region-weighting">
     Region weighting
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#setting-the-dominance-of-new-mutations">
     Setting the dominance of new mutations
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#built-in-distributions-of-effect-sizes">
     Built-in distributions of effect sizes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#labelling-mutations-from-different-regions">
     Labelling mutations from different regions
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modeling-recombination">
   Modeling recombination
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#method-1-regions-and-weights">
     Method 1: regions and weights
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#method-2-using-genetic-map-classes">
     Method 2: using “genetic map” classes
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#general-comments">
     General comments
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modeling-mutations-having-direct-effects-on-fitness">
   Modeling mutations having direct effects on fitness
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#modeling-mutations-affecting-phenotypes">
   Modeling mutations affecting phenotypes
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#changing-the-optimum-phenotype-during-a-simulation">
     Changing the optimum phenotype during a simulation
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#adding-demographic-events-involving-discrete-demes">
   Adding demographic events involving discrete demes
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-note-of-caution">
     A note of caution
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#debugging-a-demographic-model">
     Debugging a demographic model
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setting-up-the-parameters-for-a-simulation">
   Setting up the parameters for a simulation
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#mutation-and-recombination-rates">
     Mutation and recombination rates
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#running-a-simulation-with-tree-sequence-recording">
   Running a simulation with tree sequence recording
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#recording-ancient-samples-during-a-simulation">
   Recording “ancient samples” during a simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tracking-features-of-the-population-during-a-simulation">
   Tracking features of the population during a simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#starting-a-simulation-with-the-output-of-a-coalescent-simulation">
   Starting a simulation with the output of a coalescent simulation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#converting-the-data-to-a-tskit-treesequence">
   Converting the data to a
   <code class="xref py py-class docutils literal notranslate">
    <span class="pre">
     tskit.TreeSequence
    </span>
   </code>
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#saving-the-results-of-a-simulation-to-disk">
   Saving the results of a simulation to disk
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#outputting-a-tskit-trees-file">
     Outputting a
     <code class="docutils literal notranslate">
      <span class="pre">
       tskit
      </span>
     </code>
     “trees file”
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="tutorial">
<span id="id1"></span><h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">#</a></h1>
<p>The tutorial is organized around a series of tasks that one
needs to complete in order to run a simulation.</p>
<p>Each section covers a topic relevant to parameterizing a simulation.
The variable names that store the types are the same as <code class="docutils literal notranslate"><span class="pre">kwargs</span></code>
used to initialize the final model parameters class.</p>
<p>Along the way, you will see links to more detailed and/or more advanced
sections of the documentation.</p>
<p>For a first-time reader, it is probably best to go top-to-bottom through
this tutorial.</p>
<section id="initializing-a-population">
<h2>Initializing a population<a class="headerlink" href="#initializing-a-population" title="Permalink to this headline">#</a></h2>
<p>A simulation modifies the contents of instances of <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a>.
We have two basic ways to initialize a population:</p>
<ul class="simple">
<li><p>with a list of deme sizes and a genome length</p></li>
<li><p>from a tree sequence in <code class="docutils literal notranslate"><span class="pre">tskit</span></code> format</p></li>
</ul>
<p>This section deals with the former.  See <a class="reference internal" href="#pop-from-tskit"><span class="std std-ref">here</span></a>
for how to initialize from <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a> objects.</p>
<p>To initialize a population of <code class="docutils literal notranslate"><span class="pre">N</span></code> diploids from a single deme and a given
genome length:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy11</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">(</span><span class="n">N</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">length</span><span class="o">=</span><span class="mf">100.0</span><span class="p">)</span>
<span class="n">pop</span><span class="o">.</span><span class="n">N</span>
<span class="n">pop</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">genome_length</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100.0
</pre></div>
</div>
</div>
</div>
<p>The genome length is stored as an attribute of the <em>table collection</em>
that is used to record the ancestry of the simulated population.
The tables themselves are instances of <a class="reference internal" href="types.html#fwdpy11.TableCollection" title="fwdpy11.TableCollection"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.TableCollection</span></code></a>.
See <a class="reference internal" href="tsoverview.html#tsoverview"><span class="std std-ref">here</span></a> for a brief introduction to tables as well
as <span id="id2">[<a class="reference internal" href="../misc/bibliography.html#id19" title="Jerome Kelleher, Alison M Etheridge, and Gilean McVean. Efficient coalescent simulation and genealogical analysis for large sample sizes. PLoS Comput. Biol., 12(5):e1004842, May 2016. doi:10.1371/journal.pcbi.1004842.">KEM16</a>]</span> and <span id="id3">[<a class="reference internal" href="../misc/bibliography.html#id10" title="Jerome Kelleher, Kevin R Thornton, Jaime Ashander, and Peter L Ralph. Efficient pedigree recording for fast population genetics simulation. PLoS Comput. Biol., 14(11):e1006581, November 2018. doi:10.1371/journal.pcbi.1006581.">KTAR18</a>]</span>.  See <a class="reference internal" href="tstypes.html#ts-data-types"><span class="std std-ref">here</span></a>
for a more detailed discussion of relevant data structures.</p>
<p>To initialize a population with individuals in multiple demes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">(</span><span class="n">N</span><span class="o">=</span><span class="p">[</span><span class="mi">500</span><span class="p">,</span> <span class="mi">500</span><span class="p">],</span> <span class="n">length</span><span class="o">=</span><span class="mf">100.0</span><span class="p">)</span>
<span class="n">pop</span><span class="o">.</span><span class="n">N</span>
<span class="n">pop</span><span class="o">.</span><span class="n">tables</span><span class="o">.</span><span class="n">genome_length</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100.0
</pre></div>
</div>
</div>
</div>
<section id="looking-at-individual-metadata">
<h3>Looking at individual metadata<a class="headerlink" href="#looking-at-individual-metadata" title="Permalink to this headline">#</a></h3>
<p>Our population size is unchanged, but our population now has 500 individuals
in each of two demes.  We can verify this by looking at the <code class="docutils literal notranslate"><span class="pre">metadata</span></code> associated
with each individual.  Individual metadata are instances of
<a class="reference internal" href="types.html#fwdpy11.DiploidMetadata" title="fwdpy11.DiploidMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata</span></code></a> but they are also registered as a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.dtype.html#numpy.dtype" title="(in NumPy v1.24)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.dtype</span></code></a>,
meaning that we can look at the metadata as a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.recarray.html#numpy.recarray" title="(in NumPy v1.24)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.recarray</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">md</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">diploid_metadata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The field names of our array are the same as the attribute names of
<a class="reference internal" href="types.html#fwdpy11.DiploidMetadata" title="fwdpy11.DiploidMetadata"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidMetadata</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">md</span><span class="o">.</span><span class="n">dtype</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>dtype([(&#39;g&#39;, &#39;&lt;f8&#39;), (&#39;e&#39;, &#39;&lt;f8&#39;), (&#39;w&#39;, &#39;&lt;f8&#39;), (&#39;geography&#39;, &#39;&lt;f8&#39;, (3,)), (&#39;label&#39;, &#39;&lt;u8&#39;), (&#39;parents&#39;, &#39;&lt;u8&#39;, (2,)), (&#39;deme&#39;, &#39;&lt;i4&#39;), (&#39;sex&#39;, &#39;&lt;i4&#39;), (&#39;nodes&#39;, &#39;&lt;i4&#39;, (2,))])
</pre></div>
</div>
</div>
</div>
<p>We can easily confirm the number of individuals in each deme using
<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.unique.html#numpy.unique" title="(in NumPy v1.24)"><code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.unique()</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">md</span><span class="p">[</span><span class="s2">&quot;deme&quot;</span><span class="p">],</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(array([0, 1], dtype=int32), array([500, 500]))
</pre></div>
</div>
</div>
</div>
<p>We see that the deme labels are <code class="docutils literal notranslate"><span class="pre">0</span></code> and <code class="docutils literal notranslate"><span class="pre">1</span></code> and that each label
was found 500 times.  (The first 500 individuals are in deme <code class="docutils literal notranslate"><span class="pre">0</span></code>,
followed by 500 in deme <code class="docutils literal notranslate"><span class="pre">1</span></code>.)</p>
<p>See <a class="reference internal" href="../long_vignettes/tskit_metadata_vignette.html#tskit-metadata-vignette"><span class="std std-ref">here</span></a> for more on individual metadata.</p>
</section>
</section>
<section id="defining-distributions-of-mutation-effect-sizes">
<span id="mutationregions"></span><h2>Defining distributions of mutation effect sizes<a class="headerlink" href="#defining-distributions-of-mutation-effect-sizes" title="Permalink to this headline">#</a></h2>
<p>One of the main reasons to perform forward simulations is to be able
to model mutations affecting individual fitness. To do so, we need
to specify both mutation rates and the resulting effect sizes.</p>
<p><code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> works by specifying an overall mutation rate to variants
affecting fitness (see <a class="reference internal" href="#model-params"><span class="std std-ref">here</span></a>).  Given that
a mutation occurs, we need to specify its “effect size”.</p>
<p><code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> chooses the effect size of a new mutation by first
determining what <em>region</em> is mutated and then generating a mutation
from the distribution of effect size associated with that region.</p>
<p>Each region is represented by instances of classes derived from the
ABC <a class="reference internal" href="regiontypes.html#fwdpy11.Sregion" title="fwdpy11.Sregion"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Sregion</span></code></a>.  Each instance is associated with a <em>weight</em>.
These weights establish the relative probability that a mutation comes
from a given region.  Thus, given an overall mutation
rate to non-neutral variants, instances of “<code class="docutils literal notranslate"><span class="pre">sregions</span></code>” are used to set up
a multinomial distribution for generating new mutations.</p>
<p>The following sets up a model where mutations have a constant effect size
(<span class="math notranslate nohighlight">\(s=-0.01\)</span>), dominance <span class="math notranslate nohighlight">\(h=0.25\)</span>, and occur uniformly on
the interval <span class="math notranslate nohighlight">\([0, 1)\)</span>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">s</span><span class="o">=-</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">h</span><span class="o">=</span><span class="mf">0.25</span><span class="p">)]</span>
</pre></div>
</div>
</div>
</div>
<p>The previous example uses argument names for clarity, and the following is equivalent,
with the <code class="docutils literal notranslate"><span class="pre">int</span></code> values getting converted to <code class="docutils literal notranslate"><span class="pre">float</span></code> automatically:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">)]</span>
<span class="n">sregions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.ConstantS(beg=0, end=1, weight=1, s=-0.01, h=0.25, coupled=True, label=0, scaling=1.0)
</pre></div>
</div>
</div>
</div>
<p>Note that the constructor parameters for these classes often have default
values–see the specific class documentation for details.</p>
<p>In some scenarios, it is useful to think about the distribution of effect sizes
as scaled with respect to the population size.  For example, selection coefficients
may be exponentially-distributed with a mean of <span class="math notranslate nohighlight">\(2Ns\)</span>.  To do this in
<code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># ALPHA = 2Ns</span>
<span class="n">MEAN_ALPHA</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span>
<span class="n">N</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">MEAN_ALPHA</span><span class="p">,</span> <span class="n">scaling</span><span class="o">=</span><span class="mi">2</span> <span class="o">*</span> <span class="n">N</span><span class="p">)]</span>
<span class="n">sregions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.ExpS(beg=0, end=1, weight=1, mean=-10, h=1.0, coupled=True, label=0, scaling=2000)
</pre></div>
</div>
</div>
</div>
<section id="region-weighting">
<h3>Region weighting<a class="headerlink" href="#region-weighting" title="Permalink to this headline">#</a></h3>
<p>When multiple “sregion” objects are used, the default behavior is to multiply
the input <code class="docutils literal notranslate"><span class="pre">weight</span></code> by <code class="docutils literal notranslate"><span class="pre">end-beg</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span>
   <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mean</span><span class="o">=-</span><span class="mf">0.2</span><span class="p">),</span>
   <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">3.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">s</span><span class="o">=-</span><span class="mf">0.1</span><span class="p">),</span>
<span class="p">]</span>
<span class="n">sregions</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[fwdpy11.ExpS(beg=0.0, end=1.0, weight=1.0, mean=-0.2, h=1.0, coupled=True, label=0, scaling=1.0),
 fwdpy11.ConstantS(beg=1.0, end=3.0, weight=2.0, s=-0.1, h=1.0, coupled=True, label=0, scaling=1.0)]
</pre></div>
</div>
</div>
</div>
<p>Here, the input <code class="docutils literal notranslate"><span class="pre">weight</span></code> is interpreted to mean the weight “per site” is constant.
In this example, twice as many mutations will have positions in <span class="math notranslate nohighlight">\([1, 3)\)</span> as from <span class="math notranslate nohighlight">\([0, 1)\)</span>.</p>
<p>To change the default behavior, one can prevent the coupling between input <code class="docutils literal notranslate"><span class="pre">weight</span></code> and region length:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span>
   <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mean</span><span class="o">=-</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
   <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">3.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">s</span><span class="o">=-</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
<span class="p">]</span>
<span class="n">sregions</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[fwdpy11.ExpS(beg=0.0, end=1.0, weight=1.0, mean=-0.2, h=1.0, coupled=False, label=0, scaling=1.0),
 fwdpy11.ConstantS(beg=1.0, end=3.0, weight=1.0, s=-0.1, h=1.0, coupled=False, label=0, scaling=1.0)]
</pre></div>
</div>
</div>
</div>
<p>The absolute values of the <code class="docutils literal notranslate"><span class="pre">weight</span></code> parameters themselves is irrelevant.
The only thing that matters is the <em>relative</em> values from region to region.
Simulations based on the above examples would give the same results if the <code class="docutils literal notranslate"><span class="pre">weight</span></code> were 42 or 73.06.
Therefore, we can recreate our first example with code like the following:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span>
   <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">56.0</span><span class="p">,</span> <span class="n">mean</span><span class="o">=-</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
   <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">3.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">112.0</span><span class="p">,</span> <span class="n">s</span><span class="o">=-</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
<span class="p">]</span>
<span class="n">sregions</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[fwdpy11.ExpS(beg=0.0, end=1.0, weight=56.0, mean=-0.2, h=1.0, coupled=False, label=0, scaling=1.0),
 fwdpy11.ConstantS(beg=1.0, end=3.0, weight=112.0, s=-0.1, h=1.0, coupled=False, label=0, scaling=1.0)]
</pre></div>
</div>
</div>
</div>
<p>In the above example, twice as many mutations occur in the second region
because the weights have relative values of 2:1.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Different regions are allowed to overlap, allowing the simulation of
concepts like “coding regions” where the DFE are a weighted mixture
from multiple distributions, etc.</p>
</div>
</section>
<section id="setting-the-dominance-of-new-mutations">
<h3>Setting the dominance of new mutations<a class="headerlink" href="#setting-the-dominance-of-new-mutations" title="Permalink to this headline">#</a></h3>
<p>The dominance of a new mutation is set by the <code class="docutils literal notranslate"><span class="pre">h</span></code> parameter during
initialization:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mean</span><span class="o">=-</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">h</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mean</span><span class="o">=-</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">h</span><span class="o">=</span><span class="mf">0.0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.ExpS(beg=0.0, end=1.0, weight=1.0, mean=-0.2, h=0.0, coupled=True, label=0, scaling=1.0)
</pre></div>
</div>
</div>
</div>
</section>
<section id="built-in-distributions-of-effect-sizes">
<h3>Built-in distributions of effect sizes<a class="headerlink" href="#built-in-distributions-of-effect-sizes" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.ConstantS" title="fwdpy11.ConstantS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.ConstantS</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.UniformS" title="fwdpy11.UniformS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.UniformS</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.ExpS" title="fwdpy11.ExpS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.ExpS</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.GammaS" title="fwdpy11.GammaS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GammaS</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.GaussianS" title="fwdpy11.GaussianS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GaussianS</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.MultivariateGaussianEffects" title="fwdpy11.MultivariateGaussianEffects"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.MultivariateGaussianEffects</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.LogNormalS" title="fwdpy11.LogNormalS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.LogNormalS</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.DiscreteDESD" title="fwdpy11.DiscreteDESD"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiscreteDESD</span></code></a></p></li>
</ul>
</section>
<section id="labelling-mutations-from-different-regions">
<h3>Labelling mutations from different regions<a class="headerlink" href="#labelling-mutations-from-different-regions" title="Permalink to this headline">#</a></h3>
<p>It may be of use to know what region a mutation came from.  To do
so, give a nonzero value to the <code class="docutils literal notranslate"><span class="pre">label</span></code> argument:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.ConstantS(beg=0.0, end=1.0, weight=1.0, s=0.001, h=1.0, coupled=True, label=1, scaling=1.0)
</pre></div>
</div>
</div>
</div>
<p>At the end of the simulation, mutations from this region will have
the <code class="docutils literal notranslate"><span class="pre">label</span></code> value stored in the attribute <a class="reference internal" href="types.html#fwdpy11.Mutation.label" title="fwdpy11.Mutation.label"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.Mutation.label</span></code></a>.</p>
<p>The value of <code class="docutils literal notranslate"><span class="pre">label</span></code> must fit into a 16-bit unsigned integer,
<em>e.g.</em>, <a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.uint16" title="(in NumPy v1.24)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">numpy.uint16</span></code></a>. Larger values, or negative values, will result
in exceptions.  The following example tries to use a value one larger than
the maximum allowed:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span>
    <span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">iinfo</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">uint16</span><span class="p">)</span><span class="o">.</span><span class="n">max</span> <span class="o">+</span> <span class="mi">1</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">TypeError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">14</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span>     <span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">iinfo</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">uint16</span><span class="p">)</span><span class="o">.</span><span class="n">max</span> <span class="o">+</span> <span class="mi">1</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="p">)</span>

<span class="nn">File &lt;attrs generated init fwdpy11.regions.ConstantS&gt;:12,</span> in <span class="ni">__init__</span><span class="nt">(self, beg, end, weight, s, h, coupled, label, scaling)</span>
<span class="g g-Whitespace">     </span><span class="mi">10</span> <span class="n">_inst_dict</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">label</span>
<span class="g g-Whitespace">     </span><span class="mi">11</span> <span class="n">_inst_dict</span><span class="p">[</span><span class="s1">&#39;scaling&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scaling</span>
<span class="ne">---&gt; </span><span class="mi">12</span> <span class="bp">self</span><span class="o">.</span><span class="n">__attrs_post_init__</span><span class="p">()</span>

<span class="nn">File ~/work/fwdpy11/fwdpy11/fwdpy11/regions.py:137,</span> in <span class="ni">ConstantS.__attrs_post_init__</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">136</span> <span class="k">def</span> <span class="nf">__attrs_post_init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="ne">--&gt; </span><span class="mi">137</span>     <span class="nb">super</span><span class="p">(</span><span class="n">ConstantS</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">138</span>         <span class="bp">self</span><span class="o">.</span><span class="n">beg</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">139</span>         <span class="bp">self</span><span class="o">.</span><span class="n">end</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">140</span>         <span class="bp">self</span><span class="o">.</span><span class="n">weight</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">141</span>         <span class="bp">self</span><span class="o">.</span><span class="n">s</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">142</span>         <span class="bp">self</span><span class="o">.</span><span class="n">h</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">143</span>         <span class="bp">self</span><span class="o">.</span><span class="n">coupled</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">144</span>         <span class="bp">self</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">145</span>         <span class="bp">self</span><span class="o">.</span><span class="n">scaling</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">146</span>     <span class="p">)</span>

<span class="ne">TypeError</span>: __init__(): incompatible constructor arguments. The following argument types are supported:
<span class="g g-Whitespace">    </span><span class="mi">1</span><span class="o">.</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">_fwdpy11</span><span class="o">.</span><span class="n">_ll_ConstantS</span><span class="p">(</span><span class="n">beg</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">weight</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">s</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">h</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">coupled</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">label</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">scaling</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">2</span><span class="o">.</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">_fwdpy11</span><span class="o">.</span><span class="n">_ll_ConstantS</span><span class="p">(</span><span class="n">beg</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">end</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">weight</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">s</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">h</span><span class="p">:</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">_fwdpy11</span><span class="o">.</span><span class="n">MutationDominance</span><span class="p">,</span> <span class="n">coupled</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">label</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">scaling</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span>

<span class="n">Invoked</span> <span class="k">with</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.001</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="mi">65536</span><span class="p">,</span> <span class="mf">1.0</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="modeling-recombination">
<span id="geneticmaps"></span><h2>Modeling recombination<a class="headerlink" href="#modeling-recombination" title="Permalink to this headline">#</a></h2>
<p>Recombination rates are allowed to vary along genomes in a discrete fashion.  <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code>
provides two complementary methods for setting up such variation.</p>
<section id="method-1-regions-and-weights">
<span id="recregions"></span><h3>Method 1: regions and weights<a class="headerlink" href="#method-1-regions-and-weights" title="Permalink to this headline">#</a></h3>
<p>The method described in this section works in combination with a total overall recombination
rate.  This rate is the mean of a Poisson distribution and the intervals where recombination
breakpoints happen are chosen based on their relative weights.  The regions are instances
of <a class="reference internal" href="regiontypes.html#fwdpy11.Region" title="fwdpy11.Region"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Region</span></code></a>.  A region represents a continuous, half-open interval within which
crossover positions are uniformly distributed.</p>
<p>By way of example, say we want the following genetic map:</p>
<ul class="simple">
<li><p>The total recombination rate per diploid is <span class="math notranslate nohighlight">\(1e-3\)</span>, which is the mean of a Poisson process.</p></li>
<li><p>Our genome is continuous on <span class="math notranslate nohighlight">\([0,10)\)</span>.</p></li>
<li><p>The recombination rate is twice as high in one part of the “genome” than in the other.</p></li>
</ul>
<p>To initialize a region object, the following parameters may be used:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">beg</span></code>, the start of the region</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">end</span></code>, the end of the region</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">weight</span></code>, the weight assigned to the region</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">coupled</span></code> is a <code class="docutils literal notranslate"><span class="pre">bool</span></code> and determines how the weights are handled internally (see below).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">label</span></code> is an integer that defaults to <code class="docutils literal notranslate"><span class="pre">0</span></code> and is not relevant to recombination.</p></li>
</ul>
<p>The first three parameters are required.  A valid region has <span class="math notranslate nohighlight">\(beg \geq 0\)</span>,
<span class="math notranslate nohighlight">\(end &gt; beg\)</span> and <span class="math notranslate nohighlight">\(weight &gt;= 0\)</span> and defines a half-open interval <span class="math notranslate nohighlight">\([beg, end)\)</span>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A <code class="docutils literal notranslate"><span class="pre">weight</span></code> of <code class="docutils literal notranslate"><span class="pre">0</span></code> is the same as simply not defining a region! There is
no requirement that all genetic map elements cover the entire genome. We allow
zero-weight regions for those who think that it is cleaner/more explicit to write
them down.</p>
</div>
<p>By default, <code class="docutils literal notranslate"><span class="pre">coupled=True</span></code>, which means that the <em>total</em> weight assigned to a region
will be <span class="math notranslate nohighlight">\(weight\times (end-beg)\)</span>.  It is helpful to view <span class="math notranslate nohighlight">\(weight\)</span> as
the “rate per unit” and <span class="math notranslate nohighlight">\(end-beg\)</span> as the number of units in the region. (For example,
“unit” could refer to base pairs, but it need not.)</p>
<p>There are two ways to set this model up.  The first is arguably the most intuitive, which is to make
one region twice as long as the other:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy11</span>

<span class="n">recrate</span> <span class="o">=</span> <span class="mf">1e-3</span>
<span class="n">recregions</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">10.0</span> <span class="o">/</span> <span class="mf">3.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">),</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">10.0</span> <span class="o">/</span> <span class="mf">3.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">),</span>
<span class="p">]</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">recregions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.Region(beg=0.0, end=3.3333333333333335, weight=3.3333333333333335, coupled=True, label=0)
fwdpy11.Region(beg=3.3333333333333335, end=10, weight=6.666666666666666, coupled=True, label=0)
</pre></div>
</div>
</div>
</div>
<p>In the output, you see that <code class="docutils literal notranslate"><span class="pre">coupled=True</span></code>, which means that the simulation’s
back-end will assign twice as many crossovers to the second region as to the first.</p>
<p>In words, the <code class="docutils literal notranslate"><span class="pre">recregions</span></code> list and <code class="docutils literal notranslate"><span class="pre">recrate</span></code> value mean the following:</p>
<ul class="simple">
<li><p>The number of crossovers per diploid is Poisson distributed with mean
<code class="docutils literal notranslate"><span class="pre">recrate</span></code>, or <code class="docutils literal notranslate"><span class="pre">0.001</span></code>. See <a class="reference internal" href="#model-params-rate-details"><span class="std std-ref">here</span></a> for how to
send the <code class="docutils literal notranslate"><span class="pre">recrate</span></code> to a simulation.</p></li>
<li><p>Each crossover breakpoint has a <code class="docutils literal notranslate"><span class="pre">1/3</span></code> chance of being uniformly
distributed in <span class="math notranslate nohighlight">\([0, 10/3)\)</span> and a <code class="docutils literal notranslate"><span class="pre">2/3</span></code> chance of being
uniformly distributed in <span class="math notranslate nohighlight">\([10/3, 10)\)</span>.</p></li>
</ul>
<p>A more abstract approach relies on setting <code class="docutils literal notranslate"><span class="pre">coupled=False</span></code>, which means
that the “raw” weights that you input are the exact values used internally:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">recregions</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
<span class="p">]</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">recregions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.Region(beg=0, end=5, weight=1, coupled=False, label=0)
fwdpy11.Region(beg=5, end=10, weight=2, coupled=False, label=0)
</pre></div>
</div>
</div>
</div>
<p>Now, the <code class="docutils literal notranslate"><span class="pre">weight</span></code> arguments are treated as <em>absolute</em>, or exactly <code class="docutils literal notranslate"><span class="pre">1</span></code> and <code class="docutils literal notranslate"><span class="pre">2</span></code>, respectively.</p>
<p>In words, what we have is:</p>
<ul class="simple">
<li><p>The number of breakpoints per diploid is Poisson distributed with mean <span class="math notranslate nohighlight">\(1e-3\)</span></p></li>
<li><p>For each breakpoint, its position is uniform on <span class="math notranslate nohighlight">\([0, 5)\)</span> with probability <span class="math notranslate nohighlight">\(2/(2+1)\)</span>, or
it is uniform on <span class="math notranslate nohighlight">\([5, 10)\)</span> with probability <span class="math notranslate nohighlight">\(1/(2+1)\)</span>.</p></li>
</ul>
<p>In essence, instances of <a class="reference internal" href="regiontypes.html#fwdpy11.Region" title="fwdpy11.Region"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Region</span></code></a> parameterize a multinomial distribution that is used to
choose the ranges within which breakpoints are uniformly-distributed.  A limitation of this approach
is that we cannot model discrete jumps in genetic maps, such as those between chromosomes.</p>
</section>
<section id="method-2-using-genetic-map-classes">
<span id="geneticmapunit"></span><h3>Method 2: using “genetic map” classes<a class="headerlink" href="#method-2-using-genetic-map-classes" title="Permalink to this headline">#</a></h3>
<div class="versionadded">
<p><span class="versionmodified added">New in version 0.3.0.</span></p>
</div>
<p>An alternate approach uses instances of classes derived from the <code class="docutils literal notranslate"><span class="pre">ABC</span></code>
<code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GeneticMapUnit</span></code>. Here <code class="docutils literal notranslate"><span class="pre">Unit</span></code> refers to an <em>element</em> of
a genetic map rather than the actual units (<code class="docutils literal notranslate"><span class="pre">cM</span></code>, etc.).  Instances of
these classes contain their own rates and we can mix and match regions
where recombination breakpoints are Poisson and binomially distributed.</p>
<p>Let’s revisit the example from the previous section.  This time, we will
use <a class="reference internal" href="regiontypes.html#fwdpy11.PoissonInterval" title="fwdpy11.PoissonInterval"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.PoissonInterval</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">recregions</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PoissonInterval</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mf">2e-3</span> <span class="o">/</span> <span class="mi">3</span><span class="p">),</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PoissonInterval</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mf">1e-3</span> <span class="o">/</span> <span class="mi">3</span><span class="p">),</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>The number of breakpoints in each <span class="math notranslate nohighlight">\([beg, end)\)</span> interval is Poisson distributed
with the given mean. The position of each breakpoint is uniform on <span class="math notranslate nohighlight">\([beg, end)\)</span>.</p>
<p>These classes also allow us to specify breakpoints at a specific position with a specific probability.
The next example sets up 4 genomic regions, each 10 “units” long.  Within each region,
the mean number of breakpoints (per diploid, per generation) is <span class="math notranslate nohighlight">\(1e-3\)</span>.
Between each region, a single recombination occurs with probability of
one-half, meaning that each region is assorting independently (50 <code class="docutils literal notranslate"><span class="pre">cM</span></code> between each region).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">NLOCI</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">LOCUS_LENGTH</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">RECRATE_PER_LOCUS</span> <span class="o">=</span> <span class="mf">1e-3</span>
<span class="n">LOCUS_BOUNDARIES</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="n">LOCUS_LENGTH</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">NLOCI</span> <span class="o">*</span> <span class="n">LOCUS_LENGTH</span><span class="p">,</span> <span class="n">LOCUS_LENGTH</span><span class="p">)</span>
<span class="p">]</span>
<span class="n">recregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">PoissonInterval</span><span class="p">(</span><span class="o">*</span><span class="n">i</span><span class="p">,</span> <span class="n">RECRATE_PER_LOCUS</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">LOCUS_BOUNDARIES</span><span class="p">]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">LOCUS_BOUNDARIES</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
    <span class="n">recregions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">BinomialPoint</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mf">0.5</span><span class="p">))</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">recregions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.PoissonInterval(beg=0, end=10, mean=0.001, discrete=False)
fwdpy11.PoissonInterval(beg=10, end=20, mean=0.001, discrete=False)
fwdpy11.PoissonInterval(beg=20, end=30, mean=0.001, discrete=False)
fwdpy11.PoissonInterval(beg=30, end=40, mean=0.001, discrete=False)
fwdpy11.BinomialPoint(position=10, probability=0.5, discrete=False)
fwdpy11.BinomialPoint(position=20, probability=0.5, discrete=False)
fwdpy11.BinomialPoint(position=30, probability=0.5, discrete=False)
</pre></div>
</div>
</div>
</div>
<p>As an aside, this example is not creating objects in order by their positions.  Such ordering is not required.</p>
<p>Beginning in version <code class="docutils literal notranslate"><span class="pre">0.12.0</span></code>, it is possible to restrict crossover positions to integer values.
For the examples given above, crossover positions are floating-point values sampled uniformly from <span class="math notranslate nohighlight">\([beg, end)\)</span>.
To restrict positions to integer values, we pass <code class="docutils literal notranslate"><span class="pre">discrete=True</span></code> when creating object instances:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">recregions</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PoissonInterval</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mf">2e-3</span> <span class="o">/</span> <span class="mi">3</span><span class="p">,</span> <span class="n">discrete</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">PoissonInterval</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mf">1e-3</span> <span class="o">/</span> <span class="mi">3</span><span class="p">,</span> <span class="n">discrete</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Now, breakpoints from the first region will only take on values of <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">3</span></code>, or <code class="docutils literal notranslate"><span class="pre">4</span></code>.</p>
<p>Setting <code class="docutils literal notranslate"><span class="pre">discrete=True</span></code> requires the following:</p>
<ul class="simple">
<li><p>Values for <code class="docutils literal notranslate"><span class="pre">beg</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> must be <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>.  Thus, <code class="docutils literal notranslate"><span class="pre">1</span></code> is valid but <code class="docutils literal notranslate"><span class="pre">1.0</span></code> will raise a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">end</span> <span class="pre">-</span> <span class="pre">beg</span></code> must be <code class="docutils literal notranslate"><span class="pre">&gt;</span> <span class="pre">1</span></code>.  This requirement prevents you from using <code class="docutils literal notranslate"><span class="pre">beg=0</span></code> and <code class="docutils literal notranslate"><span class="pre">end=1</span></code>, for example,
which would result in the only possible crossover position being <code class="docutils literal notranslate"><span class="pre">0</span></code>.</p></li>
<li><p>You must be more careful when using <code class="docutils literal notranslate"><span class="pre">msprime</span></code> to start/finish a simulation.
See <a class="reference internal" href="../short_vignettes/initpops_vignette.html#precapitation"><span class="std std-ref">here</span></a> and <a class="reference internal" href="../short_vignettes/recapitation.html#recapitation"><span class="std std-ref">here</span></a> for details.</p></li>
</ul>
<p>The following classes are available:</p>
<ul class="simple">
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.PoissonInterval" title="fwdpy11.PoissonInterval"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.PoissonInterval</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.PoissonPoint" title="fwdpy11.PoissonPoint"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.PoissonPoint</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.BinomialInterval" title="fwdpy11.BinomialInterval"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.BinomialInterval</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.BinomialPoint" title="fwdpy11.BinomialPoint"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.BinomialPoint</span></code></a></p></li>
<li><p><a class="reference internal" href="regiontypes.html#fwdpy11.FixedCrossovers" title="fwdpy11.FixedCrossovers"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.FixedCrossovers</span></code></a></p></li>
</ul>
</section>
<section id="general-comments">
<h3>General comments<a class="headerlink" href="#general-comments" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Different <span class="math notranslate nohighlight">\([beg, end)\)</span> intervals may overlap.  The interpretation of such a setup is your problem.</p></li>
<li><p>The first method, based on <a class="reference internal" href="regiontypes.html#fwdpy11.Region" title="fwdpy11.Region"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Region</span></code></a> is slightly faster, but less flexible.  More on the flexibility
below.</p></li>
<li><p>When using classes like <a class="reference internal" href="regiontypes.html#fwdpy11.PoissonInterval" title="fwdpy11.PoissonInterval"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.PoissonInterval</span></code></a>, the recombination rate that you use to construct a
<a class="reference internal" href="model_params.html#fwdpy11.ModelParams" title="fwdpy11.ModelParams"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.ModelParams</span></code></a> instance is ignored, as the rates are stored in the individual objects.</p></li>
<li><p>You do not need to specify regions with zero recombination. Their existence is implied given the total
length of the genome being simulated (<a class="reference internal" href="types.html#fwdpy11.TableCollection.genome_length" title="fwdpy11.TableCollection.genome_length"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.TableCollection.genome_length</span></code></a>).</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Adding neutral mutations to the tables with <a class="reference internal" href="functions.html#fwdpy11.infinite_sites" title="fwdpy11.infinite_sites"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.infinite_sites()</span></code></a> will place
neutral variants in the non-recombining regions.</p>
</div>
</section>
</section>
<section id="modeling-mutations-having-direct-effects-on-fitness">
<span id="genetic-values"></span><h2>Modeling mutations having direct effects on fitness<a class="headerlink" href="#modeling-mutations-having-direct-effects-on-fitness" title="Permalink to this headline">#</a></h2>
<p>In a typical population-genetic model, mutations have direct effects on fitness.
Often, this effect is referred to as <code class="docutils literal notranslate"><span class="pre">s</span></code>, or the “selection coefficient”.</p>
<p>Once we’ve decided on our distributions of effect sizes, we need a way to obtain
a diploid’d fitness.  For these “standard” population genetic models, we will use
<a class="reference internal" href="genetic_values.html#fwdpy11.Multiplicative" title="fwdpy11.Multiplicative"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Multiplicative</span></code></a>.  Instances of this class tell the simulation
to calculate the genetic value of an individual using a multiplicative model where the
value contributed by each position with a mutation is:</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Genotype</p></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">AA</span></code></p></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">Aa</span></code></p></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">aa</span></code></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Fitness</p></td>
<td><p><span class="math notranslate nohighlight">\(1\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(1+hs\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(1 + scaling\times s\)</span></p></td>
</tr>
</tbody>
</table>
<p>In this table:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">A</span></code> refers to the ancestral/non-mutant allelic state</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">a</span></code> is the mutant allelic state</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">h</span></code> is the heterozygous effect of the mutant, the so-called dominance coefficient.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">s</span></code> is the selection coefficient.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">scaling</span></code> lets you decide between Fisher, Wright, Haldane, Kimura, etc.,
when determining the fitness of the mutant homozygote.</p></li>
</ul>
<p>The most common values for <code class="docutils literal notranslate"><span class="pre">scaling</span></code> are <code class="docutils literal notranslate"><span class="pre">1.0</span></code> or <code class="docutils literal notranslate"><span class="pre">2.0</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gvalue</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Multiplicative</span><span class="p">(</span><span class="n">scaling</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="n">gvalue</span><span class="o">.</span><span class="n">scaling</span>
<span class="n">gvalue</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Multiplicative</span><span class="p">(</span><span class="n">scaling</span><span class="o">=</span><span class="mf">2.0</span><span class="p">)</span>
<span class="n">gvalue</span><span class="o">.</span><span class="n">scaling</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2.0
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">scaling</span></code> parameter interacts with the <code class="docutils literal notranslate"><span class="pre">h</span></code> parameter
for a distribution of effect sizes! (See <a class="reference internal" href="#mutationregions"><span class="std std-ref">Defining distributions of mutation effect sizes</span></a>.)
For example, if <code class="docutils literal notranslate"><span class="pre">scaling</span> <span class="pre">=</span> <span class="pre">1.0</span></code>, then <code class="docutils literal notranslate"><span class="pre">h</span> <span class="pre">=</span> <span class="pre">1.0</span></code> results
in dominant mutations.  However, if <code class="docutils literal notranslate"><span class="pre">scaling</span> <span class="pre">=</span> <span class="pre">2.0</span></code>, then
<code class="docutils literal notranslate"><span class="pre">h</span> <span class="pre">=</span> <span class="pre">1.0</span></code> gives co-dominant mutations.  In both cases,
<code class="docutils literal notranslate"><span class="pre">h</span> <span class="pre">=</span> <span class="pre">0.0</span></code> generates fully-recessive mutations.</p>
</div>
</section>
<section id="modeling-mutations-affecting-phenotypes">
<span id="modeling-quant-traits"></span><h2>Modeling mutations affecting phenotypes<a class="headerlink" href="#modeling-mutations-affecting-phenotypes" title="Permalink to this headline">#</a></h2>
<p>The previous section discussed setting up a model where a mutation’s
effect size (<a class="reference internal" href="types.html#fwdpy11.Mutation.s" title="fwdpy11.Mutation.s"><code class="xref py py-attr docutils literal notranslate"><span class="pre">fwdpy11.Mutation.s</span></code></a>) directly affects individual fitness.
An alternative model is one where mutations affect some abstract “trait”
or “phenotype” and a separate function maps trait values to fitness.</p>
<p>Let’s consider the standard model of evolutionary quantitative genetics:</p>
<ul class="simple">
<li><p>Mutations have <strong>additive</strong> effects on trait values</p></li>
<li><p>The fitness of a trait value is a quadratic function of its distance
from an “optimum” trait value.</p></li>
</ul>
<p>In <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code>, a non-mutant individual has a phenotype of <code class="docutils literal notranslate"><span class="pre">0.0</span></code>.  Trait
values are additive over the values contributed by individual genotypes
according to the following table:</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Genotype</p></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">AA</span></code></p></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">Aa</span></code></p></th>
<th class="head"><p><code class="docutils literal notranslate"><span class="pre">aa</span></code></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Trait value</p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(hs\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(scaling\times s\)</span></p></td>
</tr>
</tbody>
</table>
<p>(If we model multiplicative effects on a trait, a non-mutant individual
still has a value of <code class="docutils literal notranslate"><span class="pre">0.0</span></code>. The internal machinery handles this so
that you don’t have to worry about it.)</p>
<p>To specify an additive effects model of a trait under Gaussian
stabilizing selection with an optimum trait value of <code class="docutils literal notranslate"><span class="pre">0.0</span></code> and
(inverse) strength of stabilizing selection <code class="docutils literal notranslate"><span class="pre">VS</span> <span class="pre">=</span> <span class="pre">1.0</span></code>, we write:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gvalue</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Additive</span><span class="p">(</span>
    <span class="n">scaling</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">gvalue_to_fitness</span><span class="o">=</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSS</span><span class="p">(</span><span class="n">optimum</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Here, we are using a second parameter to initialize a “genetic value to fitness”
map stored in an instance of <a class="reference internal" href="genetic_values.html#fwdpy11.Additive" title="fwdpy11.Additive"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Additive</span></code></a>.
(<a class="reference internal" href="genetic_values.html#fwdpy11.Multiplicative" title="fwdpy11.Multiplicative"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Multiplicative</span></code></a> also supports such maps.)
See <a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GSS" title="fwdpy11.GSS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GSS</span></code></a> for details.</p>
<p>We can also add Gaussian noise to an individual’s trait value:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gvalue</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Additive</span><span class="p">(</span>
    <span class="n">scaling</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
    <span class="n">gvalue_to_fitness</span><span class="o">=</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSS</span><span class="p">(</span><span class="n">optimum</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">2.0</span> <span class="o">/</span> <span class="mf">3.0</span><span class="p">),</span>
    <span class="n">noise</span><span class="o">=</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GaussianNoise</span><span class="p">(</span><span class="n">mean</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">sd</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="mf">3.0</span><span class="p">)),</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The last example requires some explanation:</p>
<ul class="simple">
<li><p>We want <code class="docutils literal notranslate"><span class="pre">VS</span> <span class="pre">=</span> <span class="pre">1.0</span></code>.  We can decompose <code class="docutils literal notranslate"><span class="pre">VS</span> <span class="pre">=</span> <span class="pre">VW</span> <span class="pre">+</span> <span class="pre">VE</span></code>, where <code class="docutils literal notranslate"><span class="pre">VW</span></code> and
<code class="docutils literal notranslate"><span class="pre">VE</span></code> are the additive contributions of genetic and environmental effects.</p></li>
<li><p>Here, the environmental effect is a Gaussian with mean zero and variance
<code class="docutils literal notranslate"><span class="pre">1/3</span></code>.  The class is parameterized with the standard deviation, however,
so we need to pass on the square root.</p></li>
<li><p>We then set <code class="docutils literal notranslate"><span class="pre">VS</span> <span class="pre">=</span> <span class="pre">1</span> <span class="pre">-</span> <span class="pre">1/3</span> <span class="pre">=</span> <span class="pre">2/3</span></code> when initializing <a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GSS" title="fwdpy11.GSS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GSS</span></code></a>.</p></li>
</ul>
<p>Yes, this is a nomenclature issue!  The <code class="docutils literal notranslate"><span class="pre">VS</span></code> argument to <a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GSS" title="fwdpy11.GSS"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GSS</span></code></a>
really should be called <code class="docutils literal notranslate"><span class="pre">VW</span></code> and we’ll fix that in a future version and hopefully
not break people’s code.</p>
<p>In general, there’s a good bit of subtlety to properly modeling quantitative traits.
The machinery described here was used in <span id="id4">[<a class="reference internal" href="../misc/bibliography.html#id9" title="Kevin R Thornton. Polygenic adaptation to an environmental shift: temporal dynamics of variation under gaussian stabilizing selection and additive effects on a single trait. Genetics, 213(4):1513–1530, December 2019. doi:10.1534/genetics.119.302662.">Tho19</a>]</span>. <span id="id5">[<a class="reference internal" href="../misc/bibliography.html#id22" title="R Bürger. The Mathematical Theory of Selection, Recombination, and Mutation. Wiley, November 2000. ISBN 9780471986539.">Burger00</a>]</span> is an excellent
technical reference on the topic. <span id="id6">[<a class="reference internal" href="../misc/bibliography.html#id3" title="Bruce Walsh and Michael Lynch. Evolution and Selection of Quantitative Traits. Oxford University Press, August 2018. ISBN 9780198830870.">WL18</a>]</span> also thoroughly covers a lot of
relevant material.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Under the hood, the <code class="docutils literal notranslate"><span class="pre">GSS</span></code> and <code class="docutils literal notranslate"><span class="pre">GSSmo</span></code> classes aren’t that different.
Their multivariate analogs are rather similar, too.  Thus, we envision
a future with one single <code class="docutils literal notranslate"><span class="pre">fwdpy11.GaussianStabilizingSelection</span></code> class
to handle all cases.  The types discussed here would remain as simple
Python wrappers so that we don’t break existing simulations.</p>
</div>
<p>For an example of another approach to modeling phenotypes often
attributed to <span id="id7">[<a class="reference internal" href="../misc/bibliography.html#id23" title="Adam Eyre-Walker. Evolution in health and medicine sackler colloquium: genetic architecture of a complex trait and its implications for fitness and genome-wide association studies. Proc. Natl. Acad. Sci. U. S. A., 107 Suppl 1:1752–1756, January 2010. doi:10.1073/pnas.0906182107.">EW10</a>]</span>, see <a class="reference internal" href="advancedtopics.html#eyre-walker"><span class="std std-ref">here</span></a>.</p>
<div class="admonition-todo admonition" id="id8">
<p class="admonition-title">Todo</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Write (and refer to) an advanced section on pleiotropic models.
</pre></div>
</div>
</div>
<section id="changing-the-optimum-phenotype-during-a-simulation">
<h3>Changing the optimum phenotype during a simulation<a class="headerlink" href="#changing-the-optimum-phenotype-during-a-simulation" title="Permalink to this headline">#</a></h3>
<p>The previous example set up a model where the optimum is stable for
the entire simulation.  We can parameterize a shifting optimum
using <a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.GSSmo" title="fwdpy11.GSSmo"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GSSmo</span></code></a>.  For example, to shift the optimum from <code class="docutils literal notranslate"><span class="pre">0.0</span></code>
to <code class="docutils literal notranslate"><span class="pre">1.0</span></code> at generation <code class="docutils literal notranslate"><span class="pre">100</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">moving_optimum</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSSmo</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Optimum</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">optimum</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">1.0</span><span class="p">),</span>
        <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Optimum</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">optimum</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">1.0</span><span class="p">),</span>
    <span class="p">]</span>
<span class="p">)</span>

<span class="n">gvalue</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Additive</span><span class="p">(</span><span class="n">scaling</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">gvalue_to_fitness</span><span class="o">=</span><span class="n">moving_optimum</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Since we are working in <code class="docutils literal notranslate"><span class="pre">Python</span></code>, we can take advantage of existing libraries to
implement interesting models.  Let’s consider the following model of a randomly
moving optimum:</p>
<ul class="simple">
<li><p>There is a 1% chance each generation that the optimum shifts.</p></li>
<li><p>When a shift happens, a normal deviate with mean <code class="docutils literal notranslate"><span class="pre">0.0</span></code> and variance
<code class="docutils literal notranslate"><span class="pre">0.1</span></code> is added to the current optimum.</p></li>
<li><p>The simulation will end at generation <code class="docutils literal notranslate"><span class="pre">1,000</span></code>.</p></li>
</ul>
<p>Let’s code it up:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">optima</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">Optimum</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">optimum</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">10.0</span><span class="p">)]</span>

<span class="n">last_time</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">last_optimum</span> <span class="o">=</span> <span class="mf">0.0</span>

<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">666</span><span class="p">)</span>

<span class="k">while</span> <span class="n">last_time</span> <span class="o">&lt;</span> <span class="mi">1000</span><span class="p">:</span>
    <span class="n">last_time</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">geometric</span><span class="p">(</span><span class="mf">0.01</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">last_optimum</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mf">0.1</span><span class="p">),</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">last_time</span> <span class="o">&lt;</span> <span class="mi">1000</span><span class="p">:</span>
        <span class="n">optima</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">Optimum</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="n">last_time</span><span class="p">,</span> <span class="n">optimum</span><span class="o">=</span><span class="n">last_optimum</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">10.0</span><span class="p">))</span>

<span class="n">random_moving_optimum</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSSmo</span><span class="p">(</span><span class="n">optima</span><span class="p">)</span>
<span class="n">random_moving_optimum</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.GSSmo(optima=[fwdpy11.Optimum(optimum=0.0, VS=10.0, when=0), fwdpy11.Optimum(optimum=0.14621925773046504, VS=10.0, when=120), fwdpy11.Optimum(optimum=0.43133329773747653, VS=10.0, when=250), fwdpy11.Optimum(optimum=0.42015093754764493, VS=10.0, when=552), fwdpy11.Optimum(optimum=0.3570913002785381, VS=10.0, when=557), fwdpy11.Optimum(optimum=-0.0949791266685498, VS=10.0, when=568), fwdpy11.Optimum(optimum=-0.08281396117205914, VS=10.0, when=704), fwdpy11.Optimum(optimum=-0.4194925319000761, VS=10.0, when=726), fwdpy11.Optimum(optimum=-0.09246489411724829, VS=10.0, when=875), fwdpy11.Optimum(optimum=-0.37135303951111776, VS=10.0, when=876), fwdpy11.Optimum(optimum=0.024965855928271607, VS=10.0, when=898)])
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note the cast to <code class="docutils literal notranslate"><span class="pre">int</span></code> when updating the time.  <a class="reference internal" href="gvalue_to_fitness.html#fwdpy11.Optimum" title="fwdpy11.Optimum"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Optimum</span></code></a>
is very picky about its input. It requires <code class="docutils literal notranslate"><span class="pre">int</span></code> for <code class="docutils literal notranslate"><span class="pre">when</span></code> and will
raise an exception if the <a class="reference external" href="https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.int64" title="(in NumPy v1.24)"><code class="xref py py-attr docutils literal notranslate"><span class="pre">numpy.int64</span></code></a> from <a class="reference external" href="https://numpy.org/doc/stable/reference/random/generated/numpy.random.geometric.html#numpy.random.geometric" title="(in NumPy v1.24)"><code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.random.geometric()</span></code></a>
gets passed in.</p>
</div>
</section>
</section>
<section id="adding-demographic-events-involving-discrete-demes">
<h2>Adding demographic events involving discrete demes<a class="headerlink" href="#adding-demographic-events-involving-discrete-demes" title="Permalink to this headline">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> has a flexible interface for demographic models involving multiple
discrete demes.  A full overview of the <code class="docutils literal notranslate"><span class="pre">API</span></code> is given <a class="reference internal" href="softselection.html#softselection"><span class="std std-ref">here</span></a>.
This section gives a cursory introduction.</p>
<p>Consider the following verbal description of a model:</p>
<ul class="simple">
<li><p>There is a single ancestral population of <code class="docutils literal notranslate"><span class="pre">N</span> <span class="pre">=</span> <span class="pre">100</span></code> diploids.</p></li>
<li><p>100 generations into this population’s future, it splits into two equal-sized
demes</p></li>
<li><p>The migration rate between each deme is <code class="docutils literal notranslate"><span class="pre">1e-3</span></code>.</p></li>
</ul>
<p>To set this model up, first initialize the population to the ancestral state:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This model has migration, so we need a migration matrix. The rows of a migration
matrix are the <strong>destination</strong> demes and the columns are the <strong>source</strong> demes.
A migration matrix can be interpreted as the fraction of migrants each generation
from each source deme. This definition implies that each row must sum to <code class="docutils literal notranslate"><span class="pre">1.0</span></code>.</p>
<p>This is a 2-deme model, so we need a <code class="docutils literal notranslate"><span class="pre">2x2</span></code> matrix.  Initially, there is only the
single ancestral deme, and therefore 100% of its ancestry each generation is
from itself. Thus, our initial migration matrix looks like:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">migmatrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">migmatrix</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="n">migmatrix</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[1., 0.],
       [0., 0.]])
</pre></div>
</div>
</div>
</div>
<p>At generation <code class="docutils literal notranslate"><span class="pre">100</span></code>, we move half of the ancestral population (deme <code class="docutils literal notranslate"><span class="pre">0</span></code>) to a new
deme <code class="docutils literal notranslate"><span class="pre">1</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mass_migrations</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">move_individuals</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">destination</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">fraction</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>We now need to set up our new symmetric migration rates:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">m</span> <span class="o">=</span> <span class="mf">1e-3</span>
<span class="n">set_migration_rates</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">SetMigrationRates</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">deme</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">migrates</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span> <span class="o">-</span> <span class="n">m</span><span class="p">,</span> <span class="n">m</span><span class="p">]),</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">SetMigrationRates</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">deme</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">migrates</span><span class="o">=</span><span class="p">[</span><span class="n">m</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">m</span><span class="p">]),</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>We now have the parts of our model.  To build a model, we need to create
an instance of <a class="reference internal" href="types.html#fwdpy11.DiscreteDemography" title="fwdpy11.DiscreteDemography"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiscreteDemography</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dmodel</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiscreteDemography</span><span class="p">(</span>
    <span class="n">mass_migrations</span><span class="o">=</span><span class="n">mass_migrations</span><span class="p">,</span>
    <span class="n">migmatrix</span><span class="o">=</span><span class="n">migmatrix</span><span class="p">,</span>
    <span class="n">set_migration_rates</span><span class="o">=</span><span class="n">set_migration_rates</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>This class has a nice method for pretty-printing via <code class="docutils literal notranslate"><span class="pre">black</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">dmodel</span><span class="o">.</span><span class="n">asblack</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.DiscreteDemography(
    mass_migrations=[
        fwdpy11.MassMigration(
            when=100,
            source=0,
            destination=1,
            fraction=0.5,
            move_individuals=True,
            resets_growth_rate=True,
        )
    ],
    set_growth_rates=None,
    set_deme_sizes=None,
    set_selfing_rates=None,
    migmatrix=fwdpy11.MigrationMatrix(
        migmatrix=array([[1.0, 0.0], [0.0, 0.0]]), scaled=False
    ),
    set_migration_rates=[
        fwdpy11.SetMigrationRates(when=100, deme=0, migrates=[0.999, 0.001]),
        fwdpy11.SetMigrationRates(when=100, deme=1, migrates=[0.001, 0.999]),
    ],
)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<p>Instances of <a class="reference internal" href="types.html#fwdpy11.DiscreteDemography" title="fwdpy11.DiscreteDemography"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiscreteDemography</span></code></a> are immutable, meaning that
the attributes are read-only and attempts to modify will raise exceptions:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dmodel</span><span class="o">.</span><span class="n">mass_migrations</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">FrozenInstanceError</span><span class="g g-Whitespace">                       </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">31</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">dmodel</span><span class="o">.</span><span class="n">mass_migrations</span> <span class="o">=</span> <span class="kc">None</span>

<span class="nn">File ~/.local/lib/python3.10/site-packages/attr/_make.py:642,</span> in <span class="ni">_frozen_setattrs</span><span class="nt">(self, name, value)</span>
<span class="g g-Whitespace">    </span><span class="mi">638</span> <span class="k">def</span> <span class="nf">_frozen_setattrs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">639</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">640</span><span class="sd">     Attached to frozen classes as __setattr__.</span>
<span class="g g-Whitespace">    </span><span class="mi">641</span><span class="sd">     &quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">642</span>     <span class="k">raise</span> <span class="n">FrozenInstanceError</span><span class="p">()</span>

<span class="ne">FrozenInstanceError</span>: 
</pre></div>
</div>
</div>
</div>
<p>However, you can get a <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a> representation of the data, which you
can modify:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">copy</span>

<span class="n">dmodel_dict</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">dmodel</span><span class="o">.</span><span class="n">asdict</span><span class="p">())</span>
<span class="n">dmodel_dict</span><span class="p">[</span><span class="s2">&quot;mass_migrations&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">dmodel2</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiscreteDemography</span><span class="p">(</span><span class="o">**</span><span class="n">dmodel_dict</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dmodel2</span><span class="o">.</span><span class="n">asblack</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.DiscreteDemography(
    mass_migrations=None,
    set_growth_rates=None,
    set_deme_sizes=None,
    set_selfing_rates=None,
    migmatrix=fwdpy11.MigrationMatrix(
        migmatrix=array([[1.0, 0.0], [0.0, 0.0]]), scaled=False
    ),
    set_migration_rates=[
        fwdpy11.SetMigrationRates(when=100, deme=0, migrates=[0.999, 0.001]),
        fwdpy11.SetMigrationRates(when=100, deme=1, migrates=[0.001, 0.999]),
    ],
)
</pre></div>
</div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It is best practice to use <a class="reference external" href="https://docs.python.org/3/library/copy.html#copy.deepcopy" title="(in Python v3.11)"><code class="xref py py-func docutils literal notranslate"><span class="pre">copy.deepcopy()</span></code></a> here. The
<a class="reference internal" href="types.html#fwdpy11.DiscreteDemography" title="fwdpy11.DiscreteDemography"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiscreteDemography</span></code></a> instances may contain
objects like <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.24)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a> that only get copied by
reference when the <code class="docutils literal notranslate"><span class="pre">dict</span></code> is generated.  For safety/general
happiness, making a deep copy helps here.</p>
</div>
<p><code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> contains a small collection of pre-computed demographic models.
One of them is the common modeling scenario of two recently diverged populations.
Additionally, we supply some commonly-used models of human demography.
See <a class="reference internal" href="demographic_models.html#demographic-models"><span class="std std-ref">The demography debugger</span></a> for details.</p>
<section id="a-note-of-caution">
<h3>A note of caution<a class="headerlink" href="#a-note-of-caution" title="Permalink to this headline">#</a></h3>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Programatically building up demographic models is tedious
and error-prone.  Errors end up in the literature, too (see
<span id="id9">[<a class="reference internal" href="../misc/bibliography.html#id2" title="Aaron P Ragsdale, Dominic Nelson, Simon Gravel, and Jerome Kelleher. Lessons learned from bugs in models of human history. Am. J. Hum. Genet., 107(4):583–588, October 2020. doi:10.1016/j.ajhg.2020.08.017.">RNGK20</a>]</span>).  Where possible, I <strong>strongly</strong> urge you
to test the correctness of a model against another implementation.
For example, one could compare results with no selection
and no recombination with the output of <code class="docutils literal notranslate"><span class="pre">msprime</span></code>. Another
approach is to compare to the predictions of
<a class="reference external" href="https://bitbucket.org/simongravel/moments">moments</a> <span id="id10">[<a class="reference internal" href="../misc/bibliography.html#id8" title="Julien Jouganous, Will Long, Aaron P Ragsdale, and Simon Gravel. Inferring the joint demographic history of multiple populations: beyond the diffusion approximation. Genetics, 206(3):1549–1567, July 2017. doi:10.1534/genetics.117.200493.">JLRG17</a>]</span>,
which is the approach taken by the <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code>
<a class="reference external" href="https://github.com/molpopgen/fwdpy11_statistical_tests">statistical tests</a>.</p>
</div>
<p>Given the inherent difficulty of building up models, we hope to provide
a simpler approach in a future release.  There’s a small group of people
currently hatching a plan to provide a common schema representing demographic
models of discrete demes.  The hope is that several pieces of software can
all use these schema.</p>
</section>
<section id="debugging-a-demographic-model">
<span id="demographydebugger"></span><h3>Debugging a demographic model<a class="headerlink" href="#debugging-a-demographic-model" title="Permalink to this headline">#</a></h3>
<p>The parameters of a demographic model are checked at run time at two different places:</p>
<ul class="simple">
<li><p>Upon object construction.  The various event objects try to make sure that
the parameter inputs are valid.</p></li>
<li><p>If invalid events occur during a simulation,
the simulation raises a <code class="docutils literal notranslate"><span class="pre">fwdpy11.DemographyError</span></code> exception.</p></li>
</ul>
<p>It is clearly preferable for a simulation to detect errors as early as possible.
While bad inputs can be detected almost immediately, more subtle errors are only
detected during simulation, which may take a while.
A more efficient approach to checking your models is to use <a class="reference internal" href="demographic_models.html#fwdpy11.DemographyDebugger" title="fwdpy11.DemographyDebugger"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DemographyDebugger</span></code></a>:</p>
<p>The class also generates a “report” with a verbal description of the model:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">d</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DemographyDebugger</span><span class="p">(</span>
    <span class="n">initial_deme_sizes</span><span class="o">=</span><span class="p">[</span><span class="mi">100</span><span class="p">],</span>
    <span class="n">events</span><span class="o">=</span><span class="n">dmodel</span><span class="p">,</span>
    <span class="n">simlen</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span>
    <span class="n">deme_labels</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s2">&quot;ANCESTRAL&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;DERIVED&quot;</span><span class="p">},</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">report</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DemographyDebugger::report not yet implemented
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/fwdpy11/fwdpy11/fwdpy11/_types/demography_debugger.py:307: UserWarning: report generation has not been implemented
  warnings.warn(&quot;report generation has not been implemented&quot;)
</pre></div>
</div>
</div>
</div>
<p>If the model were invalid, then an error would have been raised during initialization.</p>
<p>Let’s take a look at passing in a bad model.  If we neglect to update
the migration rates, then the model considers there to be no ancestry specified for
deme <code class="docutils literal notranslate"><span class="pre">1</span></code>:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bad_dmodel</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiscreteDemography</span><span class="p">(</span>
    <span class="n">mass_migrations</span><span class="o">=</span><span class="n">mass_migrations</span><span class="p">,</span> <span class="n">migmatrix</span><span class="o">=</span><span class="n">migmatrix</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">bad_dmodel</span><span class="o">.</span><span class="n">asblack</span><span class="p">())</span>
<span class="n">d</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DemographyDebugger</span><span class="p">([</span><span class="mi">100</span><span class="p">],</span> <span class="n">bad_dmodel</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s2">&quot;ANCESTRAL&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;DERIVED&quot;</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.DiscreteDemography(
    mass_migrations=[
        fwdpy11.MassMigration(
            when=100,
            source=0,
            destination=1,
            fraction=0.5,
            move_individuals=True,
            resets_growth_rate=True,
        )
    ],
    set_growth_rates=None,
    set_deme_sizes=None,
    set_selfing_rates=None,
    migmatrix=fwdpy11.MigrationMatrix(
        migmatrix=array([[1.0, 0.0], [0.0, 0.0]]), scaled=False
    ),
    set_migration_rates=None,
)
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">34</span><span class="p">],</span> <span class="n">line</span> <span class="mi">5</span>
<span class="g g-Whitespace">      </span><span class="mi">1</span> <span class="n">bad_dmodel</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiscreteDemography</span><span class="p">(</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span>     <span class="n">mass_migrations</span><span class="o">=</span><span class="n">mass_migrations</span><span class="p">,</span> <span class="n">migmatrix</span><span class="o">=</span><span class="n">migmatrix</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="p">)</span>
<span class="g g-Whitespace">      </span><span class="mi">4</span> <span class="nb">print</span><span class="p">(</span><span class="n">bad_dmodel</span><span class="o">.</span><span class="n">asblack</span><span class="p">())</span>
<span class="ne">----&gt; </span><span class="mi">5</span> <span class="n">d</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DemographyDebugger</span><span class="p">([</span><span class="mi">100</span><span class="p">],</span> <span class="n">bad_dmodel</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s2">&quot;ANCESTRAL&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;DERIVED&quot;</span><span class="p">})</span>

<span class="nn">File &lt;string&gt;:7,</span> in <span class="ni">__init__</span><span class="nt">(self, initial_deme_sizes, events, simlen, deme_labels)</span>

<span class="nn">File ~/work/fwdpy11/fwdpy11/fwdpy11/_types/demography_debugger.py:170,</span> in <span class="ni">DemographyDebugger.__post_init__</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">160</span>     <span class="n">event_list</span> <span class="o">=</span> <span class="n">_DemographyEvents</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">161</span>         <span class="n">temp</span><span class="o">.</span><span class="n">migmatrix</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">162</span>         <span class="n">temp</span><span class="o">.</span><span class="n">mass_migrations</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">166</span>         <span class="n">temp</span><span class="o">.</span><span class="n">set_migration_rates</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">167</span>     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">169</span> <span class="bp">self</span><span class="o">.</span><span class="n">migration_history</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_build_migration_history</span><span class="p">(</span><span class="n">event_list</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">170</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate_epochs</span><span class="p">(</span><span class="n">event_list</span><span class="p">)</span>

<span class="nn">File ~/work/fwdpy11/fwdpy11/fwdpy11/_types/demography_debugger.py:275,</span> in <span class="ni">DemographyDebugger._validate_epochs</span><span class="nt">(self, events)</span>
<span class="g g-Whitespace">    </span><span class="mi">272</span>                 <span class="n">ttl_rate_in</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">migration_matrix</span><span class="p">[</span><span class="n">epoch</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">deme</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">273</span>                 <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">274</span>                 <span class="k">if</span> <span class="ow">not</span> <span class="n">ttl_rate_in</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">275</span>                     <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">276</span>                         <span class="sa">f</span><span class="s2">&quot;the migration rantes into deme </span><span class="si">{</span><span class="n">epoch</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">deme</span><span class="si">}</span><span class="s2"> are 0.0 at time </span><span class="si">{</span><span class="n">i</span><span class="o">.</span><span class="n">start</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">277</span>                     <span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">278</span> <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">279</span>     <span class="k">if</span> <span class="n">events</span><span class="o">.</span><span class="n">migmatrix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

<span class="ne">ValueError</span>: the migration rantes into deme 1 are 0.0 at time 0
</pre></div>
</div>
</div>
</div>
<p>Our <code class="docutils literal notranslate"><span class="pre">dmodel2</span></code> from above is also invalid, as deme <code class="docutils literal notranslate"><span class="pre">1</span></code>
never gets created since we deleted the mass migration event:</p>
<div class="cell tag_raises-exception docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">DemographyDebugger</span><span class="p">([</span><span class="mi">100</span><span class="p">],</span> <span class="n">dmodel2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DemographyDebugger(initial_deme_sizes=[100], events=fwdpy11.DiscreteDemography(mass_migrations=None, set_growth_rates=None, set_deme_sizes=None, set_selfing_rates=None, migmatrix=fwdpy11.MigrationMatrix(migmatrix=array([[1., 0.],
       [0., 0.]]), scaled=False), set_migration_rates=[fwdpy11.SetMigrationRates(when=100, deme=0, migrates=[0.999, 0.001]), fwdpy11.SetMigrationRates(when=100, deme=1, migrates=[0.001, 0.999])]), simlen=None, deme_labels=None)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="setting-up-the-parameters-for-a-simulation">
<span id="model-params"></span><h2>Setting up the parameters for a simulation<a class="headerlink" href="#setting-up-the-parameters-for-a-simulation" title="Permalink to this headline">#</a></h2>
<p>Simulation parameters are stored in instances of <a class="reference internal" href="model_params.html#fwdpy11.ModelParams" title="fwdpy11.ModelParams"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.ModelParams</span></code></a>.
All of the examples shown above generated objects that we need to store in
such instances.</p>
<p>Like <a class="reference internal" href="types.html#fwdpy11.DiscreteDemography" title="fwdpy11.DiscreteDemography"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiscreteDemography</span></code></a>, <a class="reference internal" href="model_params.html#fwdpy11.ModelParams" title="fwdpy11.ModelParams"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.ModelParams</span></code></a> is
immutable after initialization.  Thus, I find it most convenient to
first put the objects into a <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a> and then “explode” it to
initialize a <a class="reference internal" href="model_params.html#fwdpy11.ModelParams" title="fwdpy11.ModelParams"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.ModelParams</span></code></a>.</p>
<p>The following example uses what we discussed above:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">mean</span><span class="o">=-</span><span class="mf">0.2</span><span class="p">)]</span>
<span class="n">recregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">PoissonInterval</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">mean</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)]</span>
<span class="n">migmatrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">migmatrix</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="n">mass_migrations</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">move_individuals</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">destination</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">fraction</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="p">]</span>
<span class="n">migrate</span> <span class="o">=</span> <span class="mf">1e-3</span>
<span class="n">set_migration_rates</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">SetMigrationRates</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">deme</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">migrates</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span> <span class="o">-</span> <span class="n">migrate</span><span class="p">,</span> <span class="n">migrate</span><span class="p">]),</span>
    <span class="n">fwdpy11</span><span class="o">.</span><span class="n">SetMigrationRates</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">deme</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">migrates</span><span class="o">=</span><span class="p">[</span><span class="n">migrate</span><span class="p">,</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="n">migrate</span><span class="p">]),</span>
<span class="p">]</span>
<span class="n">dmodel</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiscreteDemography</span><span class="p">(</span>
    <span class="n">mass_migrations</span><span class="o">=</span><span class="n">mass_migrations</span><span class="p">,</span>
    <span class="n">migmatrix</span><span class="o">=</span><span class="n">migmatrix</span><span class="p">,</span>
    <span class="n">set_migration_rates</span><span class="o">=</span><span class="n">set_migration_rates</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">dbg</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DemographyDebugger</span><span class="p">([</span><span class="mi">100</span><span class="p">],</span> <span class="n">dmodel</span><span class="p">,</span> <span class="mi">150</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gvalue</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Additive</span><span class="p">(</span>
    <span class="n">scaling</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
    <span class="n">gvalue_to_fitness</span><span class="o">=</span><span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSSmo</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Optimum</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">optimum</span><span class="o">=</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">1.0</span><span class="p">),</span>
            <span class="n">fwdpy11</span><span class="o">.</span><span class="n">Optimum</span><span class="p">(</span><span class="n">when</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">optimum</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">VS</span><span class="o">=</span><span class="mf">1.0</span><span class="p">),</span>
        <span class="p">]</span>
    <span class="p">),</span>
<span class="p">)</span>
<span class="n">pdict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;nregions&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="s2">&quot;sregions&quot;</span><span class="p">:</span> <span class="n">sregions</span><span class="p">,</span>
    <span class="s2">&quot;recregions&quot;</span><span class="p">:</span> <span class="n">recregions</span><span class="p">,</span>
    <span class="s2">&quot;rates&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1e-2</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span>
    <span class="s2">&quot;gvalue&quot;</span><span class="p">:</span> <span class="n">gvalue</span><span class="p">,</span>
    <span class="s2">&quot;prune_selected&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s2">&quot;simlen&quot;</span><span class="p">:</span> <span class="mi">150</span><span class="p">,</span>
    <span class="s2">&quot;demography&quot;</span><span class="p">:</span> <span class="n">dmodel</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">params</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">ModelParams</span><span class="p">(</span><span class="o">**</span><span class="n">pdict</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>There are a few new things here:</p>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">nregions</span></code> field specifies where neutral mutations occur.  We leave it empty
because we can add such mutations after the simulation is done.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">prune_selected</span></code> tells the simulation what to do with fixed selected mutations
after simplification. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, they will be removed from the simulation.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, they will be kept.  We keep them here because we are simulating an
additive trait, and the fixed genetic background is part of the genetic value
of an individual.  If we were instead simulating a standard population genetic model
with multiplicative fitness effects, we could use <code class="docutils literal notranslate"><span class="pre">True</span></code> because all such models
require is that relative fitnesses are preserved up to a multiplicative constant.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">rates</span></code> parameter is a list-like object with the neutral mutation rate,
selected mutation rate, and the recombination rate, respectively.  See the
next subsection for details.</p></li>
</ul>
<p>Let’s take a look at what we just built:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">params</span><span class="o">.</span><span class="n">asblack</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>fwdpy11.ModelParams(
    nregions=[],
    sregions=[
        fwdpy11.ExpS(
            beg=0.0,
            end=1.0,
            weight=1.0,
            mean=-0.2,
            h=1.0,
            coupled=True,
            label=0,
            scaling=1.0,
        )
    ],
    recregions=[fwdpy11.PoissonInterval(beg=0.0, end=0.1, mean=0.001, discrete=False)],
    rates=fwdpy11.MutationAndRecombinationRates(
        neutral_mutation_rate=0.0, selected_mutation_rate=0.01, recombination_rate=None
    ),
    gvalue=fwdpy11.Additive(
        scaling=2.0,
        gvalue_to_fitness=fwdpy11.GSSmo(
            optima=[
                fwdpy11.Optimum(optimum=0.0, VS=1.0, when=0),
                fwdpy11.Optimum(optimum=1.0, VS=1.0, when=100),
            ]
        ),
        noise=None,
        ndemes=1,
    ),
    demography=fwdpy11.DiscreteDemography(
        mass_migrations=[
            fwdpy11.MassMigration(
                when=100,
                source=0,
                destination=1,
                fraction=0.5,
                move_individuals=True,
                resets_growth_rate=True,
            )
        ],
        set_growth_rates=None,
        set_deme_sizes=None,
        set_selfing_rates=None,
        migmatrix=fwdpy11.MigrationMatrix(
            migmatrix=array([[1.0, 0.0], [0.0, 0.0]]), scaled=False
        ),
        set_migration_rates=[
            fwdpy11.SetMigrationRates(when=100, deme=0, migrates=[0.999, 0.001]),
            fwdpy11.SetMigrationRates(when=100, deme=1, migrates=[0.001, 0.999]),
        ],
    ),
    simlen=150,
    prune_selected=False,
)
</pre></div>
</div>
</div>
</div>
<section id="mutation-and-recombination-rates">
<span id="model-params-rate-details"></span><h3>Mutation and recombination rates<a class="headerlink" href="#mutation-and-recombination-rates" title="Permalink to this headline">#</a></h3>
<p>In general, the neutral mutation rate should be set to <code class="docutils literal notranslate"><span class="pre">0.0</span></code>.</p>
<p>The selected mutation rate is a non-negative <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a> representing the total mutation
rate per haploid genome per generation.</p>
<p>If you use instances of <a class="reference internal" href="regiontypes.html#fwdpy11.Region" title="fwdpy11.Region"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.Region</span></code></a> to set up the genetic map
(see <a class="reference internal" href="#recregions"><span class="std std-ref">here</span></a>), then you need to provide a non-negative <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.11)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>
representing the total recombination rate per meiosis.  If you use instances
of <code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GeneticMapUnit</span></code> (see <a class="reference internal" href="#geneticmapunit"><span class="std std-ref">here</span></a>), then use
<code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
</section>
</section>
<section id="running-a-simulation-with-tree-sequence-recording">
<span id="evolvets"></span><h2>Running a simulation with tree sequence recording<a class="headerlink" href="#running-a-simulation-with-tree-sequence-recording" title="Permalink to this headline">#</a></h2>
<p>To evolve the population with tree sequence recording, we make a call
to <a class="reference internal" href="functions.html#fwdpy11.evolvets" title="fwdpy11.evolvets"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.evolvets()</span></code></a>.  We also need an instance of <a class="reference internal" href="types.html#fwdpy11.GSLrng" title="fwdpy11.GSLrng"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.GSLrng</span></code></a>,
which is a random number generator.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
<span class="n">fwdpy11</span><span class="o">.</span><span class="n">evolvets</span><span class="p">(</span><span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span> <span class="n">pop</span><span class="o">=</span><span class="n">pop</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">params</span><span class="p">,</span> <span class="n">simplification_interval</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
<span class="n">pop</span><span class="o">.</span><span class="n">generation</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>150
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">simplification_interval</span></code> parameter directs <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> to run the tree
sequence simplification algorithm <span id="id11">[<a class="reference internal" href="../misc/bibliography.html#id10" title="Jerome Kelleher, Kevin R Thornton, Jaime Ashander, and Peter L Ralph. Efficient pedigree recording for fast population genetics simulation. PLoS Comput. Biol., 14(11):e1006581, November 2018. doi:10.1371/journal.pcbi.1006581.">KTAR18</a>]</span> every 25 generations.  Simplifying
often uses less total memory but requires a longer total run time.  Simplifying too
infrequently may consume too much memory.  In general, I use a value of <code class="docutils literal notranslate"><span class="pre">100</span></code> and
try not to think about it too much.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This simulation involves one deme splitting into two. The way that
this model is written, the <code class="docutils literal notranslate"><span class="pre">gvalue</span></code> parameter applies to both
demes. See <a class="reference internal" href="mvdes.html#mvdes"><span class="std std-ref">here</span></a> for how to simulate mutations with
different effect sizes in different demes and <a class="reference internal" href="../examples/localadaptation.html#localadaptation"><span class="std std-ref">here</span></a>
for a more complex example.</p>
</div>
<div class="admonition-todo admonition" id="id12">
<p class="admonition-title">Todo</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>How to track precise fixation times.
</pre></div>
</div>
</div>
</section>
<section id="recording-ancient-samples-during-a-simulation">
<span id="ancient-samples"></span><h2>Recording “ancient samples” during a simulation<a class="headerlink" href="#recording-ancient-samples-during-a-simulation" title="Permalink to this headline">#</a></h2>
<p>During a simulation, individuals can be marked as “preserved” or
“remembered”.  This marking means that their nodes in the node
table and their metadata are maintained, allowing you to retrieve
the data for later analysis, reconstruct genotypes, etc..</p>
<p>To mark samples as preserved, you need to define a callable object
equivalent to the following function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">record_samples</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">sampler</span><span class="p">):</span>
    <span class="k">pass</span>
</pre></div>
</div>
<p>The first argument to this callable will be the instance of
<a class="reference internal" href="types.html#fwdpy11.DiploidPopulation" title="fwdpy11.DiploidPopulation"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation</span></code></a> that you’ve passed on to
<a class="reference internal" href="functions.html#fwdpy11.evolvets" title="fwdpy11.evolvets"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.evolvets()</span></code></a> (see <a class="reference internal" href="#evolvets"><span class="std std-ref">above</span></a>).
The second argument will be an instance of <a class="reference internal" href="recorders.html#fwdpy11.SampleRecorder" title="fwdpy11.SampleRecorder"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.SampleRecorder</span></code></a>.
This second argument will be created internally by <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code>
and passed to your callable.</p>
<p>The reason why we use the term “callable” here instead of “function”
is that you will probably want to write callable classes rather
than actual functions.  For example, imagine that we only want
to record individuals whose fitnesses are in the top <code class="docutils literal notranslate"><span class="pre">X%</span></code>
of those found in deme <code class="docutils literal notranslate"><span class="pre">Y</span></code> every <code class="docutils literal notranslate"><span class="pre">10th</span></code> generation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">GetTopFitnessess</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">X</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">Y</span> <span class="o">=</span> <span class="n">Y</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pop</span><span class="p">,</span> <span class="n">sampler</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">pop</span><span class="o">.</span><span class="n">generation</span> <span class="o">%</span> <span class="mi">10</span> <span class="o">==</span> <span class="mf">0.0</span><span class="p">:</span>
            <span class="n">md</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">diploid_metadata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">inY</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">md</span><span class="p">[</span><span class="s2">&quot;deme&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">Y</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">w</span> <span class="o">=</span> <span class="n">md</span><span class="p">[</span><span class="s2">&quot;w&quot;</span><span class="p">][</span><span class="n">inY</span><span class="p">]</span>
            <span class="n">q</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">w</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">Y</span><span class="p">)</span>
            <span class="n">geqq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">w</span> <span class="o">&gt;=</span> <span class="n">q</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">sampler</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">inY</span><span class="p">[</span><span class="n">geqq</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<p>The above class is not well-implemented at all, but it will suffice for our example,
where we pass an instance on to <a class="reference internal" href="functions.html#fwdpy11.evolvets" title="fwdpy11.evolvets"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.evolvets()</span></code></a>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
<span class="n">recorder</span> <span class="o">=</span> <span class="n">GetTopFitnessess</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">fwdpy11</span><span class="o">.</span><span class="n">evolvets</span><span class="p">(</span>
    <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span> <span class="n">pop</span><span class="o">=</span><span class="n">pop</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">params</span><span class="p">,</span> <span class="n">simplification_interval</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">recorder</span><span class="o">=</span><span class="n">recorder</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>At the end of the simulation, we indeed have metadata from “ancient samples”:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">amd</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">ancient_sample_metadata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="n">amd</span><span class="p">[</span>
        <span class="o">-</span><span class="mi">5</span><span class="p">:,</span>
    <span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[(0., 0., 0.60653066, [0., 0., 0.], 45, [ 9, 13], 0, 0, [2590, 2591])
 (0., 0., 0.60653066, [0., 0., 0.], 46, [11, 19], 0, 0, [2592, 2593])
 (0., 0., 0.60653066, [0., 0., 0.], 47, [16, 17], 0, 0, [2594, 2595])
 (0., 0., 0.60653066, [0., 0., 0.], 48, [43, 28], 0, 0, [2596, 2597])
 (0., 0., 0.60653066, [0., 0., 0.], 49, [34, 10], 0, 0, [2598, 2599])]
</pre></div>
</div>
</div>
</div>
<p>The following sampler types are built in to <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code>:</p>
<ul class="simple">
<li><p><a class="reference internal" href="recorders.html#fwdpy11.RandomAncientSamples" title="fwdpy11.RandomAncientSamples"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.RandomAncientSamples</span></code></a></p></li>
</ul>
</section>
<section id="tracking-features-of-the-population-during-a-simulation">
<span id="pop-from-tskit"></span><h2>Tracking features of the population during a simulation<a class="headerlink" href="#tracking-features-of-the-population-during-a-simulation" title="Permalink to this headline">#</a></h2>
<p>The concept of passing a callable into the simulation has more uses
than recording ancient samples.  Basically, we can track pretty
much anything.  The following example tracks the mean genetic value
in each deme every generation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">class</span> <span class="nc">TrackGvalues</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pop</span><span class="p">,</span> <span class="n">sampler</span><span class="p">):</span>
        <span class="n">md</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">diploid_metadata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]:</span>
            <span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">md</span><span class="p">[</span><span class="s2">&quot;deme&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">w</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">pop</span><span class="o">.</span><span class="n">generation</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">md</span><span class="p">[</span><span class="s2">&quot;g&quot;</span><span class="p">][</span><span class="n">w</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
<span class="n">tracker</span> <span class="o">=</span> <span class="n">TrackGvalues</span><span class="p">()</span>
<span class="n">fwdpy11</span><span class="o">.</span><span class="n">evolvets</span><span class="p">(</span>
    <span class="n">rng</span><span class="o">=</span><span class="n">rng</span><span class="p">,</span> <span class="n">pop</span><span class="o">=</span><span class="n">pop</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">params</span><span class="p">,</span> <span class="n">simplification_interval</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">recorder</span><span class="o">=</span><span class="n">tracker</span>
<span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tracker</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="o">-</span><span class="mi">5</span><span class="p">:]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(148, 1, -0.004283821388400186)
(149, 0, -0.021819672969005686)
(149, 1, -0.018201916502682675)
(150, 0, -0.013648388151684962)
(150, 1, -0.011121088973856224)
</pre></div>
</div>
</div>
</div>
</section>
<section id="starting-a-simulation-with-the-output-of-a-coalescent-simulation">
<span id="starting-from-msprime"></span><h2>Starting a simulation with the output of a coalescent simulation<a class="headerlink" href="#starting-a-simulation-with-the-output-of-a-coalescent-simulation" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="n">Ne</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="o">.</span><span class="n">create_from_tskit</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
<span class="n">pop</span><span class="o">.</span><span class="n">N</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100
</pre></div>
</div>
</div>
</div>
<p>We can also start from a simulation of multiple demes:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">m</span> <span class="o">=</span> <span class="mf">1e-3</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span>
    <span class="n">Ne</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">recombination_rate</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span>
    <span class="n">population_configurations</span><span class="o">=</span><span class="p">[</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">PopulationConfiguration</span><span class="p">(</span><span class="n">sample_size</span><span class="o">=</span><span class="mi">100</span><span class="p">),</span>
        <span class="n">msprime</span><span class="o">.</span><span class="n">PopulationConfiguration</span><span class="p">(</span><span class="n">sample_size</span><span class="o">=</span><span class="mi">100</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">migration_matrix</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">])</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
<span class="p">)</span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="o">.</span><span class="n">create_from_tskit</span><span class="p">(</span><span class="n">ts</span><span class="p">)</span>
<span class="n">pop</span><span class="o">.</span><span class="n">N</span>
<span class="n">md</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pop</span><span class="o">.</span><span class="n">diploid_metadata</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">md</span><span class="p">[</span><span class="s2">&quot;deme&quot;</span><span class="p">],</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(array([0, 1], dtype=int32), array([50, 50]))
</pre></div>
</div>
</div>
</div>
<p>See <a class="reference internal" href="../short_vignettes/initpops_vignette.html#precapitation"><span class="std std-ref">here</span></a> for a detailed example.</p>
<p>Currently, <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> does <strong>not</strong> support starting from <code class="docutils literal notranslate"><span class="pre">msprime</span></code> simulations
with mutations.  The variants will simply be ignored.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When simulating very large genomic regions, it is not a good
idea to use the standard model in <code class="docutils literal notranslate"><span class="pre">msprime</span></code>. See <span id="id13">[<a class="reference internal" href="../misc/bibliography.html#id6" title="Dominic Nelson, Jerome Kelleher, Aaron P Ragsdale, Claudia Moreau, Gil McVean, and Simon Gravel. Accounting for long-range correlations in genome-wide simulations of large cohorts. PLoS Genet., 16(5):e1008619, May 2020. doi:10.1371/journal.pgen.1008619.">NKR+20</a>]</span>
for details and also <a class="reference external" href="https://tskit.dev/msprime/docs/latest/api.html#msprime.DiscreteTimeWrightFisher" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">msprime.DiscreteTimeWrightFisher</span></code></a>.
Also note that starting with <code class="docutils literal notranslate"><span class="pre">msprime</span></code> output doesn’t
guarantee that you don’t need to “burn in” the simulation
(see <a class="reference internal" href="advancedtopics.html#howlongtorun"><span class="std std-ref">here</span></a>).</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Not all genetic maps supported by <code class="docutils literal notranslate"><span class="pre">fwdpy11</span></code> are possible
in <code class="docutils literal notranslate"><span class="pre">msprime</span></code>. Specifically, discrete “jumps” in the
genetic map (<em>e.g.</em>, <a class="reference internal" href="regiontypes.html#fwdpy11.BinomialPoint" title="fwdpy11.BinomialPoint"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.BinomialPoint</span></code></a>) are
not compatible with <code class="docutils literal notranslate"><span class="pre">msprime</span></code>.  Further, the approach
outlined <a class="reference external" href="https://msprime.readthedocs.io/en/stable/tutorial.html#multiple-chromosomes">here</a>
is not sufficient to get the recombination rates right
between separate “chromosomes”.</p>
</div>
</section>
<section id="converting-the-data-to-a-tskit-treesequence">
<span id="tskittransfer"></span><h2>Converting the data to a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a><a class="headerlink" href="#converting-the-data-to-a-tskit-treesequence" title="Permalink to this headline">#</a></h2>
<p>The tree sequence data structures may be converted to the analogous <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
objects using <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation.dump_tables_to_tskit" title="fwdpy11.DiploidPopulation.dump_tables_to_tskit"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation.dump_tables_to_tskit()</span></code></a>,
which returns a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>.</p>
<p>The most basic usage is:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">dump_tables_to_tskit</span><span class="p">()</span>
</pre></div>
</div>
<p>When you have the data stored as a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>,
information about individuals, mutations, etc., is stored as table metadata.
See <a class="reference internal" href="../long_vignettes/tskit_metadata_vignette.html#tskit-metadata-vignette"><span class="std std-ref">here</span></a> to learn how to decode the metadata.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Once <code class="docutils literal notranslate"><span class="pre">tskit</span></code> 0.3.0 is released, the metadata encoding will change
quite a bit and it will be simpler and more efficient to decode.</p>
</div>
<p>You may provide a <code class="docutils literal notranslate"><span class="pre">dict</span></code> that reflects the simulation parameters used.  This <code class="docutils literal notranslate"><span class="pre">dict</span></code>
will be part of the provenance information encoded in the <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>.
For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assuming mp is a fwdpy11.ModelParams:</span>
<span class="n">ts</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">dump_tables_to_tskit</span><span class="p">(</span><span class="n">parameters</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">mp</span><span class="p">),</span> <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">12345</span><span class="p">})</span>
</pre></div>
</div>
<p>Ultimately, it is up to you to decide what to include in <code class="docutils literal notranslate"><span class="pre">parameters</span></code>.
For example, it could be a script:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Bonus points for somehow including the git commit hash corresponding</span>
<span class="c1"># to the version of the script that you used !</span>
<span class="n">parameters</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;script&quot;</span><span class="p">:</span> <span class="s2">&quot;/path/to/script&quot;</span><span class="p">,</span> <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;script&quot;</span><span class="p">,</span> <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">1234</span><span class="p">}</span>
</pre></div>
</div>
<p>You can go further than that, and even include the entire script.
It turns out that Python files know who they are and can read themselves:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">read_self</span><span class="p">():</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="vm">__file__</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">script</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">script</span>


<span class="n">script</span> <span class="o">=</span> <span class="n">read_self</span><span class="p">()</span>
<span class="n">parameters</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;script&quot;</span><span class="p">:</span> <span class="n">script</span><span class="p">,</span> <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;script&quot;</span><span class="p">,</span> <span class="s2">&quot;seed&quot;</span><span class="p">:</span> <span class="mi">1234</span><span class="p">}</span>
</pre></div>
</div>
<p>In order to get the provenance information back out from a <a class="reference external" href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" title="(in Python)"><code class="xref py py-class docutils literal notranslate"><span class="pre">tskit.TreeSequence</span></code></a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">json</span>

<span class="n">provenance</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">provenance</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>
</pre></div>
</div>
<p>If you recorded an instance of <a class="reference internal" href="model_params.html#fwdpy11.ModelParams" title="fwdpy11.ModelParams"><code class="xref py py-class docutils literal notranslate"><span class="pre">fwdpy11.ModelParams</span></code></a> as your <code class="docutils literal notranslate"><span class="pre">parameters</span></code>, you
can even reconstruct the original object (if you have the correct modules imported).
For example, if we assume that we encoded the model parameters as shown two listings ago:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">tskit</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">fwdpy11</span>

<span class="n">ts</span> <span class="o">=</span> <span class="n">tskit</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;sim.trees&quot;</span><span class="p">)</span>
<span class="n">provenance</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">ts</span><span class="o">.</span><span class="n">provenance</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>
<span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span>  <span class="c1"># Annoyance!</span>
<span class="n">tmp</span> <span class="o">=</span> <span class="nb">eval</span><span class="p">(</span><span class="n">provenance</span><span class="p">[</span><span class="s2">&quot;parameters&quot;</span><span class="p">][</span><span class="s2">&quot;model&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>It is possible for model parameters to contain <code class="docutils literal notranslate"><span class="pre">numpy</span></code> arrays.  Unfortunately, their
string representations are not namespace-qualified, meaning that they say <code class="docutils literal notranslate"><span class="pre">array</span></code> rather
than <code class="docutils literal notranslate"><span class="pre">numpy.array</span></code> or <code class="docutils literal notranslate"><span class="pre">np.array</span></code>. Thus, I made a type alias so that the <code class="docutils literal notranslate"><span class="pre">eval</span></code> would work.</p>
</section>
<section id="saving-the-results-of-a-simulation-to-disk">
<span id="savingsimstodisk"></span><h2>Saving the results of a simulation to disk<a class="headerlink" href="#saving-the-results-of-a-simulation-to-disk" title="Permalink to this headline">#</a></h2>
<p>To dump the output of a simulation to an uncompressed binary file, use
<a class="reference internal" href="types.html#fwdpy11.DiploidPopulation.dump_to_file" title="fwdpy11.DiploidPopulation.dump_to_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation.dump_to_file()</span></code></a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pop</span><span class="o">.</span><span class="n">dump_to_file</span><span class="p">(</span><span class="s2">&quot;pop.bin&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>To restore a population from such a file, call the static method
<a class="reference internal" href="types.html#fwdpy11.DiploidPopulation.load_from_file" title="fwdpy11.DiploidPopulation.load_from_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation.load_from_file()</span></code></a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy11</span><span class="o">.</span><span class="n">DiploidPopulation</span><span class="o">.</span><span class="n">load_from_file</span><span class="p">(</span><span class="s2">&quot;pop.bin&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>There are two means of storing a population in <a class="reference external" href="https://docs.python.org/3/library/pickle.html#module-pickle" title="(in Python v3.11)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pickle</span></code></a> format.
The first is:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;pickled_pop.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
</pre></div>
</div>
<p>However, this method is not the most efficient.  The following takes less
memory, which is probably important for really big simulations:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;pickled_pop.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">pop</span><span class="o">.</span><span class="n">pickle_to_file</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>These two pickling methods create very different files!
The first can be read back in with <a class="reference external" href="https://docs.python.org/3/library/pickle.html#pickle.load" title="(in Python v3.11)"><code class="xref py py-func docutils literal notranslate"><span class="pre">pickle.load()</span></code></a>
but the second <strong>must</strong> be read back in with the
static function <a class="reference internal" href="types.html#fwdpy11.DiploidPopulation.load_from_pickle_file" title="fwdpy11.DiploidPopulation.load_from_pickle_file"><code class="xref py py-func docutils literal notranslate"><span class="pre">fwdpy11.DiploidPopulation.load_from_pickle_file()</span></code></a>.</p>
</div>
<section id="outputting-a-tskit-trees-file">
<h3>Outputting a <code class="docutils literal notranslate"><span class="pre">tskit</span></code> “trees file”<a class="headerlink" href="#outputting-a-tskit-trees-file" title="Permalink to this headline">#</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ts</span> <span class="o">=</span> <span class="n">pop</span><span class="o">.</span><span class="n">dump_tables_to_tskit</span><span class="p">()</span>
<span class="n">ts</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="s2">&quot;treefile.trees&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./pages"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="definitions.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Definitions</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="advancedtopics.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Advanced topics</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Kevin Thornton<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>