{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "absolute-short",
   "metadata": {},
   "source": [
    "(advancedtopics)=\n",
    "\n",
    "# Advanced topics\n",
    "\n",
    "## Executing multiple replicates in parallel using `concurrent.futures`\n",
    "\n",
    "You have many options when it comes to running many replicates of a model.\n",
    "For example, you could write a single Python script plus a shell script\n",
    "that your local cluster understands, sending your simulation to hundreds\n",
    "of compute nodes.\n",
    "\n",
    "This section introduces {mod}`concurrent.futures` as a method to execute\n",
    "many replicates in separate Python *processes* using a single script.\n",
    "\n",
    "Fully worked-out examples of varying complexity can be found here:\n",
    "\n",
    "* {ref}`bgs_vignette`\n",
    "* {ref}`IMexample`\n",
    "* {ref}`migtest`\n",
    "* {ref}`precapitation`\n",
    "* {ref}`recapitation`\n",
    "\n",
    "A common set of idioms used by these examples are:\n",
    "\n",
    "* The function to run the simulation typically does not return\n",
    "  the instance of {class}`fwdpy11.DiploidPopulation`.  Such\n",
    "  a return requires pickling, which is very expensive.  If you\n",
    "  need to store these population, write them to files rather\n",
    "  than returning them. See {ref}`here <savingsimstodisk>`.\n",
    "* If you do return something to the main \"collector\"\n",
    "  process, keep it simple!\n",
    "* Random number seeds are set in `__main__` using\n",
    "  {func}`numpy.random.randint` after calling\n",
    "  {func}`numpy.random.seed` using a user-provided seed.\n",
    "\n",
    "(howlongtorun)=\n",
    "\n",
    "## How long to run the simulation\n",
    "\n",
    "Deciding how long to run a simulation depends on whether or not\n",
    "a \"burn in\" phase is required to get the model to statistical\n",
    "equilibrium.  For example, if you want to get the steady-state\n",
    "properties of a model with many selected mutations and linkage,\n",
    "you will need to burn in.  If you want to know how the short term\n",
    "dynamics of the neutral site-frequency spectrum are affected by suddenly\n",
    "introducing strongly-beneficial mutations, then you may not need to\n",
    "burn in.\n",
    "\n",
    "So, then, how long to burn in?  Clearly, the answer should be \"the\n",
    "least that you have to\"!  As always, there are a few different\n",
    "things to consider:\n",
    "\n",
    "* You want the *distribution* of final outcomes to be \"correct\".\n",
    "  Theory often gives us expressions for expectations and\n",
    "  sometimes for variances of things we care about, like\n",
    "  the number of mutations in a sample. However, comparing\n",
    "  distributions will often require a statistical analysis\n",
    "  comparing the output of independent simulations.\n",
    "  For example, for a neutral model, the distribution of the\n",
    "  number of mutations under an infinitely-many sites model\n",
    "  for a small sample should match `msprime` very well.\n",
    "  For a steady-state model of recurrent hitch-hiking\n",
    "  {cite}`Kaplan1989-rt`, results from small samples\n",
    "  should match coalescent simulations such as {cite}`Kern2016-wg`.\n",
    "  Clearly, all of the different methods\n",
    "  should match analytical results where possible.\n",
    "* You probably want all of your trees to be fully coalesced\n",
    "  to a single common ancestor.  The time back to a final\n",
    "  ancestor has a very long tail.  In models with recombination,\n",
    "  it is not uncommon to have a tree or three with multiple roots\n",
    "  at the end of a simulation.\n",
    "\n",
    "For the first point, a burn-in of `10N` generations or so\n",
    "is usually sufficient. Before we simulated with tree sequence\n",
    "recording, we simulated entire \"genomes\" (neutral mutations and all),\n",
    "which was rather slow.  (Here \"we\" is the field in general.)  We then\n",
    "(hopefully!) compared our results to something like `msprime`.  The\n",
    "agreement was really good, so one answer is \"about `10N`.  I'm being\n",
    "vague about `N` here--typically, it would be equal to the starting effective\n",
    "population size of your population (*e.g.* in the absence of selection).\n",
    "\n",
    "With tree sequence recording, we run into the \"uncoalesced marginal\n",
    "trees\" issue mentioned above.  We could just simulate much longer\n",
    "to get rid of this problem, but it is simply easier to start\n",
    "with a tree sequence from `msprime` (see {ref}`here <starting-from-msprime>`).\n",
    "\n",
    "(eyre-walker)=\n",
    "\n",
    "## The \"Eyre-Walker\" model of complex traits\n",
    "\n",
    "{cite}`Eyre-Walker2010-rs` describes a model relating mutations with\n",
    "fitness effect {math}`S = |2Ns|` to {math}`z`, their effect on\n",
    "a phenotype/trait according to {math}`z = \\delta S^\\tau(1+\\epsilon)`.\n",
    "Here, {math}`\\epsilon` is a draw from a Gaussian distribution with mean zero,\n",
    "{math}`\\delta` is {math}`1` or {math}`-1` with equal probability, and\n",
    "{math}`\\tau` \"tunes\" the correlation between fitness effect and effect on the trait.\n",
    "\n",
    "Implementing this model is quite straightforward, as the trait values do not affect\n",
    "the dynamics of the model.  For this flavor of a \"pure pleiotropy\" model,\n",
    "the technical details reduce to a standard population genetic model\n",
    "where we tack on the trait values at the end.\n",
    "\n",
    "Let's set up a model where {math}`S` is exponentially distributed with mean `-20`.\n",
    "We'll run a small population for a few generations.  This model will be nowhere\n",
    "near equilibrium, but we're just using it as an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "balanced-genius",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    }
   ],
   "source": [
    "import fwdpy11\n",
    "import numpy as np\n",
    "\n",
    "N = 1000\n",
    "sregions = [fwdpy11.ExpS(beg=0.0, end=1.0, weight=1.0, mean=-20, scaling=2 * N)]\n",
    "recregions = [fwdpy11.PoissonInterval(beg=0.0, end=0.1, mean=1e-3)]\n",
    "gvalue = fwdpy11.Multiplicative(scaling=2.0)\n",
    "pdict = {\n",
    "    \"nregions\": [],\n",
    "    \"sregions\": sregions,\n",
    "    \"recregions\": recregions,\n",
    "    \"rates\": (0.0, 1e-3, None),\n",
    "    \"gvalue\": gvalue,\n",
    "    \"prune_selected\": False,\n",
    "    \"simlen\": 150,\n",
    "}\n",
    "params = fwdpy11.ModelParams(**pdict)\n",
    "pop = fwdpy11.DiploidPopulation(N, 1.0)\n",
    "rng = fwdpy11.GSLrng(54321)\n",
    "fwdpy11.evolvets(rng, pop, params, 100)\n",
    "print(len(pop.tables.mutations))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "clear-badge",
   "metadata": {},
   "source": [
    "Define a function relating {math}`S` to {math}`z`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "distinct-ownership",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getz(S, tau, sigma):\n",
    "    if np.random.uniform() < 0.5:\n",
    "        delta = 1\n",
    "    else:\n",
    "        delta = -1\n",
    "    epsilon = np.random.normal(loc=0, scale=sigma, size=1)[0]\n",
    "    return delta * np.power(np.abs(S), tau) * (1.0 + epsilon)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "crazy-gentleman",
   "metadata": {},
   "source": [
    "Apply the function and look at the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "stuck-wisdom",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-0.013650793652025846 15.03525114016863\n",
      "-0.007424030753135078 -1.8733412999601573\n",
      "-0.0015970102395456 -2.628496658148799\n",
      "-0.007935780438803351 -2.7240381012841626\n",
      "-0.004531235901923532 -2.7792831274736045\n",
      "-0.004083990901721122 3.2826503807113983\n",
      "-0.0007740534491111242 -2.401179214236283\n",
      "-0.0021590316871619632 -5.171505563855891\n",
      "-0.014729668076822772 1.5835049661103993\n",
      "-0.0009353737988953688 2.4355050590839866\n",
      "-0.003567001189691813 3.909023772537182\n",
      "-0.0053709050825429296 -5.534331330636803\n",
      "-0.009093123747102732 0.8602921563750122\n"
     ]
    }
   ],
   "source": [
    "np.random.seed(101010)\n",
    "zvals = {}\n",
    "for i, m in enumerate(pop.tables.mutations):\n",
    "    zvals[m.key] = getz(2 * N * pop.mutations[m.key].s, 0.5, 1.0)\n",
    "\n",
    "for k, v in zvals.items():\n",
    "    print(pop.mutations[k].s, v)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ambient-conservation",
   "metadata": {},
   "source": [
    "Traverse the tree sequence to get individual phenotypes under a\n",
    "strictly additive model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "pointed-committee",
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes = np.zeros(pop.N)\n",
    "node_to_individual = {}\n",
    "for i, j in enumerate(pop.diploid_metadata):\n",
    "    assert j.nodes[0] not in node_to_individual\n",
    "    assert j.nodes[1] not in node_to_individual\n",
    "    node_to_individual[j.nodes[0]] = i\n",
    "    node_to_individual[j.nodes[1]] = i\n",
    "ti = fwdpy11.TreeIterator(pop.tables, pop.alive_nodes, update_samples=True)\n",
    "for t in ti:\n",
    "    for m in t.mutations():\n",
    "        for n in t.samples_below(m.node):\n",
    "            phenotypes[node_to_individual[n]] += zvals[m.key]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "urban-outreach",
   "metadata": {},
   "source": [
    "The trait value distribution is:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "seventh-district",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([-5.53433133, -3.98651989, -2.77928313, -2.62849666, -2.40117921,\n",
       "        -2.25168095,  0.        ,  0.86029216,  1.18498567,  1.58350497,\n",
       "         2.36997134,  2.43550506,  2.76849064, 13.16190984, 15.03525114]),\n",
       " array([ 30,   1,   7,   1,   1,   1, 897,   4,  32,  13,   1,   1,   1,\n",
       "          1,   9]))"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(phenotypes, return_counts=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rubber-course",
   "metadata": {},
   "source": [
    "The mean trait value and the genetic variance are:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "polished-exchange",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.3135909660572516"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotypes.mean()\n",
    "phenotypes.var()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "appreciated-accommodation",
   "metadata": {},
   "source": [
    "For our final trick, let's store them in the individual metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "subtle-agreement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0 0.0\n",
      "0.0 0.0\n",
      "-2.7792831274736045 -2.7792831274736045\n",
      "0.0 0.0\n",
      "0.0 0.0\n"
     ]
    }
   ],
   "source": [
    "md = np.array(pop.diploid_metadata, copy=False)\n",
    "md[\"g\"][:] = phenotypes\n",
    "\n",
    "for i, j in zip(md[\"g\"][:5], phenotypes[:5]):\n",
    "    print(i, j)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "extended-helmet",
   "metadata": {},
   "source": [
    "The trick is that the `numpy` array is a *non-owning* array, meaning\n",
    "that it is a simple \"view\" of the underlying `C++` data.  Further,\n",
    "it happens to be read/write, and thus we can modify it.  (Try not\n",
    "to abuse this.  More often than not, you'll just break stuff.)\n",
    "\n",
    "I feel that some comments about this model are warranted:\n",
    "\n",
    "* The model is a simplification of earlier work by Keightley and Hill\n",
    "  ({cite}`Keightley1988-rn`, {cite}`Keightley1990-az`), which should be cited\n",
    "  alongside {cite}`Eyre-Walker2010-rs`.\n",
    "* {cite}`Keightley1988-rn` and {cite}`Keightley1990-az`) show that this model predicts heritability\n",
    "  (the genetic variance) is linear-ish with `N`, with the details depending\n",
    "  somewhat on the model parameters.  The biological reasonableness\n",
    "  of that prediction is dubious.  {cite}`Johnson2005-bp` discuss this point\n",
    "  in some detail, and give other relevant references.\n",
    "* Depending somewhat on the parameters of the `getz` function, one will eventually\n",
    "  generate an intermediate-frequency variant with a massive {math}`z`,\n",
    "  meaning that it would explain a considerable amount of the genic\n",
    "  variance for the trait (high {math}`2pqz^2`).  Such outcomes are contrary\n",
    "  to the results of human GWAS.\n",
    "* It becomes tempting to start doing arbitrary things when applying this model\n",
    "  to simulation output.  For example, only treating mutations in certain frequency\n",
    "  ranges as affecting trait values.  However, doing so means the predictions\n",
    "  made by such procedures are not natural outcomes of evolutionary models.\n",
    "  Consider only applying the `getz` function to mutations with frequency\n",
    "  {math}`< x` in order to say something about \"rare alleles\".\n",
    "  This treatment of the data actually changes the model: mutations that are common\n",
    "  now (at the end of the simulation) were rare at some point in the past,\n",
    "  and had frequencies {math}`< x`.  Therefore, the model is one where mutations\n",
    "  suddenly stop affecting the trait once they hit some critical frequency.\n",
    "  Presumably, the same variants would affect the trait again should they drift to\n",
    "  a frequency below {math}`x` if the simulation is run a bit longer.\n",
    "\n",
    "(gvalues_python)=\n",
    "\n",
    "## Writing genetic value classes in Python\n",
    "\n",
    "Version `0.9.0` added the ability to write custom genetic value types in\n",
    "Python.  Such classes are useful for prototyping new models\n",
    "and/or generating example models for teaching/illustrating concepts.\n",
    "\n",
    ":::{versionadded} 0.9.0\n",
    "\n",
    ":::\n",
    "\n",
    "### Concepts\n",
    "\n",
    "(python_genetic_value_dimensions)=\n",
    "\n",
    "#### The \"dimensionality\" of a model\n",
    "\n",
    "`fwdpy11` supports both univariate and multivariate trait simulations.\n",
    "Thus, a genetic value implicitly has a dimensionality. For the\n",
    "case of a standard population-genetic simulation where mutations\n",
    "directly affect fitness, that dimensionality is `1`.  Simply put,\n",
    "the genetic value is fitness.\n",
    "\n",
    "For a simulation of mutations affecting `n` traits, the dimensionality\n",
    "is `n`.\n",
    "\n",
    "The dimensionality matters because we need to define a single genetic value\n",
    "to populate the {attr}`fwdpy11.DiploidMetadata.g` field for an offspring.\n",
    "Doing so is simple for a one-dimensional trait.  For a simulation with\n",
    "pleiotropy, we have a decision to make.  Should {attr}`fwdpy11.DiploidMetadata.g`\n",
    "store the Euclidean distance from the optima?  Or should it store the\n",
    "individual's value for some \"focal\" trait?  The `API` described here is\n",
    "flexible enough to accommodate both.\n",
    "\n",
    "The dimensionality of your genetic value to fitness map and your genetic value\n",
    "object must match.  For the trait examples discussed above, this requirement\n",
    "should make good sense.  For the case of modeling correlated effect sizes\n",
    "of mutations in different demes (see {ref}`here <mvdes>`), we still require\n",
    "this although it makes less sense. (Such simulations are limited to either\n",
    "the standard population genetic case or the case of selection on a single\n",
    "trait.  But relaxing this requirement for some models but not others becomes\n",
    "problematic.)\n",
    "\n",
    "The current version of this section only discusses one-dimensional genetic\n",
    "values.\n",
    "\n",
    "```{eval-rst}\n",
    ".. todo:: \n",
    "\n",
    "    Update for multi-dimensional genetic values once we get documentation\n",
    "    in place for simulations of Gaussian stabilizing selection with pleiotropy.\n",
    "```\n",
    "\n",
    "(python_genetic_value_update_desc)=\n",
    "\n",
    "#### Supporting time-dependent models\n",
    "\n",
    "Genetic value methods may have dynamic states that depend on some\n",
    "state of the population.  For example, something may change once\n",
    "`pop.generation > x`.\n",
    "\n",
    "To support such time-dependent states, classes have a method with\n",
    "the following signature:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "def update(self, pop: fwdpy11.DiploidPopulation) -> None:\n",
    "    pass\n",
    "\n",
    "```\n",
    "\n",
    "A type with no time-dependent state can in fact use the function shown\n",
    "above. A concrete example is shown {ref}`below <python-gvalue-to-fitness>`,\n",
    "in an example class called `PyGSSRandomOptimum`.  That example does not\n",
    "use any information from `pop` to update its state.\n",
    "\n",
    "### Performance\n",
    "\n",
    "It will be hard to get \"awesome\" performance with Python genetic\n",
    "value classes.  The culprit are abstract methods that get called\n",
    "once per offspring.  That's simply a lot of Python/C++ boundary\n",
    "crossing.  The `update` function described above is called only\n",
    "once per generation, making it much less of a performance killer.\n",
    "\n",
    "We've tried hard to make the data exchange from C++ to your Python\n",
    "code as efficient as possible.  We accomplished this by carefully\n",
    "considering how we implement the `data` argument that gets passed\n",
    "in.  Benchmarking showed that carefully designed proxies to the\n",
    "underlying C++ data greatly improved run times.\n",
    "\n",
    "Importantly, you do not have to write every component in Python!\n",
    "If you are interested in additive effects models but a new \"noise\"\n",
    "model, then use {class}`fwdpy11.Additive` along with your custom\n",
    "type.\n",
    "\n",
    "`fwdpy11` provides the following helper functions to improve efficiency:\n",
    "\n",
    "```{py:function} fwdpy11.strict_additive_effects\n",
    "\n",
    "This function computes the sum of effect sizes in a\n",
    "diploid genome, ignoring dominance.  The return value\n",
    "is an offset from zero, so add 1.0 to convert to a\n",
    "fitness if needed.\n",
    "\n",
    ":param pop: The population\n",
    ":type pop: fwdpy11.DiploidPopulation\n",
    ":param metadata: The offspring metadata\n",
    ":type metadata: fwdpy11.DiploidMetadata\n",
    ":returns: Sum of effect sizes\n",
    ":rtype: float\n",
    "\n",
    "```\n",
    "\n",
    "```{py:function} fwdpy11.additive_effects\n",
    "\n",
    "This function computes the sum of effect sizes in a\n",
    "diploid genome, accounting for dominance.  The return value\n",
    "is an offset from zero, so add 1.0 to convert to a\n",
    "fitness if needed.\n",
    "\n",
    ":param pop: The population\n",
    ":type pop: fwdpy11.DiploidPopulation\n",
    ":param metadata: The offspring metadata\n",
    ":type metadata: fwdpy11.DiploidMetadata\n",
    ":type scaling: float\n",
    ":returns: Sum of effect sizes (accounting for heterozygous effects)\n",
    ":rtype: float\n",
    "\n",
    "```\n",
    "\n",
    "The first function is equivalent to the following Python code:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "def strict_additive_effects(\n",
    "    pop: fwdpy11.DiploidPopulation,\n",
    "    metadata: fwdpy11.DiploidMetadata,\n",
    ") -> float:\n",
    "    g = 0.0\n",
    "    dip = pop.diploids[metadata.label]\n",
    "    for i in [dip.first, dip.second]:\n",
    "        for k in pop.haploid_genomes[i].smutations:\n",
    "            g += pop.mutations[k].s\n",
    "    return g\n",
    "\n",
    "```\n",
    "\n",
    "The second function uses C++ code from `fwdpp` to do the calculation\n",
    "accounting for dominance effects.\n",
    "\n",
    "### Technical issues\n",
    "\n",
    "This section discusses some important technicalities.\n",
    "\n",
    "#### Abstract base classes\n",
    "\n",
    "In object oriented lingo, the `fwdpy11` base classes here are *abstract\n",
    "base classes* or `ABCs`.  This means that they simply describe a required\n",
    "interface.  When you inherit from an ABC, you must fill in the details\n",
    "of that interface.\n",
    "\n",
    "#### Initializing the super class\n",
    "\n",
    "When writing object-oriented code, it is important that a derived\n",
    "class properly initialize the base class.  In Python, base\n",
    "classes are often called `superclasses` and derived classes are\n",
    "`subclasses`.\n",
    "\n",
    "The idiomatic approach in `Python 3` is:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cloudy-carter",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Base(object):\n",
    "   def __init__(self):\n",
    "       self.b = \"I am a Base\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "adjusted-basic",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Derived(Base):\n",
    "   def __init__(self):\n",
    "       self.d = \"I am a Derived\"\n",
    "       super(Derived, self).__init__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "secondary-praise",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I am a Base\n"
     ]
    }
   ],
   "source": [
    "d = Derived()\n",
    "print(d.b)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "qualified-shipping",
   "metadata": {},
   "source": [
    "Failure to do that `super(...)` business would result in an\n",
    "`AttributeError` when trying to print `d.b`.\n",
    "\n",
    "However, the classes that you are writing are **not** straightforward\n",
    "Python classes!  While `super(...)` *may* work, it is not\n",
    "guaranteed.  To quote the `pybind11` documentation:\n",
    "\n",
    "> Note that a direct __init__ constructor should be called,\n",
    "> and super() should not be used. For simple cases of linear inheritance,\n",
    "> super() may work, but once you begin mixing Python and C++ multiple inheritance,\n",
    "> things will fall apart due to differences between Python’s MRO and C++’s mechanisms.\n",
    "> \n",
    "The source of this quote is the `pybind11` manual section on\n",
    "[Classes](<https://pybind11.readthedocs.io/en/stable/advanced/classes.html#>).\n",
    "\n",
    "The examples that follow take their advice.\n",
    "\n",
    "#### `attrs`\n",
    "\n",
    "The [attrs](<https://www.attrs.org>) package provides a library\n",
    "of convenient class decorators that significantly reduce the amount\n",
    "of \"boiler plate\" code needed to write Python\n",
    "classes.  Some of the examples below will use this library.  You must\n",
    "*not* use `slots=True`!  The `superclasses` already use the slots mechanism,\n",
    "and it seems difficult for a derived class to also use slots and properly\n",
    "initialize the base.\n",
    "\n",
    "Other than this caveat regarding slots, we enthusiastically recommend\n",
    "`attrs`.\n",
    "\n",
    "#### Custom decorators\n",
    "\n",
    "Speaking of decorators, we provide several that make writing genetic\n",
    "value types a bit easier.  You will see concrete examples using them\n",
    "later.\n",
    "\n",
    "```{py:decorator} fwdpy11.custom_genetic_value_decorators.genetic_value_is_trait_default_clone(ndim=1)\n",
    "\n",
    "This decorator is a *class* that adds the ability\n",
    "for custom types to be passed to classes derived from\n",
    ":class:`fwdpy11.DiploidGeneticValue`.\n",
    "\n",
    "Because this decorator is a class, you need the ``()``\n",
    "to apply it.\n",
    "\n",
    "This decorator is required because of the very different\n",
    "object models of C++ and Python.\n",
    "See `here <https://github.com/pybind/pybind11/issues/1049>`_\n",
    "for some of the gory details.\n",
    "\n",
    ".. py:method:: __init__\n",
    "\n",
    "    :param ndim: The dimensionality of the model\n",
    "    :type ndim: int\n",
    "\n",
    "    The default is ``ndim=1``.\n",
    "\n",
    "```\n",
    "\n",
    "```{py:decorator} fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone\n",
    "\n",
    "This is the analog to\n",
    ":attr:`fwdpy11.custom_genetic_value_decorators.genetic_value_is_trait_default_clone`\n",
    "for noise objects inheriting from :class:`fwdpy11.GeneticValueNoise`.\n",
    "\n",
    "Unlike its analog, this decorator is a *function* and not a class.\n",
    "\n",
    "```\n",
    "\n",
    "```{py:decorator} fwdpy11.custom_genetic_value_decorators.default_update\n",
    "\n",
    "This decorator adds the following function to ``cls``,\n",
    "which is a no-op function:\n",
    "\n",
    ".. code-block:: python\n",
    "\n",
    "    def update(pop: fwdpy11.DiploidPopulation):\n",
    "        pass\n",
    "\n",
    ".. note::\n",
    "\n",
    "    In the future, we will experiment with adding a C++\n",
    "    implementation of this no-op.\n",
    "\n",
    "```\n",
    "\n",
    "(python-gvalue-to-fitness)=\n",
    "\n",
    "### Genetic value to fitness maps\n",
    "\n",
    "Here, one derives a new class from {class}`fwdpy11.GeneticValueIsTrait`.\n",
    "The basic form of such a class is:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "import fwdpy11\n",
    "import fwdpy11.custom_genetic_value_decorators\n",
    "\n",
    "\n",
    "@fwdpy11.custom_genetic_value_decorators.genetic_value_is_trait_default_clone()\n",
    "class MyGeneticValueToFitness(fwdpy11.GeneticValueIsTrait):\n",
    "    def __init__(self):\n",
    "        # Do NOT call super()!\n",
    "        fwdpy11.GeneticValueIsTrait.__init__(self)\n",
    "\n",
    "    def __call__(self, data: fwdpy11.DiploidGeneticValueToFitnessData) -> float:\n",
    "        pass\n",
    "\n",
    "    def update(pop: fwdpy11.DiploidPopulation) -> None:\n",
    "        pass\n",
    "\n",
    "```\n",
    "\n",
    "If the object does not need to manage an internal state that changes over time,\n",
    "we can simplify a bit with another decorator:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "import fwdpy11\n",
    "import fwdpy11.custom_genetic_value_decorators\n",
    "\n",
    "\n",
    "@fwdpy11.custom_genetic_value_decorators.default_update\n",
    "@fwdpy11.custom_genetic_value_decorators.genetic_value_is_trait_default_clone()\n",
    "class MyGeneticValueToFitness(fwdpy11.GeneticValueIsTrait):\n",
    "    def __init__(self):\n",
    "        # Do NOT call super()!\n",
    "        fwdpy11.GeneticValueIsTrait.__init__(self)\n",
    "\n",
    "    def __call__(self, data: fwdpy11.DiploidGeneticValueToFitnessData) -> float:\n",
    "        pass\n",
    "\n",
    "```\n",
    "\n",
    "Two complete examples come from the test suite.  The first example reimplements\n",
    "{class}`fwdpy11.GSS` in Python.  The second example implements a model\n",
    "where the optimum changes to a new value each generation. (Note that\n",
    "the second model requires that {func}`numpy.random.seed` be called elsewhere\n",
    "so that results are reproducible.)\n",
    "\n",
    "```{literalinclude} ../../tests/pygss.py\n",
    ":lines: 19-\n",
    "\n",
    "```\n",
    "\n",
    "#### The data type\n",
    "\n",
    "The type passed into the `__call__` function is:\n",
    "\n",
    "```{py:class} fwdpy11.DiploidGeneticValueToFitnessData\n",
    "\n",
    ".. versionadded:: 0.9.0\n",
    "\n",
    "This class supports the buffer protocol, which exposes the\n",
    "genetic values array.  The most efficient access will\n",
    "be via a :class:`memoryview`.\n",
    "\n",
    "Instances of this class have the following attributes:\n",
    "\n",
    ".. py:attribute:: offspring_metadata\n",
    "\n",
    "    An instance of :class:`fwdpy11.DiploidMetadata`, giving you\n",
    "    access to those fields that have been assigned to the offspring.\n",
    "    (This is a *copy* of the metadata from the C++ side.)\n",
    "\n",
    ".. py:attribute:: offspring_metadata_index\n",
    "\n",
    "    A 64 bit integer that gives the location (index) of\n",
    "    ``offspring_metadata`` in :attr:`fwdpy11.DiploidPopulation.diploid_metadata`.\n",
    "    This index is useful in the event of mass migrations via copies,\n",
    "    which can cause a mismatch between :attr:`fwdpy11.DiploidMetadata.label`\n",
    "    and this value.\n",
    "\n",
    ".. py:attribute:: parent1_metadata\n",
    "\n",
    "    The first parent's metadata (:class:`fwdpy11.DiploidMetadata`).\n",
    "\n",
    ".. py:attribute:: parent2_metadata\n",
    "\n",
    "    The second parent's metadata (:class:`fwdpy11.DiploidMetadata`).\n",
    "\n",
    "```\n",
    "\n",
    "### Genetic value \"noise\"\n",
    "\n",
    "A custom \"noise\" class inherits from {class}`fwdpy11.GeneticValueNoise`.\n",
    "A minimal implementation has the following form:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "@fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone\n",
    "class MyGeneticValueNoise(fwdpy11.GeneticValueNoise):\n",
    "    def __call__(self, data: fwdpy11.DiploidGeneticValueNoiseData) -> float:\n",
    "        pass\n",
    "\n",
    "    def update(pop: fwdpy11.DiploidPopulation) -> None:\n",
    "        pass\n",
    "\n",
    "```\n",
    "\n",
    "If your model can allow a no-op `update` function, you may apply\n",
    "the decorator\n",
    "{attr}`fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone`.\n",
    "\n",
    "A simple example from the test suite implements Gaussian noise with mean zero\n",
    "and standard deviation `0.1`:\n",
    "\n",
    "```{literalinclude} ../../tests/pynoise.py\n",
    ":lines: 19-\n",
    "\n",
    "```\n",
    "\n",
    "See {func}`fwdpy11.gsl_ran_gaussian_ziggurat` for details on that function.\n",
    "\n",
    "#### The data type\n",
    "\n",
    "```{py:class} fwdpy11.DiploidGeneticValueNoiseData\n",
    "\n",
    ".. versionadded:: 0.9.0\n",
    "\n",
    "Instances of this class have the following attributes:\n",
    "\n",
    ".. py:attribute:: rng\n",
    "\n",
    "    The simulation's random number generation, an\n",
    "    instance of :class:`fwdpy11.GSLrng`\n",
    "\n",
    ".. py:attribute:: offspring_metadata\n",
    "\n",
    "    An instance of :class:`fwdpy11.DiploidMetadata`, giving you\n",
    "    access to those fields that have been assigned to the offspring.\n",
    "    (This is a *copy* of the metadata from the C++ side.)\n",
    "\n",
    ".. py:attribute:: offspring_metadata_index\n",
    "\n",
    "    A 64 bit integer that gives the location (index) of\n",
    "    ``offspring_metadata`` in :attr:`fwdpy11.DiploidPopulation.diploid_metadata`.\n",
    "    This index is useful in the event of mass migrations via copies,\n",
    "    which can cause a mismatch between :attr:`fwdpy11.DiploidMetadata.label`\n",
    "    and this value.\n",
    "\n",
    ".. py:attribute:: parent1_metadata\n",
    "\n",
    "    The first parent's metadata (:class:`fwdpy11.DiploidMetadata`).\n",
    "\n",
    ".. py:attribute:: parent2_metadata\n",
    "\n",
    "    The first parent's metadata (:class:`fwdpy11.DiploidMetadata`).\n",
    "\n",
    "```\n",
    "\n",
    "### New genetic value models\n",
    "\n",
    "#### The abstract base class\n",
    "\n",
    "The ABC type is {class}`fwdpy11.PyDiploidGeneticValue`:\n",
    "\n",
    "```{py:class} fwdpy11.PyDiploidGeneticValue\n",
    "\n",
    ".. py:method:: __init__()\n",
    "\n",
    "    :param ndim:\n",
    "    :type ndim: int\n",
    "    :param genetic_value_to_fitness:\n",
    "    :type genetic_value_to_fitness: fwdpy11.GeneticValueIsTrait or None\n",
    "    :param noise:\n",
    "    :type noise: fwdpy11.GeneticValueNoise or None\n",
    "\n",
    ".. py:method:: calculate_gvalue\n",
    "    :abstractmethod:\n",
    "\n",
    "    :param data: Input data\n",
    "    :type data: fwdpy11.PyDiploidGeneticValueData\n",
    "    :returns: The value to be stored in the offspring's\n",
    "              :attr:`fwdpy11.DiploidMetadata.g`\n",
    "    :rtype: float\n",
    "\n",
    ".. py:method:: genetic_value_to_fitness\n",
    "\n",
    "    :param data: Input data\n",
    "    :type data: fwdpy11.DiploidGeneticValueToFitnessData\n",
    "\n",
    "    :returns: fitness\n",
    "    :rtype: float\n",
    "\n",
    "    .. note::\n",
    "\n",
    "        This function does not need to be defined\n",
    "        by derived classes most of the time.\n",
    "        The default behavior is to apply\n",
    "        the instance of :class:`fwdpy11.GeneticValueToFitnessMap`\n",
    "        stored by the instance of :class:`fwdpy11.DiploidGeneticValue`\n",
    "        (or :class:`fwdpy11.PyDiploidGeneticValue`).  Defining\n",
    "        this function in a derived class skips calling held instance\n",
    "        in favor of the derived class implementation.\n",
    "        In general, one only needs to derive this class for models\n",
    "        where either individual genetic values depend on genotypes of the\n",
    "        rest of the population. See :ref:`here <more-complex-gvalue-models>`.\n",
    "\n",
    "\n",
    ".. py:method:: update\n",
    "\n",
    "    :param pop: The population being simulated\n",
    "    :type pop: fwdpy11.DiploidPopulation\n",
    "    :rtype: None\n",
    "\n",
    "    .. note::\n",
    "\n",
    "        A default implementation can be defined using\n",
    "        the decorator\n",
    "        :attr:`fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone`.\n",
    "\n",
    "```\n",
    "\n",
    "#### Outline of user-defined classes\n",
    "\n",
    "The form of a custom genetic value type inherits from the ABC\n",
    "described above and would look something like this:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "class MyGeneticValueObject(fwdpy11.PyDiploidGeneticValue):\n",
    "    def __init__(self, *args):\n",
    "        # Do not call super()!\n",
    "        fwdpy11.PyDiploidGeneticValue.__init__(...)\n",
    "\n",
    "    def calculate_gvalue(self, data: fwdpy11.PyDiploidGeneticValueData) -> float:\n",
    "        pass\n",
    "\n",
    "    def update(self, pop: fwdpy11.DiploidPopulation) -> None:\n",
    "        pass\n",
    "\n",
    "```\n",
    "\n",
    "#### Method requirements\n",
    "\n",
    "{class}`fwdpy11.PyDiploidGeneticValue` *does* provide a default `update`\n",
    "function that updates the genetic-value-to-fitness-map and noise objects.\n",
    "\n",
    "If your model can allow a no-op `update` function, you may apply\n",
    "the decorator\n",
    "{attr}`fwdpy11.custom_genetic_value_decorators.genetic_value_noise_default_clone`.\n",
    "\n",
    "The `genetic_value_to_fitness` function must completely replace the\n",
    "functionality of an instance of {class}`fwdpy11.GeneticValueIsTrait`.\n",
    "See {ref}`here <python-gvalue-to-fitness>` for details.\n",
    "\n",
    "The `calculate_gvalue` function is more complex because it has more\n",
    "responsibilities:\n",
    "\n",
    "1. It must return a value that will populate the genetic value\n",
    "   field of the offspring's metadata ({attr}`fwdpy11.DiploidMetadata.g`).\n",
    "2. It also needs to populate an array of {class}`float` that represents\n",
    "   the genetic values for all trait dimensions. See {ref}`here <python_genetic_value_dimensions>`.\n",
    "   For the common case of a one-dimensional genetic value,\n",
    "   you populate element `0` of this array with value that you\n",
    "   will return from this function. The \"genetic values array\" is\n",
    "   accessible via the Python buffer protocol.\n",
    "   The next section shows a concrete example.\n",
    "\n",
    ":::{note}\n",
    "\n",
    "The steps described in `2` are populating\n",
    "{attr}`fwdpy11.DiploidGeneticValue.genetic_values`.  In the\n",
    "future, we may provide read-write access to that field, which\n",
    "would constitute an API change.  However, the path of least\n",
    "resistance in the short term was to implement the current approach.\n",
    "\n",
    ":::\n",
    "\n",
    "#### Example: strictly additive effects on a trait\n",
    "\n",
    "Here is an example of a strict additive model of a trait,\n",
    "taken from the test suite. You'll notice that we are accessing\n",
    "arrays on the `C++` using Python {class}`memoryview` objects.\n",
    "This is a *very* efficient method, and much faster\n",
    "than creating a temporary, non-owning {class}`numpy.ndarray`.\n",
    "\n",
    "```{literalinclude} ../../tests/pyadditive.py\n",
    ":lines: 20-\n",
    "\n",
    "```\n",
    "\n",
    "If you want to account for the heterozygous effect of mutations, then the\n",
    "most explicit approach is probably to use {class}`collections.Counter`:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "def calculate_gvalue(self, data):\n",
    "    s = 0.0\n",
    "    c = Counter()\n",
    "    pop = data.pop\n",
    "    dip = data.diploids[data.offspring_metadata.label]\n",
    "    for g in [dip.first, dip.second]:\n",
    "        for k in pop.haploid_genomes[g].smutations:\n",
    "            m = pop.mutations[k]\n",
    "            c.update([(m.pos, m.s, m.h)])\n",
    "    for i, j in c.items():\n",
    "        if j == 1:  # Aa\n",
    "            s += i[1] * i[2]\n",
    "        else:  # aa\n",
    "            # This is a hard-coded \"scaling\"\n",
    "            # of 2\n",
    "            s += 2.0 * i[1]\n",
    "    # Use a memory view to update\n",
    "    # the genetic values array\n",
    "    memoryview(data)[0] = s\n",
    "    return s\n",
    "\n",
    "```\n",
    "\n",
    "For additive cases like this, though, you'll get better performance\n",
    "with {func}`fwdpy11.additive_effects`.\n",
    "\n",
    "(more-complex-gvalue-models)=\n",
    "\n",
    "#### More complex scenarios (such as social interactions)\n",
    "\n",
    "Some models are not easily implemented with the genetic value +\n",
    "genetic value to fitness map + noise trio of types defined here.\n",
    "The reason has to do with how these types are applied during a simulation,\n",
    "which is described in more detail {ref}`here <gvalue-technical-details>`, and\n",
    "in {ref}`this section <gvalue-simulation-behavior>` specifically.\n",
    "\n",
    "#### The data types\n",
    "\n",
    "```{eval-rst}\n",
    ".. class:: fwdpy11.PyDiploidGeneticValueData\n",
    "\n",
    ".. versionadded:: 0.9.0\n",
    "\n",
    "This class supports the buffer protocol, which exposes the\n",
    "genetic values array.  The most efficient access will\n",
    "be via a :class:`memoryview`.\n",
    "\n",
    "Instances of this class have the following attributes:\n",
    "\n",
    ".. py:attribute:: rng\n",
    "\n",
    "    The simulation's random number generation, an\n",
    "    instance of :class:`fwdpy11.GSLrng`\n",
    "\n",
    ".. py:attribute:: pop\n",
    "\n",
    "    The population, an instance of :class:`fwdpy11.DiploidPopulation`\n",
    "\n",
    ".. py:attribute:: offspring_metadata\n",
    "\n",
    "    The offspring's :class:`fwdpy11.DiploidMetadata`\n",
    "\n",
    ".. py:attribute:: parent1_metadata\n",
    "\n",
    "    The first parent's :class:`fwdpy11.DiploidMetadata`\n",
    "\n",
    ".. py:attribute:: parent2_metadata\n",
    "\n",
    "    The second parent's :class:`fwdpy11.DiploidMetadata`\n",
    "\n",
    ".. py:attribute:: offspring_metadata_index\n",
    "\n",
    "    A 64 bit integer that gives the location (index) of\n",
    "    ``offspring_metadata`` in :attr:`fwdpy11.DiploidPopulation.diploid_metadata`.\n",
    "    This index is useful in the event of mass migrations via copies,\n",
    "    which can cause a mismatch between :attr:`fwdpy11.DiploidMetadata.label`\n",
    "    and this value.\n",
    "```\n",
    "\n",
    "(msprime-subtleties)=\n",
    "\n",
    "## Subtleties of starting/finishing tree sequences using `msprime`\n",
    "\n",
    "Some care is required starting a simulation from a tree sequence (see {ref}`here <precapitation>`) or finishing a simulation with one (see {ref}`here <recapitation>`).\n",
    "\n",
    "If simulating with discrete recombination positions (see {ref}`here <geneticmapunit>`), it may be important to realize that the current stable releases of `msprime` model continuous recombination breakpoints.\n",
    "A future `msprime` release will support discretized positions.\n",
    "\n",
    "When simulating with genome lengths longer than unity (1.0), determining the correct recombination rate for `msprime` needs additional work.\n",
    "The examples shown {ref}`here <recapitation>` and {ref}`here <precapitation>` show the following method for determining the `msprime` recombination rate:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "recombination_rate = rho / Ne / 4\n",
    "\n",
    "```\n",
    "\n",
    "That code implies a genome length of 1.0.\n",
    "A more general example, for a genome of length `L` would look like:\n",
    "\n",
    "```{code-block} python\n",
    "\n",
    "recombination_rate = rho / Ne / 4 / L\n",
    "\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "md:myst",
   "text_representation": {
    "extension": ".md",
    "format_name": "myst"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "source_map": [
   11,
   119,
   141,
   145,
   153,
   157,
   165,
   170,
   183,
   187,
   189,
   193,
   196,
   200,
   206,
   412,
   418,
   425,
   428
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}